---
title: ""
output: 
  github_document:
    html_preview: false
---

## Characterizing the between-host transmission of SARS-CoV-2 minor-variants in a linked cluster of infections
___
```{r setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)

## ==== Important file paths ==== ##

preprint.filepath =  "../../config/data/preprint_samples.csv" # Samples from pre-print
runselector.filepath =  "../../config/data/PRJNA610428.csv" # Samples from Bioproject
samples.filepath =  "../../config/samples.csv" # Samples that ran analysis
annotations.filepath = "../../config/data/sars_cov_2_annot.csv" # SARS-CoV-2 gene locations
multiqc.filepath = "../../config/data/multiQC-data.csv" # Data formatted from MultiQC
variant.filepath = "../../results/variant/variants.csv" # Combined Variant Calls
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "kableExtra", "gridExtra", "grid", "UpSetR", "ggridges")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

### Project Overview

  The influence of genetic diversity accumulated within an infected individual on the global evolutionary dynamics of SARS-CoV-2 is unknown. Characterizing this dynamic requires an estimate of the amount of genetic variation transferred between individuals during an infection. Here, we leverage a unique dataset to measure this bottleneck. Deep-sequencing of RNA collected from a linked cluster of infections on an isolated fishing vessel enables the observation of the transmission of minor variants between individuals. We propose using this data to provide the highest resolution estimate of the inter-host transmission bottleneck for SARS-CoV-2 to date. 

### Methods

#### Cohort

There was an outbreak on a fishing boat in Seattle. Individuals (120/122) were screened for SARS-CoV-2 RNA by RT-qPCR and seroconversion by the Abbot IgG assay before departure and after the return of the vessel. Although all were negative after the initial testing, an outbreak on the boat occurred. A total of 104/122 crew members had viral RNA or seroconverted upon return to shore (84% Attack Rate). 

#### Sequencing Data

From the 104 reportedly infected crew members, there are 39 metagenomic RNA-sequencing runs available. Libraries were sequenced on a 1x75 bp Illumina NextSeq run. A median of 509,551 sequencing reads were obtained for each sample. The data is publicly available at NCBI BioProject [`PRJNA610428`](https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=2&WebEnv=MCID_5f344f743a76fab9ab22e42b&o=acc_s%3Aa). 

#### Variant Calling

To measure intra-host variation in each sample, the data was processed through a variant calling pipeline. The raw sequencing reads were processed using `Fastp` to trim adaptor sequences, low-quality bases, filter low-quality reads, and trim poly-A tails. Processed reads were subsequently aligned to a reference (NC_045512) using either `BWA` or `STAR`, and variants were called using either `Varscan2` or `Lofreq`. 

### Analysis

#### Included Samples

```{r Included Samples, message=F, warning=F, echo=F}

## ==== Read-in sample data and get SRA accessions ==== ##

# Pre-print samples
samples = read_csv(preprint.filepath)
# Bioproject samples from Run Selector
bioproject = read_csv(runselector.filepath)
# Join the data-frames by 'Isolate' (derived from Strain)
combined_samples = left_join(samples, bioproject, by="Isolate") %>%
  select(Isolate, Run, LibraryLayout, Bases) %>% 
  mutate(Virus = "SARS2", Host = "none", Source = "public", Path = "[]", Experiment = "Cluster")
# From manual inspection, some samples are missing from the Bioproject
missing_samples = combined_samples %>% filter(is.na(Run))
# Samples that aren't missing
final_samples = combined_samples %>% filter(!is.na(Run))

## ==== Download the final samples used to run the analysis ==== ##
# All Samples
analysis_samples = read_csv(samples.filepath)

```

The paper has a supplementary table (Supplementary Table 2.) that includes the strain names for each individual and the GISAID ID for the consensus, but not the SRA accession. I used this information to search the Bioproject ([`PRJNA610428`](https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=2&WebEnv=MCID_5f344f743a76fab9ab22e42b&o=acc_s%3Aa)) for the relevant samples. This Bioproject contains all of the samples sequenced by the UW Virology Department (~1000 Runs). Of the 39 samples referenced in the paper, there are `r nrow(final_samples)` that had corresponding Runs deposited in this project and `r nrow(missing_samples)` that did not. As a quasi-control, I randomly sampled an equal number of sequencing runs from the rest of the UW bioproject that were done on the same sequencer, were also single-ended, and were within one standard deviation of the mean library size for the transmission cluster Runs. 

```{r Data Formatting, message=FALSE, warning=FALSE, echo=FALSE}
## ==== Read in the variant calls ==== ##
variant.df = read_csv(variant.filepath)

## ====  Import genome annotations ==== ##
annotations.df = read_csv(annotations.filepath)

# Get the interval positions for these genes 
ORF1ab = annotations.df %>% dplyr::filter(gene == "Orf1ab")
S = annotations.df %>% dplyr::filter(gene == "Spike")
ORF3a = annotations.df %>% dplyr::filter(gene == "ORF3a")
E = annotations.df %>% dplyr::filter(gene == "Envelope")
M = annotations.df %>% dplyr::filter(gene == "Membrane")
ORF6 = annotations.df %>% dplyr::filter(gene == "ORF6")
ORF7a = annotations.df %>% dplyr::filter(gene == "ORF7a")
ORF8 = annotations.df %>% dplyr::filter(gene == "ORF8")
N = annotations.df %>% dplyr::filter(gene == "Nucleocapsid")
ORF10 = annotations.df %>% dplyr::filter(gene == "ORF10")

# Rename genes for continuity
annotations.df = annotations.df %>% 
  mutate(gene = case_when(gene == "Orf1ab" ~ "ORF1ab",
                          gene == "Spike" ~ "Spike",
                          gene == "ORF3a" ~ "ORF3a",
                          gene == "Envelope" ~ "Envelope",
                          gene == "Membrane" ~ "Membrane",
                          gene == "ORF6" ~ "ORF6",
                          gene == "ORF7a" ~ "ORF7a",
                          gene == "ORF8" ~ "ORF8",
                          gene == "Nucleocapsid" ~ "Nucelocapsid",
                          gene == "ORF10" ~ "ORF10"))

# Add gene names to the variant.df
variant.df = variant.df %>% 
  mutate(Gene = case_when(POS < ORF1ab$start ~ "3'UTR",
                          POS >= ORF1ab$start & POS <= ORF1ab$end  ~ "ORF1ab",
                          POS >= S$start & POS <= S$end  ~ "Spike",
                          POS >= ORF3a$start & POS <= ORF3a$end  ~ "ORF3a",
                          POS >= E$start & POS <= E$end  ~ "Envelope",
                          POS >= M$start & POS <= M$end  ~ "Membrane",
                          POS >= ORF6$start & POS <= ORF6$end  ~ "ORF6",
                          POS >= ORF7a$start & POS <= ORF7a$end  ~ "ORF7a",
                          POS >= ORF8$start & POS <= ORF8$end  ~ "ORF8",
                          POS >= N$start & POS <= N$end  ~ "Nucelocapsid",
                          POS >= ORF10$start & POS <= ORF10$end  ~ "ORF10",
                          POS > ORF10$end ~ "5'UTR")) %>% 
  select(!c("LibraryLayout", "Virus", "Host", "Source", "Path")) %>% 
  mutate(ALT = ifelse(ALT == "True", "T", ALT)) %>% 
  mutate(Var_Type = case_when(nchar(REF) == 1 & nchar(ALT) == 1 ~ "SNP",
                              nchar(REF) > 1 ~ "Deletion",
                              nchar(ALT) > 1 ~ "Insertion")) %>% 
  mutate(SNP = paste0(REF,POS,ALT), Mutation = paste0(REF, ">", ALT))

## ==== Get the Experiment that corresponds to each Accession ==== ##
experiment.table = variant.df %>% 
  select(Accession, Experiment) %>% 
  distinct()

```


#### Quality Contol

Data for the quality control portion of the analysis was collated by the tool `multiqc`. It consists of statistics produced by `samtools`, `STAR`, `fastp`, and `fastQC.` I've also included the coverage over each sample in this part of the analyis.

**The percentage of reads that mapped to SARS-CoV-2.**

```{r Percentage Mapped, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=8, fig.align='center'}
# Experiment Colors
experiment.colors = c("#FFC20A", "#0C7BDC")

# Get the % of reads mapped 
multiqc.df = read_csv(multiqc.filepath) %>% left_join(., experiment.table, by = "Accession")

multiqc.df %>% 
  ggplot(aes(x = reorder(Accession, -Mapped), y = Mapped*100, fill = Experiment)) +
    geom_bar(stat = "identity", position = position_dodge()) +
    facet_wrap(~Aligner) + 
    scale_fill_manual(values = experiment.colors) +
    xlab("Accession") +
    ylab("Percent") +
    ggtitle("Percentage of Reads Mapped") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=13,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1))

```

There is a tremendous amount of variation in the proportion of reads that correctly map. It's fairly well distributed between the control and cluster samples, but on average the cluster samples have a lower mapping rate. Interestingly, there is a bigger effect of aligner choice (`STAR` v. `BWA`) on mapping rate in the cluster than the control. I wonder if this speaks to some underlying issues with sample quality that isn't well summarized by mapping rate?

**The mean coverage over the SARS-CoV-2 reference genome for each sample.**

```{r Average Coverage, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=8, fig.align='center'}
## ==== Import genome coverage files ==== ##
# Path to coverage data
coverage.path = "../../results/split/"

# Get paths to *.table output
coverage_file_paths = dir(path = coverage.path, pattern = "\\.coverage$", full.names = TRUE, recursive = TRUE)

# Make empty df to store coverage info
coverage.df = data.frame()

# Loop over all file paths
for (i in 1:length(coverage_file_paths)){

  # get path to file
  coverage.path = coverage_file_paths[i]

  # extract metadata from filepath
  coverage.info = strsplit(basename(coverage.path), "[.]")[[1]]
  coverage.name = coverage.info[1]
  coverage.aligner = coverage.info[2]

  # Read in the coverage file
  sample.coverage = read_tsv(coverage_file_paths[i],
                             col_names = c("Ref", "POS", "Coverage")) %>%
    mutate(Accession = coverage.name) %>%
    mutate(Aligner = coverage.aligner) %>%
    sample_n(., 10000, replace = F)

  # Combine the dataframes
  coverage.df = rbind(sample.coverage, coverage.df)
}

coverage.df %>%
  group_by(Accession, Aligner) %>%
  summarize(mean_coverage = mean(Coverage)) %>%
  left_join(., experiment.table, by = "Accession") %>%
  ggplot(aes(x = reorder(Accession, -mean_coverage), y = mean_coverage, fill = Experiment)) +
    geom_bar(stat = "identity", position = position_dodge()) +
    facet_wrap(~Aligner) +
    scale_fill_manual(values = experiment.colors) +
    xlab("Accession") +
    ylab("Mean Coverage") +
    ggtitle("Average Coverage per Sample" ) +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=13,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1))




```

Coverage is more consistent than the mapping rate between different aligners. The same relationship holds that the controls have higher coverage than the cluster samples on average. 

**The ratio of transitions to transversions in each sample.**

```{r Ts/Tv Ratio, message=FALSE, warning=FALSE, echo=FALSE, fig.width=7, fig.height=5, fig.align='center'}

variant.df %>% 
  filter(Var_Type == "SNP", Aligner == "BWA", Caller =="lofreq") %>% 
  mutate(TsTv = case_when(Mutation == "C>T" ~ "Transition",
                          Mutation == "T>C" ~ "Transition",
                          Mutation == "A>G" ~ "Transition",
                          Mutation == "G>A" ~ "Transition",
                          TRUE ~ "Tansversion")) %>% 
  group_by(Accession, TsTv) %>% 
  summarise(Count = n()) %>% 
  spread(key = TsTv, value = Count) %>% 
  mutate(TsTv = Transition/Tansversion) %>% 
  left_join(., experiment.table, by = "Accession") %>% 
  ggplot(aes(x = reorder(Accession, -TsTv), y = TsTv, fill = Experiment)) +
    geom_bar(stat = "identity", position = position_dodge()) +
    geom_hline(yintercept = 1, size = 2, col = "red", linetype = 2) +
    scale_fill_manual(values = experiment.colors) +
    xlab("Accession") +
    ylab("Ts/Tv Ratio") +
    ggtitle("Transition v. Transversions" ) +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=14,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1))

```

In theory, transitions are much more common than transversions biologically, however Illumina sequencing errors tend to bais towards transversions. Above, I plotted the ratio transitions (Ts) to transversions (Tv) for each sample and colored them by the experiment (cluster v. control). The cluster samples seem to have a higher ratio of transitions. I'm not sure what to make of this yet. This is just for BWA and Lofreq. From here on, I'll just use that aligner-caller combo. 

**The concordance between the pipeline for annotated SNPs.**

```{r Pipeline Comparison, , message=FALSE, warning=FALSE, echo=FALSE, fig.width=7, fig.height=5, fig.align='center'}

variant.df %>% 
  mutate(Identifier = paste(Accession, SNP, sep = "-"), Approach = paste(Aligner, Caller, sep = "-")) %>% 
  select(Identifier, Approach) %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "Approach", values_from = "Tally", values_fill = 0) %>% 
  as.data.frame(row.names = Identifier) %>%
  upset(., sets = c("BWA-varscan",
                          "BWA-lofreq",
                          "STAR-varscan",
                          "STAR-lofreq"), sets.bar.color = "#56B4E9",
      order.by = "freq", empty.intersections = "on", text.scale = 1.5)

variant.df %>% 
  filter(AF >= 0.2) %>% 
  mutate(Identifier = paste(Accession, SNP, sep = "-"), Approach = paste(Aligner, Caller, sep = "-")) %>% 
  select(Identifier, Approach) %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "Approach", values_from = "Tally", values_fill = 0) %>% 
  as.data.frame(row.names = Identifier) %>%
  upset(., sets = c("BWA-varscan",
                          "BWA-lofreq",
                          "STAR-varscan",
                          "STAR-lofreq"), sets.bar.color = "#56B4E9",
      order.by = "freq", empty.intersections = "on", text.scale = 1.5)

```

The difference in concordance between all variants (first plot) and just variants that have an allele frequency above 2% (second plot) is huge. It's worth taking this into consideration in down stream analyses and perhaps filtering out alleles with frequencies less than 2%. 

#### Mutational Spectra

**The ratio of Synonymous/Non-Synonymous/Missense mutations - all variants and only minor variants.**

```{r Effect Distribution, message=FALSE, warning=FALSE, echo=FALSE, fig.width=8, fig.height=8, fig.align='center'}

# Fix the factor levels
variant.df$Effect = factor(variant.df$Effect, levels = c("Synonymous", "Missense", "Nonsense")) 

# Color scheme
effect_colors = c("#13a13b","#e2e627","#990505")

variant.df %>% 
  filter(Var_Type == "SNP" & !is.na(Effect)) %>% 
  group_by(Aligner, Caller, Effect) %>% 
  summarize(Count = n())  %>%
  mutate(Frequnecy = Count / sum(Count)) %>% 
  ggplot(aes(x = Effect, y = Frequnecy, fill = Effect)) +
    geom_bar(stat = "identity") + 
    facet_grid(rows = vars(Caller), cols = vars(Aligner)) +
    geom_text(aes(label = paste0(round(Frequnecy, 4) * 100, "%")), size = 6, hjust = 0.5, vjust = -2, position = "stack", family="mono") +
    xlab("Percent (%)") +
    ylab("Effect") +
    ggtitle("SNP Effect") +
    ylim(0, 0.8) +
    scale_fill_manual(values = effect_colors) +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) 

```

The ratio of mutational effects is very similar between the four combinations and consistent with other studies that I've seen. 

**The distribution of nucleotide changes - all variants and only minor variants.**

```{r Mutational Spectra, message=FALSE, warning=FALSE, echo=FALSE, fig.width=8, fig.height=8, fig.align='center'}

variant.df %>% 
  filter(Var_Type == "SNP") %>% 
  ggplot(aes(x = Mutation)) +
    geom_bar(stat = "count") + 
    facet_grid(rows = vars(Caller), cols = vars(Aligner)) +
    xlab("SNP Count") +
    ylab("Mutations") +
    ggtitle("Mutational Spectra - All Variants") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=35, hjust=1))

variant.df %>% 
  filter(Var_Type == "SNP", AF <= 0.5) %>% 
  ggplot(aes(x = Mutation)) +
    geom_bar(stat = "count") + 
    facet_grid(rows = vars(Caller), cols = vars(Aligner)) +
    xlab("SNP Count") +
    ylab("Mutations") +
    ggtitle("Mutational Spectra - Minor Variants") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=35, hjust=1))
  
```

Similar to the Ts/Tv plot above, I wanted to see the mutational spectra for minor and major variants. Overall, there seems to be a slight bias towards C>T transtions, which is also consistent with other studies, but it's not as apparent in only the minor variants. 

**The distribution of nucleotide changes across genes, normalized to size - all variants and only minor variants.**

```{r Gene Distribution, message=FALSE, warning=FALSE, echo=FALSE, fig.width=8, fig.height=8, fig.align='center'}

variant.df %>% 
  filter(Var_Type == "SNP"  & !is.na(Effect)) %>% 
  group_by(Aligner, Caller, Gene) %>% 
  summarize(Count = n())  %>%
  left_join(., annotations.df, by = c("Gene"="gene")) %>% 
  filter(Gene != "3'UTR", Gene != "5'UTR") %>% 
  mutate(length = end - start) %>% 
  mutate(Normalized = Count / length) %>% 
  ggplot(aes(x = Gene, y = Normalized)) +
    geom_bar(stat = "identity") + 
    facet_grid(rows = vars(Caller), cols = vars(Aligner)) +
    xlab("Gene") +
    ylab("SNPs/Kb") +
    ggtitle("SNPs per Gene") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=35, hjust=1))
  

```

#### Shared Variation

**Dirstribution of minor variants.**

```{r Minor SNP Distribution, message=FALSE, warning=FALSE, echo=FALSE, fig.width=8, fig.height=8, fig.align='center'}

variant.df %>% 
  filter(Var_Type == "SNP", Aligner == "BWA", Caller == "lofreq", AF <= 0.5) %>% 
  group_by(SNP, Experiment) %>% 
  summarize(Count = n()) %>% 
  filter(Count >= 2) %>% 
  ggplot(aes(x = Count, y = Experiment, fill = Experiment)) + 
    geom_density_ridges() +
    scale_fill_manual(values = experiment.colors) +
    xlab("# of Individuals Sharing SNP") +
    ylab("Experiment") +
    ggtitle("Distribution of Shared Minor SNPs") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed"))


```

**Upset plot of shared Minor SNPs.**

```{r Minor SNP Uspet, message=FALSE, warning=FALSE, echo=FALSE, fig.width=8, fig.height=8, fig.align='center'}

variant.df %>% 
  filter(Var_Type == "SNP", Aligner == "BWA", Caller == "lofreq", AF <= 0.5) %>% 
  group_by(SNP, Experiment) %>% 
  summarize(Count = n()) %>% 
  filter(Count >= 2) %>% 
  select(SNP, Experiment) %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "Experiment", values_from = "Tally", values_fill = 0) %>% 
  as.data.frame(row.names = SNP) %>%
  upset(., sets = c("Control", "Cluster"), sets.bar.color = "#56B4E9",
      order.by = "freq", empty.intersections = "on", text.scale = 1.5)
  

shared.minor.variants = variant.df %>% 
  filter(Var_Type == "SNP", Aligner == "BWA", Caller == "lofreq", AF <= 0.5) %>% 
  group_by(SNP, Experiment) %>% 
  summarize(Count = n()) %>% 
  filter(Count >= 2) %>% 
  select(SNP, Experiment) %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "Experiment", values_from = "Tally", values_fill = 0) %>% 
  mutate(Sum = Control + Cluster) %>% 
  filter(Sum == 2) %>% 
  pull(SNP)


# variant.df[which(variant.df$SNP %in% shared.minor.variants),] %>% 
#   filter(Aligner == "BWA", Caller == "lofreq") %>% 
#   arrange(SNP) %>% 
#   view()
```

** Correlation between frequency of overlap and allele frequency in minor SNPs. **

```{r Correlation between AF and Overlap, message=FALSE, warning=FALSE, echo=FALSE, fig.width=8, fig.height=8, fig.align='center'}

variant.df %>% 
  filter(Var_Type == "SNP", Aligner == "BWA", Caller == "lofreq", AF <= 0.5) %>% 
  group_by(SNP, Experiment) %>%
  summarize(Count = n(), MeanAF = mean(AF)) %>% 
  filter(Count >= 2) %>% 
  ggplot(aes(x = MeanAF, y = Count, col = Experiment)) + 
    geom_point(size = 2) +
    scale_color_manual(values = experiment.colors) +
    xlab("Mean Allele Frequency") +
    ylab("# of Times Shared") +
    ggtitle("Correlation between AF and Overlap") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed"))

  
```


```{r Fishers Exact Test, message=F, warning=F, echo=F}

shared.cluster = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq") %>% 
  filter(Experiment == "Cluster", AF <= 0.5) %>% 
  group_by(SNP) %>% 
  summarize(Count = n()) %>% 
  filter(Count >= 2) %>% 
  nrow(.)

shared.control = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq") %>% 
  filter(Experiment == "Control", AF <= 0.5) %>% 
  group_by(SNP) %>% 
  summarize(Count = n()) %>% 
  filter(Count >= 2) %>% 
  nrow(.)

unshared.cluster = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq") %>% 
  filter(Experiment == "Cluster", AF <= 0.5) %>% 
  group_by(SNP) %>% 
  summarize(Count = n()) %>% 
  filter(Count == 1) %>% 
  nrow(.)

unshared.control = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq") %>% 
  filter(Experiment == "Control", AF <= 0.5) %>% 
  group_by(SNP) %>% 
  summarize(Count = n()) %>% 
  filter(Count == 1) %>% 
  nrow(.)

matrix(c(shared.cluster,unshared.cluster,shared.control,unshared.control), ncol=2)

fisher.test(matrix(c(shared.cluster,unshared.cluster,shared.control,unshared.control), ncol=2))

```

```{r Pairwise Comparison, message=F, warning=F, echo=F}

varint.cluster = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq", Experiment == "Cluster") 

varint.cluster.wide = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq", Experiment == "Cluster") %>% 
  select(SNP, Experiment, Accession) %>% 
  mutate(count = 1) %>%
  spread(Accession, count, fill = 0) 

variant.pairwise.cluster = varint.cluster %>% 
    pull(Accession) %>%
    unique() %>%
    combn(2) %>% 
    t() %>%
    as_data_frame() %>%
    mutate(count = map2_int(V1, V2, ~sum(apply(varint.cluster.wide %>% select(.x, .y), 1, sum) == 2))) %>% 
    mutate(Experiment = "Cluster")


varint.control = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq", Experiment == "Control") 

varint.control.wide = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq", Experiment == "Control") %>% 
  select(SNP, Experiment, Accession) %>% 
  mutate(count = 1) %>%
  spread(Accession, count, fill = 0) 

variant.pairwise.control = varint.control %>% 
    pull(Accession) %>%
    unique() %>%
    combn(2) %>% 
    t() %>%
    as_data_frame() %>%
    mutate(count = map2_int(V1, V2, ~sum(apply(varint.control.wide %>% select(.x, .y), 1, sum) == 2))) %>% 
    mutate(Experiment = "Control")

variant.pairwise = rbind(variant.pairwise.cluster, variant.pairwise.control)

variant.pairwise %>% 
  ggplot(aes(x = count, fill = Experiment)) +
    geom_histogram(alpha = 0.5)  +
    scale_fill_manual(values = experiment.colors) +
    xlab("Number of Pairwise Shared Variants") +
    ylab("Count of Pairs") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed"))


variant.pairwise %>% 
  group_by(Experiment) %>% 
  summarize(mean = mean(count), median = median(count))
```





