---
title: "1. Coverage Analysis"
author: "Will Hannon"
date: "1/29/2021"
output: html_document
---

The goal of this notebook is to look at the coverage of reads over the SARS-CoV-2 genome for each deep-sequencing sample and replicate. The percentage of each genome covered by an adequate depth (100X) is assessed. Additionally, the pattern of read coverage is exmained. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "scales")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```


```{r Paths to data, echo = FALSE}
# Data from `samtools stats`
stats.data = "../../results/coverage/coverage.stats" 

#  Data from `samtools depth` summed for any position >= 100X / NR
average.depth.data = "../../results/coverage/merged.average.depth" 

# Data from `samtools depth` averaged over 50 BP bins 
depth.of.coverage.data = "../../results/coverage/merged.depth" 

# Annotation of SARS-CoV-2 gene positions
annotation.data = "../../config/data/sars_cov_2_annot.csv" 

```


```{r Setup}
# Colors for plotting
experiment.colors = c("#000000", "#7a7a7a")

# Make the figure directory if it doesn't exist
dir.create("../../results/figures", showWarnings = TRUE, recursive = FALSE, mode = "0777")

```


```{r Format data, warning=FALSE, message=FALSE}

# Samtools stats
stats.df = read_tsv(stats.data) %>% 
  separate(filename, into = c("X1", "X2", "X3", "Run", "X4"), sep = "/") %>% 
  select(!starts_with('X')) %>% 
  separate(Run, into = c("Accession", "Replicate"), remove = F)
# Average Depth
average.depth.df = read.table(average.depth.data, header = T) %>% 
  separate(Accession, into = c("Accession", "Replicate")) %>% 
  mutate(Run = paste0(Accession, Replicate))
# Coverage Mins
depth.of.coverage.df = read.table(depth.of.coverage.data, header = T) %>% 
  separate(Accession, into = c("Accession", "Replicate")) %>% 
  mutate(Run = paste0(Accession, Replicate))

```

## Completeness 

This was the first exclusion criteria for samples in the experiment. They had to have at least 100 reads covering 80% of the genomes. Samples with less completeness were excluded. We know that depth/completeness impact how accuratley we can call variants. Samples with less completeness than this fail comparisons between the replicates anyhow. 

```{r Coverage Completeness, message=FALSE, warning=FALSE, fig.width=11, fig.height=5, fig.align='center'}

# What is the percent of the genome covered by more than 200 reads?
average.depth.df %>% 
  mutate(Accession = paste0("Crewmember-", Accession)) %>% 
  ggplot(aes(x = reorder(Accession, -Percent), y = Percent/100, fill = Replicate)) + 
    geom_bar(stat = "identity", position = position_dodge()) + 
    scale_fill_manual(values = experiment.colors, name = "Replicate") +
    scale_y_continuous(labels = percent) + 
    xlab("Sample") +
    ylab("Completeness") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=25, hjust=1)) +
    theme(plot.margin = unit(c(1.5,1.5,1.5,1.5), "cm"))

ggsave("../../results/figures/coverage_completeness_100X.svg", device = svg(), width = 11, height = 5)

```

## Coverage Distribution

Next, I wanted to look at the pattern of coverage over the genome. To do this, I counted the number of reads in 50 BP bins over the genome using `samtools depth` and `AWK`.  

The coverage pattern is pretty erratic since this isn't amplicon sequencing - this is to be expected. Red samples are colored as those that don't have more than 80% of the genome covered by more than 100 reads. 

```{r Failed QC}

# Low coverage samples - These are the samples that failed our quality control. The coverage was less than 100X in 20% of the genome.
low.coverage.samples = average.depth.df %>% filter(Percent < 80) %>% pull(Accession)
low.coverage.runs = average.depth.df %>% filter(Percent < 80) %>% pull(Run)

print(paste0("There are ", length(low.coverage.samples), " low coverage samples: "))
print(low.coverage.samples)

```

This leads to the exclusion of 6/24 crewmembers: `10027, 10114, 10028, 10138, 10039, and 10107`

```{r Coverage Distribution, message=FALSE, warning=FALSE, fig.width=15, fig.height=6, fig.align='center'}

# Read in the SARS-CoV-2 gene annotations
annotation.df = read_csv(annotation.data)

# Generate positions for lables 
label = annotation.df %>% 
  mutate(POS = (start+end)/2) %>% 
  mutate(Depth = -(1200/2)) %>% 
  select(gene, POS, Depth) 

# Get the interval positions for these genes 
ORF1ab = annotation.df %>% dplyr::filter(gene == "Orf1ab")
Spike = annotation.df %>% dplyr::filter(gene == "Spike")
ORF3a = annotation.df %>% dplyr::filter(gene == "ORF3a")
Envelope = annotation.df %>% dplyr::filter(gene == "Envelope")
Membrane = annotation.df %>% dplyr::filter(gene == "Membrane")
ORF6 = annotation.df %>% dplyr::filter(gene == "ORF6")
ORF7a = annotation.df %>% dplyr::filter(gene == "ORF7a")
ORF8 = annotation.df %>% dplyr::filter(gene == "ORF8")
Nucleocapsid = annotation.df %>% dplyr::filter(gene == "Nucleocapsid")
ORF10 = annotation.df %>% dplyr::filter(gene == "ORF10")


# Excluding the last 100 or so nucleotides which is a poly-A run. 
depth.of.coverage.df %>% 
  mutate(`Low Coverage` = ifelse(Run %in% low.coverage.runs, "Yes", "No"),
         POS =(Start+Stop)/2) %>% 
  filter(Start < 29850) %>%
  ggplot(aes(x = POS, y = Depth, group = Run, col = `Low Coverage`))+
    geom_line() + 
    scale_color_manual(values = c("#080000","#b80000")) +
    xlab("Position") +
    ylab("Depth") +
    scale_x_continuous(breaks=c(0,5000,10000,15000,20000,25000,30000)) +
    annotate("rect", xmin = 0, xmax = ORF1ab[[2]]-1, ymin = -1200, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate("rect", xmin = ORF1ab[[2]], xmax = ORF1ab[[3]], ymin = -1200, ymax = -.01 ,
               alpha = .2, fill = "dodgerblue2", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][1], y = label[[3]][1], label = label[[1]][1], family = "mono") +
    annotate("rect", xmin = Spike[[2]], xmax = Spike[[3]], ymin =  -1200, ymax = -.01 ,
             alpha = .2, fill = "yellow3", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][2], y = label[[3]][2], label = label[[1]][2], family = "mono") +
    annotate("rect", xmin = ORF3a[[2]], xmax = ORF3a[[3]], ymin =  -1200, ymax = -.01 ,
             alpha = .2, fill = "#FF7F00", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][3], y = label[[3]][3], label = "3a", family = "mono") +
    annotate("rect", xmin = Envelope[[2]], xmax = Envelope[[3]], ymin =  -1200, ymax = -.01 ,
             alpha = .2, fill =  "#6A3D9A", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][4], y = label[[3]][4], label = "E", family = "mono") +
    annotate("rect", xmin = Membrane[[2]], xmax = Membrane[[3]], ymin =  -1200, ymax = -.01 ,
             alpha = .2, fill =  "green4", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][5], y = label[[3]][5], label = "M", family = "mono") +
    annotate("rect", xmin = ORF6[[2]], xmax = ORF6[[3]], ymin = -1200, ymax = -.01 ,
             alpha = .2, fill = "#E31A1C", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][6], y = label[[3]][6], label = "6", family = "mono") +
    annotate("rect", xmin = ORF7a[[2]], xmax = ORF7a[[3]], ymin =  -1200, ymax = -.01 ,
             alpha = .2, fill = "#FB9A99", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][7], y = label[[3]][7], label = "7a", family = "mono") +
    annotate("rect", xmin = ORF8[[2]], xmax = ORF8[[3]], ymin =  -1200, ymax = -.01 ,
             alpha = .2, fill = "orchid1", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][8], y = label[[3]][8], label = "8", family = "mono") +
    annotate("rect", xmin = Nucleocapsid[[2]], xmax = Nucleocapsid[[3]], ymin =  -1200, ymax = -.01 ,
             alpha = .2, fill = "maroon", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][9], y = label[[3]][9], label = "N", family = "mono") +
    annotate("rect", xmin = ORF10[[2]], xmax = ORF10[[3]], ymin =  -1200, ymax = -.01 ,
             alpha = .2, fill =  "skyblue2", col = "black", size = .1) +
    annotate("rect", xmin = ORF10[[3]]+1, xmax = 29903, ymin = -1200, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate(geom = "text", x = -200, y = label[[3]][1], label = "5'", family = "mono") +
    annotate(geom = "text", x = 30200, y = label[[3]][1], label = "3'", family = "mono") +
    #annotate(geom = "text", x = label[[2]][10], y = label[[3]][10], label = "10", family = "mono") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(aspect.ratio = .25)  +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed"),
        strip.text.x = element_blank()) 


ggsave("../../results/figures/single_plot_coverage_dist.svg", device = svg(), width = 11, height = 5)

  
```

Below is the same plot as above, except that it's split between the individual samples. The maximum number of reads is capped at 1000 reads and a line is drawn at 100X coverage, which was our requirement for minimum coverage. 

```{r Faceted Coverage Distribution, message=FALSE, warning=FALSE, fig.width=20, fig.height=10, fig.align='center'}

# Generate positions for lables 
label = annotation.df %>% 
  mutate(POS = (start+end)/2) %>% 
  mutate(Depth = -(100/2)) %>% 
  select(gene, POS, Depth) 

# Excluding the last 100 or so nucleotides which is a poly-A run. 
depth.of.coverage.df %>% 
  mutate(`Low Coverage` = ifelse(Run %in% low.coverage.runs, "Yes", "No")) %>% 
  mutate(Accession = paste0("Crewmember-", Accession)) %>% 
  filter(Start < 29850) %>% 
  mutate(Depth = ifelse(Depth > 1000, 1000, Depth)) %>% 
  ggplot(aes(x = (Start+Stop)/2, y = Depth, group = Run, col = `Low Coverage`))+
    geom_line() + 
    geom_hline(yintercept = 100, col = "blue", linetype = 2) + 
    facet_wrap(~Accession, ncol = 3) + 
    scale_color_manual(values = c("#080000","#b80000")) +
    xlab("Genomic Position") +
    ylab("Depth") +
    scale_x_continuous(breaks=c(0,5000,10000,15000,20000,25000,30000)) +
    annotate("rect", xmin = 0, xmax = ORF1ab[[2]]-1, ymin = -100, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate("rect", xmin = ORF1ab[[2]], xmax = ORF1ab[[3]], ymin = -100, ymax = -.01 ,
               alpha = .2, fill = "dodgerblue2", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][1], y = label[[3]][1], label = label[[1]][1], family = "mono") +
    annotate("rect", xmin = Spike[[2]], xmax = Spike[[3]], ymin =  -100, ymax = -.01 ,
             alpha = .2, fill = "yellow3", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][2], y = label[[3]][2], label = label[[1]][2], family = "mono") +
    annotate("rect", xmin = ORF3a[[2]], xmax = ORF3a[[3]], ymin =  -100, ymax = -.01 ,
             alpha = .2, fill = "#FF7F00", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][3], y = label[[3]][3], label = "3a", family = "mono") +
    annotate("rect", xmin = Envelope[[2]], xmax = Envelope[[3]], ymin =  -100, ymax = -.01 ,
             alpha = .2, fill =  "#6A3D9A", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][4], y = label[[3]][4], label = "E", family = "mono") +
    annotate("rect", xmin = Membrane[[2]], xmax = Membrane[[3]], ymin = -100, ymax = -.01 ,
             alpha = .2, fill =  "green4", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][5], y = label[[3]][5], label = "M", family = "mono") +
    annotate("rect", xmin = ORF6[[2]], xmax = ORF6[[3]], ymin = -100, ymax = -.01 ,
             alpha = .2, fill = "#E31A1C", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][6], y = label[[3]][6], label = "6", family = "mono") +
    annotate("rect", xmin = ORF7a[[2]], xmax = ORF7a[[3]], ymin =  -100, ymax = -.01 ,
             alpha = .2, fill = "#FB9A99", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][7], y = label[[3]][7], label = "7a", family = "mono") +
    annotate("rect", xmin = ORF8[[2]], xmax = ORF8[[3]], ymin = -100, ymax = -.01 ,
             alpha = .2, fill = "orchid1", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][8], y = label[[3]][8], label = "8", family = "mono") +
    annotate("rect", xmin = Nucleocapsid[[2]], xmax = Nucleocapsid[[3]], ymin =  -100, ymax = -.01 ,
             alpha = .2, fill = "maroon", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][9], y = label[[3]][9], label = "N", family = "mono") +
    annotate("rect", xmin = ORF10[[2]], xmax = ORF10[[3]], ymin =  -100, ymax = -.01 ,
             alpha = .2, fill =  "skyblue2", col = "black", size = .1) +
    annotate("rect", xmin = ORF10[[3]]+1, xmax = 29903, ymin = -100, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate(geom = "text", x = -200, y = label[[3]][1], label = "5'", family = "mono") +
    annotate(geom = "text", x = 30200, y = label[[3]][1], label = "3'", family = "mono") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(legend.position="none") +
    theme(axis.text.x=element_text(angle=25, hjust=1)) 
    
ggsave("../../results/figures/faceted_plot_coverage_dist.svg", device = svg(), width = 20, height = 20)

```


