---
title: "Lab Meeting Figures"
author: "Will Hannon"
date: "9/27/2020"
output: html_document
---

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)

## ==== Important file paths ==== ##
multiqc.filepath = "../../config/data/2020-09-30_multiQC-data.csv" # QC data (manually prepared from MultiQC) from replicates
coverage.filepath = "../../results/split/merged.bedgraph" # Data on the average depth of coverage in 50 KB bins
variant.filepath = "../../results/variant/variants.csv" # Combined variant calls from varscan, lofreq, STAR, and BWA
pileup.filepath = "../../results/pileup/raw-variants.csv" # Variants identified from pileup files
annotations_path = "../../../SARS-CoV-2_cell_culture_deepseq/data/ANNOT/" 
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "kableExtra", "gridExtra", "grid", "UpSetR", "ggridges", "ggpubr", "scales", "gplots", "ggrepel")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

### Overview 

This notebook contains all of the figrues for primary lab meeting 09-30-2020

```{r Data Import, echo=F, message=FALSE, warning=FALSE, cache=T}

## =============================================== ##
#
# This chunk contains code for importing and 
#  formatting the data used in the analysis.
#
## =============================================== ##

## =========== Import the variant calls ========== ##

# Filtering out everything but SNPs from Varscan/BWA
variant.df = read_csv(variant.filepath)  %>%
      mutate(Type = case_when(nchar(REF) < nchar(ALT) ~ "Insertion",
                              nchar(REF) > nchar(ALT) ~ "Deletion",
                              nchar(REF) == nchar(ALT) ~ "SNP")) %>% 
      filter(Aligner == "BWA", Caller == "varscan", Type == "SNP") %>% 
      separate(col = Accession, into = c("Accession", "Replicate", "Percent")) %>% 
      mutate(
             SNP = paste0(REF, POS, ALT),
             Percent = ifelse(is.na(Percent), 100, Percent))

# Reshape the variant data to combine replicates
combined.df = full_join(filter(variant.df, Replicate == 1),
                        filter(variant.df, Replicate == 2),
                        by = c("Accession", "Percent", "POS", "REF", "ALT"),
                        suffix = c(".one", ".two")) %>% 
              select(!c("Replicate.one", "Replicate.two", "Aligner.one", "Aligner.two")) %>% 
              mutate_each(funs(replace(., which(is.na(.)), 0))) # Full Join, replace NA with 0

## =========== Import the run quality information ========== ##

multiqc.df = read.csv(multiqc.filepath) %>% 
  separate(col = Sample, into = c("Accession", "Aligner", "Sample"), sep = "\\.") %>% 
  filter(Sample == "virus") %>% 
  separate(col = Accession, into = c("Accession", "Replicate", "Percent"), sep = "-") %>% 
  mutate(Replicate = ifelse(is.na(Replicate), 1, Replicate)) %>% 
  mutate(Depth =(reads_mapped*average_length) / 29903) %>% 
  filter(Aligner == "BWA")  %>% 
  mutate(Percent = ifelse(is.na(Percent), 100, Percent))


## =========== Pileup statistics ========== ##

# pileup.df = read_csv(pileup.filepath) %>% 
#   select(!starts_with("Ins") & !starts_with("Del")) %>% 
#   separate(col = Accession, into = c("Accession", "Replicate", "Percent"), sep = "-") %>% 
#   mutate(Percent = ifelse(is.na(Percent), 100, Percent)) %>% 
#   mutate(Replicate = ifelse(is.na(Replicate), 1, Replicate)) %>% 
#   rename(REF = "Reference Base", CONS = "Consensus Base", POS = "Position") %>% 
#   mutate(`A Total` = `A For` + `A Rev`,
#          `C Total` = `C For` + `C Rev`,
#          `T Total` = `T For` + `T Rev`,
#          `G Total` = `G For` + `G Rev`) %>% 
#   pivot_longer(cols = c("A For", "A Rev",
#                         "C For", "C Rev",
#                         "G For", "G Rev",
#                         "T For", "T Rev",
#                         "A Total", "C Total",
#                         "T Total", "G Total"),
#                names_to = "ALT", values_to = "Count") %>% 
#   separate(col = ALT, into = c("ALT", "Strand")) %>% 
#   mutate(Strand = case_when(Strand == "For" ~ "+",
#                             Strand == "Rev" ~  "-",
#                             Strand == "Total" ~ "+/-"),
#          SNP = paste0(REF, POS, ALT)) %>% 
#   mutate(AF = Count/Depth, Type = case_when(ALT != REF ~ "SNP",
#                                             ALT == REF ~ "REF")) %>% 
#   filter(Type == "SNP") %>% 
#   filter(AF != 0, Strand == "+/-")
# 
# 
# saveRDS(pileup.df, file = "pileup_df.RDS") 

pileup.df = readRDS("pileup_df.RDS")
  
# Reshape the variant data to combine replicates
combined.pileup.df = full_join(filter(pileup.df, Replicate == 1),
                                filter(pileup.df, Replicate == 2),
                                by = c("Accession", "Percent", "POS", "REF", "ALT", "CONS", "Strand"),
                                suffix = c(".one", ".two")) %>% 
  select(!c("Replicate.one", "Replicate.two")) %>% 
  filter(Strand == "+/-", AF.one > 0 & AF.two > 0) %>% 
  rename(DP.one = "Depth.one", DP.two = "Depth.two") %>% 
  mutate_each(funs(replace(., which(is.na(.)), 0))) # Full Join, replace NA with 0


## =========== Import the variant calls, keep indels ========== ##
# Keep only Varscan/BWA
variant.df.indels = read_csv(variant.filepath)  %>%
      mutate(Type = case_when(nchar(REF) < nchar(ALT) ~ "Insertion",
                              nchar(REF) > nchar(ALT) ~ "Deletion",
                              nchar(REF) == nchar(ALT) ~ "SNP")) %>% 
      filter(Aligner == "BWA", Caller == "varscan") %>% 
      separate(col = Accession, into = c("Accession", "Replicate", "Percent")) %>% 
      mutate(Replicate = ifelse(is.na(Replicate), 1, Replicate),
             SNP = paste0(REF, POS, ALT),
             Percent = ifelse(is.na(Percent), 100, Percent))

# Reshape the variant data to combine replicates
combined.df.indels = full_join(filter(variant.df.indels, Replicate == 1),
                        filter(variant.df.indels, Replicate == 2),
                        by = c("Accession", "Percent", "POS", "REF", "ALT"),
                        suffix = c(".one", ".two")) %>% 
              select(!c("Replicate.one", "Replicate.two", "Aligner.one", "Aligner.two")) %>% 
              mutate_each(funs(replace(., which(is.na(.)), 0))) # Full Join, replace NA with 0

```


```{r Consensus, echo=F, warning=F, message=F}

combined.pileup.df %>% 
  filter(Percent == "100") %>% 
  filter(AF.one >= 0.9 | AF.two >= 0.9) %>% 
  select(SNP = "SNP.one", Accession) %>% 
  group_by(SNP) %>% 
  summarize(Count = n()) %>% 
  arrange(-Count) %>% 
  mutate(Consesus = case_when(Count >= 23 ~ "Yes",
                              Count == 22 ~ "Almost",
                              Count < 22 ~ "No"))  %>% 
  ggplot(aes(x = reorder(SNP, -Count), y = Count, fill = Consesus)) + 
    geom_bar(stat = "identity", position = position_dodge()) +
    scale_fill_manual(values = c("#dfe615", "#757575", "#b51714")) + 
    xlab("SNP") +
    ylab("Number of Samples") +
    ggtitle("Consensus Variants" ) +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=13,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1)) + 
    theme(legend.position = "none")
  
true.consensus = combined.pileup.df %>% 
  filter(Percent == "100") %>% 
  filter(AF.one >= 0.9 | AF.two >= 0.9) %>% 
  select(SNP = "SNP.one", Accession) %>% 
  group_by(SNP) %>% 
  summarize(Count = n()) %>% 
  arrange(-Count) %>% 
  filter(Count > 22) %>% 
  pull(SNP)


combined.pileup.df %>% filter(Percent == "100", SNP.one == "G11083T", Accession == "10136")

```

Interestingly, there is one allele `G11083T` that is absecent from one sample only (`SpID: 10136`). But it's clearly low frequency (7/108 reads and 2/22 reads respectivley), so that's interesting.

```{r Fix Consensus, echo=F, warning=F, message=F}

# Mask the consensus SNPs
combined.df = combined.df[which(!combined.df$SNP.one %in% true.consensus),]
combined.df.indels = combined.df.indels[which(!combined.df.indels$SNP.one %in% true.consensus),]
variant.df = variant.df[which(!variant.df$SNP %in% true.consensus),]
combined.pileup.df = combined.pileup.df[which(!combined.pileup.df$SNP.one %in% true.consensus),]

# Raw variant frequencies with line of best fit and correlation coeff.
combined.df$Percent = factor(combined.df$Percent, levels = c("25", "50", "75", "100"))
combined.pileup.df$Percent = factor(combined.pileup.df$Percent, levels = c("25", "50", "75", "100"))
```

## Figures

```{r Depth between Replicates}

multiqc.df %>% 
  filter(Percent == "100") %>% 
  ggplot(aes(x = reorder(Accession, -reads_mapped), y = reads_mapped, fill = Replicate)) +
    geom_bar(stat = "identity", position = position_dodge()) +
    scale_fill_manual(values = c("#3271a8", "#bf6021")) + 
    scale_y_continuous(label=scientific_format()) +
    xlab("SpID") +
    ylab("Reads Mapped") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=13,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=30, hjust=1))

```


```{r Depth relative to SNPs, echo=F, warning=F, message=F, fig.width=4, fig.height=3, fig.align='center'}

variant.df %>% 
  group_by(Accession, Replicate, Percent) %>% 
  count() %>% 
  full_join(., multiqc.df, by = c("Accession", "Replicate", "Percent")) %>% 
  mutate(Sample = paste(Accession, Replicate)) %>% 
  ggplot(aes(x = as.numeric(as.character(Percent)), y = n, group = Sample)) +
    geom_point() + 
    geom_line() + 
    xlab("Percent of Library Size") +
    ylab("Number of SNPs") +
    scale_x_continuous(breaks = c(25, 50, 75, 100)) + 
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) + 
    theme(legend.position = "none")


```




```{r Coverage and SNPs, echo=F, warning=F, message=F, fig.width=4, fig.height=3, fig.align='center'}

pileup.df %>% 
  #filter(AF > 0.01, Depth > 200, Count > 5) %>% 
  group_by(Accession, Replicate, Percent) %>% 
  filter(Percent == "100") %>% 
  count() %>% 
  full_join(., filter(multiqc.df, Percent == "100"), by = c("Accession", "Replicate")) %>% 
  mutate(Sample = paste(Accession, Replicate)) %>% 
  ggplot(aes(x = reads_mapped, y = n, col = Replicate)) +
    geom_point(size = 3) +
    scale_color_manual(values = c("#3271a8", "#bf6021")) + 
    scale_x_continuous(label=scientific_format()) +  
    xlab("Number of Viral Reads") +
    ylab("Number of SNPs") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) + 
    theme(legend.position = "none")

pileup.df %>% 
  #filter(AF > 0.01, Depth > 200, Count > 5) %>% 
  group_by(Accession, Replicate, Percent) %>% 
  filter(Percent == "100") %>% 
  count() %>% 
  full_join(., filter(multiqc.df, Percent == "100"), by = c("Accession", "Replicate")) %>% 
  mutate(Sample = paste(Accession, Replicate)) %>% 
  ggplot(aes(x = reads_mapped, y = n)) +
    geom_point(size = 3) + 
    xlab("Number of Viral Reads") +
    ylab("Number of SNPs") + 
    scale_x_continuous(label=scientific_format()) +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed"))

pileup.df %>% 
  filter(AF >= 0.001, Depth >= 300) %>% 
  group_by(Accession, Replicate, Percent) %>% 
  filter(Percent == "100") %>% 
  count() %>% 
  full_join(., filter(multiqc.df, Percent == "100"), by = c("Accession", "Replicate")) %>% 
  mutate(Sample = paste(Accession, Replicate)) %>% 
  ggplot(aes(x = reads_mapped, y = n, col = Replicate)) +
    geom_point(size = 3) + 
    xlab("Number of Viral Reads") +
    ylab("Number of SNPs (>= 0.001)") + 
    scale_x_continuous(label=scientific_format()) +
    scale_color_manual(values = c("#3271a8", "#bf6021")) + 
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed"))

pileup.df %>% 
  filter(AF >= 0.01, Depth >= 300) %>% 
  group_by(Accession, Replicate, Percent) %>% 
  filter(Percent == "100") %>% 
  count() %>% 
  full_join(., filter(multiqc.df, Percent == "100"), by = c("Accession", "Replicate")) %>% 
  mutate(Sample = paste(Accession, Replicate)) %>% 
  ggplot(aes(x = reads_mapped, y = n, col = Replicate)) +
    geom_point(size = 3) + 
    xlab("Number of Viral Reads") +
    ylab("Number of SNPs (>= 0.01)") + 
    scale_x_continuous(label=scientific_format()) +
    scale_color_manual(values = c("#3271a8", "#bf6021")) + 
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed"))


pileup.df %>% 
  filter(AF >= 0.05, Depth >= 300) %>% 
  group_by(Accession, Replicate, Percent) %>% 
  filter(Percent == "100") %>% 
  count() %>% 
  full_join(., filter(multiqc.df, Percent == "100"), by = c("Accession", "Replicate")) %>% 
  mutate(Sample = paste(Accession, Replicate)) %>% 
  ggplot(aes(x = reads_mapped, y = n, col = Replicate)) +
    geom_point(size = 3) + 
    xlab("Number of Viral Reads") +
    ylab("Number of SNPs (>= 0.05)") + 
    scale_x_continuous(label=scientific_format()) +
    scale_color_manual(values = c("#3271a8", "#bf6021")) + 
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed"))

```

```{r AF Cov, echo=F, warning=F, message=F, fig.width=4, fig.height=3, fig.align='center'}

## ==================================================================================== ##

identity.df = combined.df %>% mutate(Identity = paste0(Accession, SNP.one))

combined.pileup.df.plotting = combined.pileup.df %>% 
  mutate(Identity = paste0(Accession, SNP.one)) %>% 
  mutate(Shared = case_when(Identity %in% identity.df$Identity ~ "Varscan",
                            !Identity %in% identity.df$Identity ~ "Pileup-Only"))

combined.pileup.df.plotting %>% 
  filter(Percent == "100", Accession == "10118", Shared == "Varscan") %>% 
  ggplot(aes(x = (AF.one), y = (AF.two))) + 
    facet_wrap(~Accession) +
    geom_point(size = 3, col = "#a30000")+
    geom_smooth(method='lm', se = F, col = "black") + 
    stat_cor(method="pearson") +
    scale_color_manual(values = c("#9c9c9c", "#a30000")) + 
    xlab("Allele Frequency Replicate One") +
    ylab("Allele Frequency Replicate Two") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(panel.background = element_rect(fill = NA, color = "black"))


combined.pileup.df.plotting %>% 
  filter(Percent == "100", Accession == "10118", Shared == "Varscan") %>% 
  ggplot(aes(x = log10(AF.one), y = log10(AF.two))) + 
    facet_wrap(~Accession) +
    geom_point(size = 3, col = "#a30000") +
    geom_smooth(method='lm', se = F, col = "black") + 
    stat_cor(method="pearson") +
    scale_color_manual(values = c("#9c9c9c", "#a30000")) + 
    xlab("Allele Frequency Replicate One (Log10)") +
    ylab("Allele Frequency Replicate Two (Log10)") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(panel.background = element_rect(fill = NA, color = "black"))


combined.pileup.df.plotting %>% 
  filter(Percent == "100", Accession == "10118") %>% 
  ggplot(aes(x = log10(AF.one), y = log10(AF.two), col = Shared)) + 
    facet_wrap(~Accession) +
    geom_point(size = 3) +
    geom_smooth(method='lm', se = F, col = "black") + 
    stat_cor(method="pearson") +
    scale_color_manual(values = c("#9c9c9c", "#a30000")) + 
    xlab("Allele Frequency Replicate One (Log10)") +
    ylab("Allele Frequency Replicate Two (Log10)") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(panel.background = element_rect(fill = NA, color = "black"))

```



```{r AF Cov All, echo=F, warning=F, message=F, fig.width=9, fig.height=7, fig.align='center'}

combined.pileup.df.plotting %>% 
  filter(Percent == "100") %>% 
  ggplot(aes(x = log10(AF.one), y = log10(AF.two), col = Shared)) + 
    facet_wrap(~Accession, ncol = 4) +
    geom_point(size = 3) +
    geom_smooth(method='lm', se = F, col = "black") + 
    stat_cor(method="pearson") +
    scale_color_manual(values = c("#9c9c9c", "#a30000")) + 
    xlab("Allele Frequency Replicate One (Log10)") +
    ylab("Allele Frequency Replicate Two (Log10)") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(panel.background = element_rect(fill = NA, color = "black"))

```


```{r AF Cov All 2, echo=F, warning=F, message=F, fig.width=9, fig.height=7, fig.align='center'}

combined.pileup.df.plotting %>% 
  filter(Percent == "100") %>% 
  ggplot(aes(x = (AF.one), y = (AF.two), col = Shared)) + 
    facet_wrap(~Accession, ncol = 4) +
    geom_point(size = 3) +
    geom_smooth(method='lm', se = F, col = "black") + 
    stat_cor(method="pearson") +
    scale_color_manual(values = c("#9c9c9c", "#a30000")) + 
    xlab("Allele Frequency Replicate One") +
    ylab("Allele Frequency Replicate Two") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(panel.background = element_rect(fill = NA, color = "black"))

```


```{r AF Cov All 3, echo=F, warning=F, message=F, fig.width=9, fig.height=7, fig.align='center'}

AF = 0.01
DP = 300

combined.pileup.df %>% 
  filter(Percent == "100") %>% 
  ggplot(aes(x = log10(AF.one), y = log10(AF.two))) + 
    facet_wrap(~Accession, ncol = 4) +
    geom_point(size = 3, alpha = 0.5, col = "#9c9c9c") +
    geom_point(data = filter(combined.pileup.df, DP.one >= DP & DP.two >= DP, AF.one >= AF & AF.two >= AF), 
               mapping = aes(x = log10(AF.one), y = log10(AF.two)), size = 3, col = "#a30000") +
    geom_smooth(data = filter(combined.pileup.df, DP.one >= DP & DP.two >= DP, AF.one >= AF & AF.two >= AF), 
                mapping = aes(x = log10(AF.one), y = log10(AF.two)),
                method='lm', se = F, col = "#a30000") +
    stat_cor(data = filter(combined.pileup.df, DP.one >= DP & DP.two >= DP, AF.one >= AF & AF.two >= AF), 
             mapping = aes(x = log10(AF.one), y = log10(AF.two)),
             method="pearson") +
    geom_hline(yintercept = log10(AF), size = 0.5, linetype = 2) +
    geom_vline(xintercept = log10(AF), size = 0.5, linetype = 2) +
    xlab("Allele Frequency Replicate One") +
    ylab("Allele Frequency Replicate Two") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(panel.background = element_rect(fill = NA, color = "black"))



```


```{r Joined, echo=F, warning=F, message=F, fig.width=9, fig.height=7, fig.align='center'}

good.samples = c("10029", "10040", "10042", "10102", "10118", "10127", "10128", "10129", "10130", "10131")


shared.df = combined.pileup.df %>% 
  filter(Percent == "100") %>% 
  mutate(Quality = case_when(Accession %in% good.samples ~ "Good",
                             !Accession %in% good.samples ~ "Bad")) %>% 
  filter(Quality == "Good",
         DP.one >= DP & DP.two >= DP,
         AF.one >= AF &  AF.two >= AF)

results = data.frame()
joined_results = data.frame()

accessions.to.include = unique(shared.df$Accession)

for (i in 1:length(accessions.to.include)){
  
  for (j in 1:length(accessions.to.include)){
    
    accession.i = accessions.to.include[i]
    accession.j = accessions.to.include[j]
    
    variants.i = filter(shared.df, Accession == accession.i) %>% 
      mutate(AF =(AF.one+AF.two)/2) %>% 
      select(POS, REF, ALT, SNP = "SNP.one", Accession, AF)
    
    variants.j = filter(shared.df, Accession == accession.j) %>% 
      mutate(AF =(AF.one+AF.two)/2) %>% 
      select(POS, REF, ALT, SNP = "SNP.one", Accession, AF)
    
    join.df = full_join(variants.i, variants.j, by = c("POS", "REF", "ALT", "SNP" ), suffix = c(".i", ".j")) %>% 
      mutate_each(funs(replace(., which(is.na(.)), 0))) %>% 
      mutate(Accession.i = if_else(Accession.i == 0, accession.i, Accession.i),
             Accession.j = if_else(Accession.j == 0, accession.j, Accession.j))
  
    dot = join.df$AF.i %*% join.df$AF.j
    
    row = data.frame("accession.i" = join.df$Accession.i, "accession.j" = join.df$Accession.j, "dot" = dot[1,1])
    
    results = rbind(results, row)
    
    joined_results = rbind(joined_results, join.df)
    
  }
  
}

joined_results %>% 
  ggplot(aes(x = log10(AF.i), y = log10(AF.j))) +
    geom_point() +
    geom_smooth(method = "lm", se = F) +
    stat_cor(method="pearson") +
    facet_grid(cols = vars(Accession.i), rows = vars(Accession.j)) +
    xlab("Allele Frequency") +
    ylab("Allele Frequency") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(panel.background = element_rect(fill = NA, color = "black")) +
    theme(axis.text.x=element_text(angle=45, hjust=1)) 




# View(joined_results)
# 
# all.df  = spread(distinct(results), accession.i, dot)[,-1]
# 
# rownames(all.df) = spread(distinct(results), accession.i, dot)[,1]
# 
# all.df = as.matrix(all.df)
# 
# heatmap.2(all.df, scale = "none", col = viridis_pal(), cellnote = round(all.df, 2), main= "All Variants",
#           trace = "none", density.info = "none", cexRow=1,cexCol=1,margins=c(12,8),srtCol=45)

```
```{r Annotations, message=FALSE, warning=FALSE, echo=FALSE}
# Import genome annotations
annot = read_csv(paste0(annotations_path,"sars_cov_2_annot.csv"))

# Generate positions for lables 
label = annot %>% 
  mutate(POS = (start+end)/2) %>% 
  mutate(AF = -(750/2)) %>% 
  select(gene, POS, AF) 

# Get the interval positions for these genes 
ORF1ab = annot %>% dplyr::filter(gene == "Orf1ab")
Spike = annot %>% dplyr::filter(gene == "Spike")
ORF3a = annot %>% dplyr::filter(gene == "ORF3a")
Envelope = annot %>% dplyr::filter(gene == "Envelope")
Membrane = annot %>% dplyr::filter(gene == "Membrane")
ORF6 = annot %>% dplyr::filter(gene == "ORF6")
ORF7a = annot %>% dplyr::filter(gene == "ORF7a")
ORF8 = annot %>% dplyr::filter(gene == "ORF8")
Nucleocapsid = annot %>% dplyr::filter(gene == "Nucleocapsid")
ORF10 = annot %>% dplyr::filter(gene == "ORF10")



```

```{r Shared SNPs, message=FALSE, warning=FALSE, echo=FALSE, fig.width=6, fig.height=2.5, fig.align='center'}

unique.df = joined_results %>% 
  mutate(Same = if_else(Accession.i == Accession.j, "Yes", "No")) %>% 
  filter(AF.i > 0 & AF.j > 0, Same == "No")


unique.count.df = unique.df[!duplicated(t(apply(unique.df, 1, sort))),] %>% 
  group_by(Accession.i, Accession.j) %>% 
  count()

mean(unique.count.df$n) 

unique.minor.count.df = unique.df[!duplicated(t(apply(unique.df, 1, sort))),] %>% 
  filter(AF.i < 0.5 & AF.j < 0.5) %>% 
  group_by(Accession.i, Accession.j) %>% 
  count()

mean(unique.minor.count.df$n) 


consensus.level = unique.df[!duplicated(t(apply(unique.df, 1, sort))),] %>% 
  mutate(meanAF =(AF.i+AF.j)/2) %>% 
  group_by(SNP) %>% 
  summarise(MeanAF = mean(meanAF)) %>% 
  filter(MeanAF >= 0.5) %>% 
  pull(SNP)


minor.level = unique.df[!duplicated(t(apply(unique.df, 1, sort))),] %>% 
  mutate(meanAF =(AF.i+AF.j)/2) %>% 
  group_by(SNP) %>% 
  summarise(MeanAF = mean(meanAF)) %>% 
  filter(MeanAF < 0.5) %>% 
  pull(SNP)


unique.df[!duplicated(t(apply(unique.df, 1, sort))),] %>% 
  mutate(Level = case_when(SNP %in% consensus.level ~ "Consensus",
                          SNP %in% minor.level ~ "Minor")) %>% 
  group_by(REF, POS, ALT, SNP, Level) %>% 
  count() %>% 
  ggplot(aes(x = POS, y = n, col = Level)) + 
    geom_point(size = 3) + 
    geom_hline(yintercept = 45, linetype = 2, size = .5, col = "black")+
    geom_text_repel(aes(label = SNP),
                    size = 3, box.padding = 0.4)  +
    xlab("Position") +
    ylab("Number of Pairs") +
    scale_color_manual(values = c("#2326a8", "#ad1c1c")) +
    scale_x_continuous(breaks=c(0,5000,10000,15000,20000,25000,30000)) +
    annotate("rect", xmin = 0, xmax = ORF1ab[[2]]-1, ymin = -10, ymax = -1.5 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate("rect", xmin = ORF1ab[[2]], xmax = ORF1ab[[3]], ymin = -10, ymax = -1.5 ,
               alpha = .2, fill = "dodgerblue2", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][1], y = -5, label = label[[1]][1], family = "Helvetica") +
    annotate("rect", xmin = Spike[[2]], xmax = Spike[[3]], ymin =  -10, ymax = -1.5 ,
             alpha = .2, fill = "yellow3", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][2], y = -5, label = label[[1]][2], family = "Helvetica") +
    annotate("rect", xmin = ORF3a[[2]], xmax = ORF3a[[3]], ymin =  -10, ymax = -1.5 ,
             alpha = .2, fill = "#FF7F00", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][3], y = -5, label = "3a", family = "Helvetica") +
    annotate("rect", xmin = Envelope[[2]], xmax = Envelope[[3]], ymin =  -10, ymax = -1.5 ,
             alpha = .2, fill =  "#6A3D9A", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][4], y = -5, label = "E", family = "Helvetica") +
    annotate("rect", xmin = Membrane[[2]], xmax = Membrane[[3]], ymin =  -10, ymax = -1.5 ,
             alpha = .2, fill =  "green4", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][5], y = -5, label = "M", family = "Helvetica") +
    annotate("rect", xmin = ORF6[[2]], xmax = ORF6[[3]], ymin = -10, ymax = -1.5 ,
             alpha = .2, fill = "#E31A1C", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][6], y = -5, label = "6", family = "Helvetica") +
    annotate("rect", xmin = ORF7a[[2]], xmax = ORF7a[[3]], ymin =  -10, ymax = -1.5 ,
             alpha = .2, fill = "#FB9A99", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][7], y = label[[3]][7], label = "7a", family = "mono") +
    annotate("rect", xmin = ORF8[[2]], xmax = ORF8[[3]], ymin =  -10, ymax = -1.5 ,
             alpha = .2, fill = "orchid1", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][8], y = -5, label = "8", family = "Helvetica") +
    annotate("rect", xmin = Nucleocapsid[[2]], xmax = Nucleocapsid[[3]], ymin =  -10, ymax = -1.5,
             alpha = .2, fill = "maroon", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][9], y = -5, label = "N", family = "Helvetica") +
    annotate("rect", xmin = ORF10[[2]], xmax = ORF10[[3]], ymin =  -10, ymax = -1.5 ,
             alpha = .2, fill =  "skyblue2", col = "black", size = .1) +
    annotate("rect", xmin = ORF10[[3]]+1, xmax = 29903, ymin = -10, ymax = -1.5 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate(geom = "text", x = -200, y = -5, label = "5'", family = "Helvetica") +
    annotate(geom = "text", x = 30200, y = -5, label = "3'", family = "Helvetica") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(aspect.ratio = .35)  +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) 

  



```



