---
title: "Phylogenetic Analysis"
author: "Will Hannon"
date: "1/11/2021"
output: html_document
---

This notebook contains an analysis of phylogenetic relationship between consensus sequences from the `SARS-CoV-2_Bottleneck` project. The main goals of this notebook are to characterize: 

1. Build an easily interpretable phylogenic tree of avaliable samples.
2. Infere the transmission chain between patients.  

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("ggtree", "tidyverse")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Filepaths, include=FALSE}

## ==== File paths input data ==== ##

# Phylogenetic tree from iQtree
tree.data = "../../results/consensus/merged.consenus.fa.treefile"
# Edit distance
editdistance.data = "../../config/edit_distance.csv"
# Tree data from GISAD
GISAID.tree.data = "../../config/gisaid/phylogeny.aligned.fasta.treefile"
# Tree metadata 
metadata.data = "../../config/gisaid/phylogeny_metadata.csv"

```

For reference, the longest branch is only like 11 substitutions in the 30,000 base genome. 

```{r Phylogeny, message=FALSE, warning=FALSE, fig.width=12, fig.height=11, fig.align='center'}

tree = read.tree(tree.data)
ggtree(tree) + 
  geom_tiplab(color='firebrick', size = 10) 

```

```{r GISAID phylogeny, fig.width=12, fig.height=35, fig.align='center'}


GISAID.tree = read.tree(GISAID.tree.data)
info = as.data.frame(read.csv(metadata.data) %>% 
                       mutate(study = if_else(study == 'internal', 'internal', 'external')) %>% 
                       select(id = "strain",
                              clade = "Nextstrain_clade", 
                              study))

 
ggtree(GISAID.tree, size = 1.5) %<+% info +
  geom_tiplab() +
  geom_label(aes(x=branch, label=node)) +
  geom_tippoint(aes(color = clade), size=2) +
  geom_tippoint(aes(subset=(study=="internal")), size = 3, color = 'red') + 
  scale_color_manual(values=c("pink", "green", "blue", "purple", "black", "orange"))  + 
  theme(legend.position = 'none')


# to.investigate = c("EPI_ISL_570234", "EPI_ISL_570278", "EPI_ISL_570325", "EPI_ISL_570326", "EPI_ISL_570328")
# 
# read.csv(metadata.data) %>%  
#   filter(gisaid_epi_isl %in% to.investigate) %>% 
#   select(strain, gisaid_epi_isl, date, authors)


```

```{r GISAID phylogeny, fig.width=6, fig.height=6, fig.align='center'}
 
ggtree(GISAID.tree, size = 1.5, layout="circular") %<+% info +
  geom_tippoint(aes(color = clade), size=2) +
  geom_tippoint(aes(subset=(study=="internal")), size = 3, color = 'red') + 
  scale_color_manual(values=c("pink", "green", "blue", "purple", "black", "orange"))  + 
  geom_hilight(node = 332, fill = "gold") +
  theme(legend.position = 'none')

ggsave(filename = "../../config/figures/figure-1-C.svg", width = 8, height = 8)

```


```{r Distance Heatmap, message = FALSE, warning = FALSE, fig.width=5, fig.height=5, fig.align='center'}

edit.distance.df = read_csv(editdistance.data) %>% 
  mutate(Sample_1 = paste0("Patient-", Sample_1)) %>% 
  mutate(Sample_2 = paste0("Patient-", Sample_2))

ggplot(edit.distance.df, aes(Sample_1, Sample_2)) +
    geom_tile(aes(fill = Edit_Distance)) + 
    geom_text(aes(label = round(Edit_Distance, 1)), size = 6) +
    scale_fill_gradient(low = "white", high = "red") +
    xlab("") +
    ylab("") +
    theme_minimal(22) +
    theme(legend.position="none") +
    theme(axis.text.x=element_text(angle=90, hjust=1)) 
    

```

