---
title: "SARS-CoV-2 Bottleneck Preliminary Analysis"
author: "Will Hannon"
date: "8/13/2020"
output: html_document
---


```{r setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)

## ==== Important file paths ==== ##

preprint.filepath =  "../../config/data/preprint_samples.csv" # Samples from pre-print
runselector.filepath =  "../../config/data/PRJNA610428.csv" # Samples from Bioproject
variant.filepath = "../../results/variant/variants.csv" # Combined Variant Calls
annotations.filepath = "../../config/data/sars_cov_2_annot.csv" # SARS-CoV-2 gene locations
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}
## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "kableExtra", "gridExtra", "grid", "UpSetR", "ggridges")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```


## Notebook Overview
___

#### Project Aim

This analysis aims to see if there is a substantial transmission of intra-host minor variants between individuals in a large cluster of infections. This information could help characterize the inter-host transmission bottleneck in SARS-CoV-2 with the highest resolution achieved so far.

#### Cohort Overview

There was an outbreak on a fishing boat in Seattle. Individuals (120/122) were screened for SARS-CoV-2 RNA by RT-qPCR and seroconversion by the Abbot IgG assay both before departure and after return. Although all were negative, an outbreak occured nonetheless. A total of 104/122 crew members had viral RNA or seroconverted upon return to shore (84% Attack Rate). 

#### Data Overview 

From the 104 reportedly infected crew memebers there are 39 metagenomic RNA-sequencing runs avaliable. The data is avaliable publically at NCBI BioProject [`PRJNA610428`](https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=2&WebEnv=MCID_5f344f743a76fab9ab22e42b&o=acc_s%3Aa). Metadata for the identity of samples is avaliable in Supplementary Table 2.

The metadata only contains the GISAID-ID for the consensus and the strain. I might be able to use the strain to figure out the SRR# using the metadata on the SRA run selector.

#### Experimental Overview 

Libraries were sequenced on a 1x75 bp Illumina NextSeq run. A median of 509,551 sequencing reads were obtained for each sample.

```{r Samples, message=F, warning=F, echo=F}

## ==== Read-in sample data and get SRA accessions ==== ###
# Pre-print samples
samples = read_csv(preprint.filepath)
# Bioproject samples from Run Selector
bioproject = read_csv(runselector.filepath)
# Join the data-frames by 'Isolate' (derived from Strain)
combined_samples = left_join(samples, bioproject, by="Isolate") %>%
  select(Isolate, Run, LibraryLayout, Bases) %>% 
  mutate(Virus = "SARS2", Host = "none", Source = "public", Path = "[]", Experiment = "Cluster")
# From manual inspection, some samples are missing from the Bioproject
missing_samples = combined_samples %>% filter(is.na(Run))
# Samples that aren't missing
final_samples = combined_samples %>% filter(!is.na(Run))

## ==== Randomly get control samples ==== ###
# How many control samples to get? 
sample.size = nrow(final_samples)
# Libary size window
max = mean(final_samples$Bases) + sd(final_samples$Bases)
min = mean(final_samples$Bases) - sd(final_samples$Bases)
# Get only runs from the bioproject not already included with similar experimental design
control.pool = bioproject[which(!(bioproject$Run %in% final_samples$Run)),] %>% 
   filter(LibraryLayout == "SINGLE", Instrument == 	"NextSeq 500", Bases >= min & Bases <= max)
# Randomly sample the controls
control_samples = sample_n(control.pool, sample.size) %>% 
  select(Isolate, Run, LibraryLayout, Bases)  %>% 
  mutate(Virus = "SARS2", Host = "none", Source = "public", Path = "[]", Experiment = "Control")

# Join the controls and samples into final sample csv
analysis.samples = bind_rows(final_samples, control_samples)

## ==== Write-out the sample data to run the analysis ==== ###
## write.csv(analysis.samples, "../../config/samples.csv", row.names = FALSE)

```

#### Included Samples

The paper has a supplementary table (Supplementary Table 2.) that includes the strain names for each individual and the GISAID ID for the consensus, but not the SRA accession. I used this information to search the Bioproject ([`PRJNA610428`](https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=2&WebEnv=MCID_5f344f743a76fab9ab22e42b&o=acc_s%3Aa)) for the relevant samples. This Bioproject contains all of the samples sequenced by the UW Virology Department (~1000 Runs). Of the 39 samples referenced in the paper, there are `r nrow(final_samples)` that had corresponding Runs deposited in this project and `r nrow(missing_samples)` that did not. As a quasi-control, I randomly sampled an equal number of sequencing runs from the rest of the UW bioproject that were done on the same sequencer, were also single-ended, and were within one standard deviation of the mean library size for the transmission cluster Runs. 

## Analysis
___

#### Overview

I processed all of this data through the `viral-deepseq` pipeline, although I got rid of the de-duplication step. After running things initially with de-deuplication, and looking at the BAM files manually, I found that the majority of reads got removed. Also, for reasons I have yet to determine, a small percentage of the BAM files aligned with STAR failed (threw an error) de-duplication. 

```{r Formatting, message=FALSE, warning=FALSE, echo=FALSE}
## ==== Read in the variant calls ==== ##
variant.df = read_csv(variant.filepath)

## ====  Import genome annotations ==== ##
annotations.df = read_csv(annotations.filepath)

# Get the interval positions for these genes 
ORF1ab = annotations.df %>% dplyr::filter(gene == "Orf1ab")
S = annotations.df %>% dplyr::filter(gene == "Spike")
ORF3a = annotations.df %>% dplyr::filter(gene == "ORF3a")
E = annotations.df %>% dplyr::filter(gene == "Envelope")
M = annotations.df %>% dplyr::filter(gene == "Membrane")
ORF6 = annotations.df %>% dplyr::filter(gene == "ORF6")
ORF7a = annotations.df %>% dplyr::filter(gene == "ORF7a")
ORF8 = annotations.df %>% dplyr::filter(gene == "ORF8")
N = annotations.df %>% dplyr::filter(gene == "Nucleocapsid")
ORF10 = annotations.df %>% dplyr::filter(gene == "ORF10")

# Rename genes for continuity
annotations.df = annotations.df %>% 
  mutate(gene = case_when(gene == "Orf1ab" ~ "ORF1ab",
                          gene == "Spike" ~ "Spike",
                          gene == "ORF3a" ~ "ORF3a",
                          gene == "Envelope" ~ "Envelope",
                          gene == "Membrane" ~ "Membrane",
                          gene == "ORF6" ~ "ORF6",
                          gene == "ORF7a" ~ "ORF7a",
                          gene == "ORF8" ~ "ORF8",
                          gene == "Nucleocapsid" ~ "Nucelocapsid",
                          gene == "ORF10" ~ "ORF10"))

# Add gene names to the variant.df
variant.df = variant.df %>% 
  mutate(Gene = case_when(POS < ORF1ab$start ~ "3'UTR",
                          POS >= ORF1ab$start & POS <= ORF1ab$end  ~ "ORF1ab",
                          POS >= S$start & POS <= S$end  ~ "Spike",
                          POS >= ORF3a$start & POS <= ORF3a$end  ~ "ORF3a",
                          POS >= E$start & POS <= E$end  ~ "Envelope",
                          POS >= M$start & POS <= M$end  ~ "Membrane",
                          POS >= ORF6$start & POS <= ORF6$end  ~ "ORF6",
                          POS >= ORF7a$start & POS <= ORF7a$end  ~ "ORF7a",
                          POS >= ORF8$start & POS <= ORF8$end  ~ "ORF8",
                          POS >= N$start & POS <= N$end  ~ "Nucelocapsid",
                          POS >= ORF10$start & POS <= ORF10$end  ~ "ORF10",
                          POS > ORF10$end ~ "5'UTR")) %>% 
  select(!c("LibraryLayout", "Virus", "Host", "Source", "Path")) %>% 
  mutate(ALT = ifelse(ALT == "True", "T", ALT)) %>% 
  mutate(Var_Type = case_when(nchar(REF) == 1 & nchar(ALT) == 1 ~ "SNP",
                              nchar(REF) > 1 ~ "Deletion",
                              nchar(ALT) > 1 ~ "Insertion")) %>% 
  mutate(SNP = paste0(REF,POS,ALT), Mutation = paste0(REF, ">", ALT))

```

#### Quality Contol

Data for the quality control portion of the analysis was collated by the tool `multiqc`. It consists of stats from `samtools`, `STAR`, `fastp`, and `fastQC.` I've also included the coverage over each sample in this part of the analyis.

How good is the condordance between the different pipelines for SNPs that get called? This excludes InDels, because this will be heavily influenced by the aligner and these aren't the main focus of this analysis. Big difference between 1% and 2% AF in terms of concordance. Very interesting.. 

```{r Pipeline Comparison, , message=FALSE, warning=FALSE, echo=FALSE, fig.width=7, fig.height=5, fig.align='center'}
variant.df %>% 
  filter(AF >= 0.2) %>% 
  mutate(Identifier = paste(Accession, SNP, sep = "-"), Approach = paste(Aligner, Caller, sep = "-")) %>% 
  select(Identifier, Approach) %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "Approach", values_from = "Tally", values_fill = 0) %>% 
  as.data.frame(row.names = Identifier) %>%
  upset(., sets = c("BWA-varscan",
                          "BWA-lofreq",
                          "STAR-varscan",
                          "STAR-lofreq"), sets.bar.color = "#56B4E9",
      order.by = "freq", empty.intersections = "on", text.scale = 1.5)

```

What is the mutational spectra? Looks like there are lots of C>T mutations which maybe indicates some host RNA editing? The difference is less clear if you look only at the minor alleles (> 50%).

```{r Mutational Spectra, message=FALSE, warning=FALSE, echo=FALSE, fig.width=7, fig.height=5, fig.align='center'}

variant.df %>% 
  filter(Var_Type == "SNP") %>% 
  ggplot(aes(x = Mutation)) +
    geom_bar(stat = "count") + 
    facet_grid(rows = vars(Caller), cols = vars(Aligner)) +
    xlab("SNP Count") +
    ylab("Mutations") +
    ggtitle("Mutational Spectra - All Variants") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=35, hjust=1))

variant.df %>% 
  filter(Var_Type == "SNP", AF <= 0.5) %>% 
  ggplot(aes(x = Mutation)) +
    geom_bar(stat = "count") + 
    facet_grid(rows = vars(Caller), cols = vars(Aligner)) +
    xlab("SNP Count") +
    ylab("Mutations") +
    ggtitle("Mutational Spectra - Minor Variants") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=35, hjust=1))
  
```


What's the distribution of effects? Looking at all variants this is very consistent with literature.  

```{r Effect Distribution, message=FALSE, warning=FALSE, echo=FALSE, fig.width=7, fig.height=5, fig.align='center'}

# Fix the factor levels
variant.df$Effect = factor(variant.df$Effect, levels = c("Synonymous", "Missense", "Nonsense")) 

# Color scheme
effect_colors = c("#13a13b","#e2e627","#990505")

variant.df %>% 
  filter(Var_Type == "SNP" & !is.na(Effect)) %>% 
  group_by(Aligner, Caller, Effect) %>% 
  summarize(Count = n())  %>%
  mutate(Frequnecy = Count / sum(Count)) %>% 
  ggplot(aes(x = Effect, y = Frequnecy, fill = Effect)) +
    geom_bar(stat = "identity") + 
    facet_grid(rows = vars(Caller), cols = vars(Aligner)) +
    geom_text(aes(label = paste0(round(Frequnecy, 4) * 100, "%")), size = 6, hjust = 0.5, vjust = -2, position = "stack", family="mono") +
    xlab("Percent (%)") +
    ylab("Effect") +
    ggtitle("SNP Effect") +
    ylim(0, 0.8) +
    scale_fill_manual(values = effect_colors) +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) 

```

What's the distribution over all of the genes? 
```{r Gene Distribution, message=FALSE, warning=FALSE, echo=FALSE, fig.width=7, fig.height=5, fig.align='center'}

variant.df %>% 
  filter(Var_Type == "SNP"  & !is.na(Effect)) %>% 
  group_by(Aligner, Caller, Gene) %>% 
  summarize(Count = n())  %>%
  left_join(., annotations.df, by = c("Gene"="gene")) %>% 
  filter(Gene != "3'UTR", Gene != "5'UTR") %>% 
  mutate(length = end - start) %>% 
  mutate(Normalized = Count / length) %>% 
  ggplot(aes(x = Gene, y = Normalized)) +
    geom_bar(stat = "identity") + 
    facet_grid(rows = vars(Caller), cols = vars(Aligner)) +
    xlab("Gene") +
    ylab("SNPs/Kb") +
    ggtitle("SNPs per Gene") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=35, hjust=1))
  


```


```{r SNP Count, message=FALSE, warning=FALSE, echo=FALSE, fig.width=5, fig.height=4, fig.align='center'}


variant.df %>% 
  filter(Var_Type == "SNP", Aligner == "BWA", Caller == "lofreq", AF <= 0.5) %>% 
  group_by(SNP, Experiment) %>% 
  summarize(Count = n()) %>% 
  filter(Count >= 2) %>% 
  ggplot(aes(x = Count, y = Experiment, fill = Experiment)) + 
    geom_density_ridges()

  



```

```{r}

variant.df %>% 
  filter(Var_Type == "SNP", Aligner == "BWA", Caller == "lofreq", AF <= 0.5) %>% 
  group_by(SNP, Experiment) %>% 
  summarize(Count = n()) %>% 
  filter(Count >= 2) %>% 
  select(SNP, Experiment) %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "Experiment", values_from = "Tally", values_fill = 0) %>% 
  as.data.frame(row.names = SNP) %>%
  upset(., sets = c("Control", "Cluster"), sets.bar.color = "#56B4E9",
      order.by = "freq", empty.intersections = "on", text.scale = 1.5)
  

shared.minor.variants = variant.df %>% 
  filter(Var_Type == "SNP", Aligner == "BWA", Caller == "lofreq", AF <= 0.5) %>% 
  group_by(SNP, Experiment) %>% 
  summarize(Count = n()) %>% 
  filter(Count >= 2) %>% 
  select(SNP, Experiment) %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "Experiment", values_from = "Tally", values_fill = 0) %>% 
  mutate(Sum = Control + Cluster) %>% 
  filter(Sum == 2) %>% 
  pull(SNP)


variant.df[which(variant.df$SNP %in% shared.minor.variants),] %>% 
  filter(Aligner == "BWA", Caller == "lofreq") %>% 
  arrange(SNP) %>% 
  view()
```

```{r}

variant.df %>% 
  filter(Var_Type == "SNP", Aligner == "BWA", Caller == "lofreq", AF <= 0.5) %>% 
  group_by(SNP, Experiment) %>%
  summarize(Count = n(), MeanAF = mean(AF)) %>% 
  filter(Count >= 2) %>% 
  ggplot(aes(x = MeanAF, y = Count, col = Experiment)) + 
    geom_point(size = 2) 

  
```

```{r Fishers Exact Test}

shared.cluster = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq") %>% 
  filter(Experiment == "Cluster", AF <= 0.5) %>% 
  group_by(SNP) %>% 
  summarize(Count = n()) %>% 
  filter(Count >= 2) %>% 
  nrow(.)

shared.control = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq") %>% 
  filter(Experiment == "Control", AF <= 0.5) %>% 
  group_by(SNP) %>% 
  summarize(Count = n()) %>% 
  filter(Count >= 2) %>% 
  nrow(.)

unshared.cluster = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq") %>% 
  filter(Experiment == "Cluster", AF <= 0.5) %>% 
  group_by(SNP) %>% 
  summarize(Count = n()) %>% 
  filter(Count == 1) %>% 
  nrow(.)

unshared.control = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq") %>% 
  filter(Experiment == "Control", AF <= 0.5) %>% 
  group_by(SNP) %>% 
  summarize(Count = n()) %>% 
  filter(Count == 1) %>% 
  nrow(.)

fisher.test(matrix(c(shared.cluster,unshared.cluster,shared.control,unshared.control), ncol=2))

matrix(c(shared.cluster,unshared.cluster,shared.control,unshared.control), ncol=2)

```


```{r}
# Re-do label to refect new y-axis positions
label = annotations.df %>% 
  mutate(POS = (start+end)/2) %>% 
  mutate(AF = (-.01/-.1)/-2) %>% 
  select(gene, POS, AF) 

variant.df %>% 
  filter(Aligner == "BWA" & Caller == "varscan") %>% 
  filter(AF < 0.5) %>% 
  ggplot(aes(x = POS, y = AF, col = Accession, fill = Accession))+
    geom_point(size =2) + 
    geom_bar(stat="identity", width=.01, position = position_identity()) +
    facet_wrap(~Accession) +
    xlab("Position") +
    ylab("Allele Frequency") +
    ggtitle("Distribution of Variants") + 
    scale_x_continuous(breaks=c(0,5000,10000,15000,20000,25000,30000)) +
    annotate("rect", xmin = 0, xmax = ORF1ab[[2]]-1, ymin = -.1, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate("rect", xmin = ORF1ab[[2]], xmax = ORF1ab[[3]], ymin = -.1, ymax = -.01 ,
               alpha = .2, fill = "dodgerblue2", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][1], y = label[[3]][1], label = label[[1]][1], family = "mono") +
    annotate("rect", xmin = S[[2]], xmax = S[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "yellow3", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][2], y = label[[3]][2], label = label[[1]][2], family = "mono") +
    annotate("rect", xmin = ORF3a[[2]], xmax = ORF3a[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "#FF7F00", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][3], y = label[[3]][3], label = "3a", family = "mono") +
    annotate("rect", xmin = E[[2]], xmax = E[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill =  "#6A3D9A", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][4], y = label[[3]][4], label = "E", family = "mono") +
    annotate("rect", xmin = M[[2]], xmax = M[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill =  "green4", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][5], y = label[[3]][5], label = "M", family = "mono") +
    annotate("rect", xmin = ORF6[[2]], xmax = ORF6[[3]], ymin = -.1, ymax = -.01 ,
             alpha = .2, fill = "#E31A1C", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][6], y = label[[3]][6], label = "6", family = "mono") +
    annotate("rect", xmin = ORF7a[[2]], xmax = ORF7a[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "#FB9A99", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][7], y = label[[3]][7], label = "7a", family = "mono") +
    annotate("rect", xmin = ORF8[[2]], xmax = ORF8[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "orchid1", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][8], y = label[[3]][8], label = "8", family = "mono") +
    annotate("rect", xmin = N[[2]], xmax = N[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "maroon", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][9], y = label[[3]][9], label = "N", family = "mono") +
    annotate("rect", xmin = ORF10[[2]], xmax = ORF10[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill =  "skyblue2", col = "black", size = .1) +
    annotate("rect", xmin = ORF10[[3]]+1, xmax = 29903, ymin = -.1, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate(geom = "text", x = -200, y = label[[3]][1], label = "5'", family = "mono") +
    annotate(geom = "text", x = 30200, y = label[[3]][1], label = "3'", family = "mono") +
    #annotate(geom = "text", x = label[[2]][10], y = label[[3]][10], label = "10", family = "mono") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed"),
        strip.text.x = element_blank()) 
  
```

```{r}

varint.cluster = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq", Experiment == "Cluster") 

varint.cluster.wide = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq", Experiment == "Cluster") %>% 
  select(SNP, Experiment, Accession) %>% 
  mutate(count = 1) %>%
  spread(Accession, count, fill = 0) 

variant.pairwise.cluster = varint.cluster %>% 
    pull(Accession) %>%
    unique() %>%
    combn(2) %>% 
    t() %>%
    as_data_frame() %>%
    mutate(count = map2_int(V1, V2, ~sum(apply(varint.cluster.wide %>% select(.x, .y), 1, sum) == 2))) %>% 
    mutate(Experiment = "Cluster")


varint.control = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq", Experiment == "Control") 

varint.control.wide = variant.df %>% 
  filter(Aligner == "BWA", Caller == "lofreq", Experiment == "Control") %>% 
  select(SNP, Experiment, Accession) %>% 
  mutate(count = 1) %>%
  spread(Accession, count, fill = 0) 

variant.pairwise.control = varint.control %>% 
    pull(Accession) %>%
    unique() %>%
    combn(2) %>% 
    t() %>%
    as_data_frame() %>%
    mutate(count = map2_int(V1, V2, ~sum(apply(varint.control.wide %>% select(.x, .y), 1, sum) == 2))) %>% 
    mutate(Experiment = "Control")


variant.pairwise = rbind(variant.pairwise.cluster, variant.pairwise.control)

variant.pairwise %>% 
  ggplot(aes(x = count, fill = Experiment)) +
    geom_histogram(alpha = 0.5)

```



