---
title: "SARS-CoV-2 Bottleneck Preliminary Analysis"
author: "Will Hannon"
date: "8/13/2020"
output: html_document
---


```{r setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)

## ==== Important file paths ==== ##

preprint.filepath =  "../../config/data/preprint_samples.csv" # Samples from pre-print
runselector.filepath =  "../../config/data/PRJNA610428.csv" # Samples from Bioproject
variant.filepath = "../../results/variant/variants.csv" # Combined Variant Calls
annotations.filepath = "../../config/data/sars_cov_2_annot.csv" # SARS-CoV-2 gene locations
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}
## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "kableExtra", "gridExtra", "grid")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```


## Notebook Overview
___

#### Project Aim

This analysis aims to see if there is a substantial transmission of intra-host minor variants between individuals in a large cluster of infections. This information could help characterize the inter-host transmission bottleneck in SARS-CoV-2 with the highest resolution achieved so far.

#### Cohort Overview

There was an outbreak on a fishing boat in Seattle. Individuals (120/122) were screened for SARS-CoV-2 RNA by RT-qPCR and seroconversion by the Abbot IgG assay both before departure and after return. Although all were negative, an outbreak occured nonetheless. A total of 104/122 crew members had viral RNA or seroconverted upon return to shore (84% Attack Rate). 

#### Data Overview 

From the 104 reportedly infected crew memebers there are 39 metagenomic RNA-sequencing runs avaliable. The data is avaliable publically at NCBI BioProject [`PRJNA610428`](https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=2&WebEnv=MCID_5f344f743a76fab9ab22e42b&o=acc_s%3Aa). Metadata for the identity of samples is avaliable in Supplementary Table 2.

The metadata only contains the GISAID-ID for the consensus and the strain. I might be able to use the strain to figure out the SRR# using the metadata on the SRA run selector.

#### Experimental Overview 

Libraries were sequenced on a 1x75 bp Illumina NextSeq run. A median of 509,551 sequencing reads were obtained for each sample.

```{r Samples, message=F, warning=F, echo=F}

## ==== Read-in sample data and get SRA accessions ==== ###
# Pre-print samples
samples = read_csv(preprint.filepath)
# Bioproject samples from Run Selector
bioproject = read_csv(runselector.filepath)
# Join the data-frames by 'Isolate' (derived from Strain)
combined_samples = left_join(samples, bioproject, by="Isolate") %>%
  select(Isolate, Run, LibraryLayout, Bases) %>% 
  mutate(Virus = "SARS2", Host = "none", Source = "public", Path = "[]", Experiment = "Cluster")
# From manual inspection, some samples are missing from the Bioproject
missing_samples = combined_samples %>% filter(is.na(Run))
# Samples that aren't missing
final_samples = combined_samples %>% filter(!is.na(Run))

## ==== Randomly get control samples ==== ###
# How many control samples to get? 
sample.size = nrow(final_samples)
# Libary size window
max = mean(final_samples$Bases) + sd(final_samples$Bases)
min = mean(final_samples$Bases) - sd(final_samples$Bases)
# Get only runs from the bioproject not already included with similar experimental design
control.pool = bioproject[which(!(bioproject$Run %in% final_samples$Run)),] %>% 
   filter(LibraryLayout == "SINGLE", Instrument == 	"NextSeq 500", Bases >= min & Bases <= max)
# Randomly sample the controls
control_samples = sample_n(control.pool, sample.size) %>% 
  select(Isolate, Run, LibraryLayout, Bases)  %>% 
  mutate(Virus = "SARS2", Host = "none", Source = "public", Path = "[]", Experiment = "Control")

# Join the controls and samples into final sample csv
analysis.samples = bind_rows(final_samples, control_samples)

## ==== Write-out the sample data to run the analysis ==== ###
## write.csv(analysis.samples, "../../config/samples.csv", row.names = FALSE)

```

#### Included Samples

The paper has a supplementary table (Supplementary Table 2.) that includes the strain names for each individual and the GISAID ID for the consensus, but not the SRA accession. I used this information to search the Bioproject ([`PRJNA610428`](https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=2&WebEnv=MCID_5f344f743a76fab9ab22e42b&o=acc_s%3Aa)) for the relevant samples. This Bioproject contains all of the samples sequenced by the UW Virology Department (~1000 Runs). Of the 39 samples referenced in the paper, there are `r nrow(final_samples)` that had corresponding Runs deposited in this project and `r nrow(missing_samples)` that did not. As a quasi-control, I randomly sampled an equal number of sequencing runs from the rest of the UW bioproject that were done on the same sequencer, were also single-ended, and were within one standard deviation of the mean library size for the transmission cluster Runs. 

## Analysis
___

#### Overview

I processed all of this data through the `viral-deepseq` pipeline, although I got rid of the de-duplication step. After running things initially with de-deuplication, and looking at the BAM files manually, I found that the majority of reads got removed. Also, for reasons I have yet to determine, a small percentage of the BAM files aligned with STAR failed (threw an error) de-duplication. 

```{r Formatting, message=FALSE, warning=FALSE, echo=FALSE}
## ==== Read in the variant calls ==== ##
variant.df = read_csv(variant.filepath)

## ====  Import genome annotations ==== ##
annotations.df = read_csv(annotations.filepath)

# Get the interval positions for these genes 
ORF1ab = annotations.df %>% dplyr::filter(gene == "Orf1ab")
S = annotations.df %>% dplyr::filter(gene == "Spike")
ORF3a = annotations.df %>% dplyr::filter(gene == "ORF3a")
E = annotations.df %>% dplyr::filter(gene == "Envelope")
M = annotations.df %>% dplyr::filter(gene == "Membrane")
ORF6 = annotations.df %>% dplyr::filter(gene == "ORF6")
ORF7a = annotations.df %>% dplyr::filter(gene == "ORF7a")
ORF8 = annotations.df %>% dplyr::filter(gene == "ORF8")
N = annotations.df %>% dplyr::filter(gene == "Nucleocapsid")
ORF10 = annotations.df %>% dplyr::filter(gene == "ORF10")

# Add gene names to the variant.df
variant.df = variant.df %>% 
  mutate(Gene = case_when(POS < ORF1ab$start ~ "3'UTR",
                          POS >= ORF1ab$start & POS <= ORF1ab$end  ~ "ORF1ab",
                          POS >= S$start & POS <= S$end  ~ "Spike",
                          POS >= ORF3a$start & POS <= ORF3a$end  ~ "ORF3a",
                          POS >= E$start & POS <= E$end  ~ "Envelope",
                          POS >= M$start & POS <= M$end  ~ "Matrix",
                          POS >= ORF6$start & POS <= ORF6$end  ~ "ORF6",
                          POS >= ORF7a$start & POS <= ORF7a$end  ~ "ORF7a",
                          POS >= ORF8$start & POS <= ORF8$end  ~ "ORF8",
                          POS >= N$start & POS <= N$end  ~ "Nucelocapsid",
                          POS >= ORF10$start & POS <= ORF10$end  ~ "ORF10",
                          POS > ORF10$end ~ "5'UTR")) %>% 
  select(!c("LibraryLayout", "Virus", "Host", "Source", "Path")) %>% 
  mutate(Var_Type = case_when(nchar(REF) == 1 & nchar(ALT) == 1 ~ "SNP",
                              nchar(REF) > 1 ~ "Deletion",
                              nchar(ALT) > 1 ~ "Insertion")) %>% 
  mutate(ALT = ifelse(ALT == "True", "T", ALT))

```

#### Quality Contol

Data for the quality control portion of the analysis was collated by the tool `multiqc`. It consists of stats from `samtools`, `STAR`, `fastp`, and `fastQC.` I've also included the coverage over each sample in this part of the analyis.


```{r }

```




```{r}
# Re-do label to refect new y-axis positions
label = annotations.df %>% 
  mutate(POS = (start+end)/2) %>% 
  mutate(AF = (-.01/-.1)/-2) %>% 
  select(gene, POS, AF) 

variant.df %>% 
  filter(Aligner == "BWA" & Caller == "varscan") %>% 
  filter(AF < 0.5) %>% 
  ggplot(aes(x = POS, y = AF, col = Accession, fill = Accession))+
    geom_point(size =2) + 
    geom_bar(stat="identity", width=.01, position = position_identity()) +
    facet_wrap(~Accession) +
    xlab("Position") +
    ylab("Allele Frequency") +
    ggtitle("Distribution of Variants") + 
    scale_x_continuous(breaks=c(0,5000,10000,15000,20000,25000,30000)) +
    annotate("rect", xmin = 0, xmax = ORF1ab[[2]]-1, ymin = -.1, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate("rect", xmin = ORF1ab[[2]], xmax = ORF1ab[[3]], ymin = -.1, ymax = -.01 ,
               alpha = .2, fill = "dodgerblue2", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][1], y = label[[3]][1], label = label[[1]][1], family = "mono") +
    annotate("rect", xmin = S[[2]], xmax = S[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "yellow3", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][2], y = label[[3]][2], label = label[[1]][2], family = "mono") +
    annotate("rect", xmin = ORF3a[[2]], xmax = ORF3a[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "#FF7F00", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][3], y = label[[3]][3], label = "3a", family = "mono") +
    annotate("rect", xmin = E[[2]], xmax = E[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill =  "#6A3D9A", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][4], y = label[[3]][4], label = "E", family = "mono") +
    annotate("rect", xmin = M[[2]], xmax = M[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill =  "green4", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][5], y = label[[3]][5], label = "M", family = "mono") +
    annotate("rect", xmin = ORF6[[2]], xmax = ORF6[[3]], ymin = -.1, ymax = -.01 ,
             alpha = .2, fill = "#E31A1C", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][6], y = label[[3]][6], label = "6", family = "mono") +
    annotate("rect", xmin = ORF7a[[2]], xmax = ORF7a[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "#FB9A99", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][7], y = label[[3]][7], label = "7a", family = "mono") +
    annotate("rect", xmin = ORF8[[2]], xmax = ORF8[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "orchid1", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][8], y = label[[3]][8], label = "8", family = "mono") +
    annotate("rect", xmin = N[[2]], xmax = N[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "maroon", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][9], y = label[[3]][9], label = "N", family = "mono") +
    annotate("rect", xmin = ORF10[[2]], xmax = ORF10[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill =  "skyblue2", col = "black", size = .1) +
    annotate("rect", xmin = ORF10[[3]]+1, xmax = 29903, ymin = -.1, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate(geom = "text", x = -200, y = label[[3]][1], label = "5'", family = "mono") +
    annotate(geom = "text", x = 30200, y = label[[3]][1], label = "3'", family = "mono") +
    #annotate(geom = "text", x = label[[2]][10], y = label[[3]][10], label = "10", family = "mono") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="mono")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed"),
        strip.text.x = element_blank()) 
  
```




