---
title: "Quality Control"
author: "Will Hannon"
date: "1/11/2021"
output: html_document
---

This notebook contains an QC analysis of sequencing runs for the `SARS-CoV-2_Bottleneck` project. The main goals of this notebook are to characterize: 

1. Identify samples and regions of low-depth in the sequencing. 
2. Determine the contribution of extra sequencing to the quality of samples. 
3. Identify any oddities in the sequencing runs.

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("tidyverse")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Paths to data, echo = FALSE}

# Path to a processed data frame from the tool Multiqc.
# This isn't exhaustive for all QC data collated by this tool. 
# For a quick look at individual samples, there is also an HTML
# repot generated in the `results/qc` directory.
multiqc.data = "../../results/qc/formatted_multiqc_data.csv"
# Path to the data for what % of each library are filted out by BBduk
filter.data = "../../results/qc/BBduk_filtered_reads.csv"
# Path to sample information
sample.data = "../../config/samples.csv"
# Path to boat metadata
boat.data = "../../config/boat_metadata.csv"

```

```{r Load in data, echo = FALSE, message = FALSE}

# QC data from multiqc
multiqc.df = read_csv(multiqc.data)
# QC data from BBduk, collated separatley
filter.df = read_csv(filter.data) %>% mutate(library = as.character(library))
# Join the BBduk data with the multiqc data
qc.df = left_join(multiqc.df, filter.df, by = c("sample", "replicate", "library"))
# Data from before the individual libraries (same) were merged into one.
qc.before.merge = qc.df %>% filter(library != "merged")
# Data from after the individual libraries (same) were merged into one.
qc.after.merge = qc.df %>% filter(library == "merged")

# Sample and boat information
sample.df = read_csv(sample.data) %>% select(LibraryLayout, Virus, Host, Source, spID) %>% distinct()
boat.df = read_csv(boat.data) %>% rename(spID = "SpID.x")
metadata.df = full_join(boat.df, sample.df, by = "spID") %>% 
  mutate(Included = if_else(is.na(Source), "Not Included", "Included"))

# Colors 
experiment.colors = c("#000000", "#7a7a7a")
```

Out of all 39 genomes that were reconstucted for the phylogeny, how many did we include in the analysis? 

```{r Samples included by Ct, echo = F, message = F, warning = F, fig.width=7, fig.height=3, fig.align='center'}

metadata.df %>% 
  ggplot(aes(x = reorder(spID, -CT), y = CT, col = Included)) + 
    geom_hline(yintercept = 20, col = "black", linetype = 2, size = 1.5) + 
    geom_point(size = 5) +
    xlab("Sample") + 
    ylab("Cycle Threshold (Ct)") + 
    scale_color_manual(values = c("#DC3220","#005AB5")) + 
    theme_classic(35) +
    theme(axis.text.x=element_blank(), 
          axis.ticks.x=element_blank(),
          legend.position = "none") 

```

The following plot is the size of each library (how many mapped SARS-CoV-2 reads are there per patient per replicate) both before and after merging the separate sequencing attempts. The facet labled `1` shows each library from the initial round of sequencing. The facet labled `2` shows the extra sequencing we requested from Pavitra to even out the coverage between replicates. It appears the have made most samples more evenly covered, with the notable exception of patient `10027`.

```{r Run Size, echo = FALSE, message = FALSE, fig.width=15, fig.height=5, fig.align='center'}

# The number of mapped reads for each sample. (1) correspond to the initial sequencing. (2) corresponds to the supplementary sequencing.
# `merged` corresponds to the combindation of the two libraries. 
qc.df %>% 
  ggplot(aes(x = as.factor(sample), y = reads_mapped, fill = as.factor(replicate))) +
    geom_bar(stat = "identity", position = position_dodge()) + 
    facet_wrap(~library) +
    scale_fill_manual(values = experiment.colors) +
    xlab("Sample") +
    ylab("Number of Mapped Reads") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(legend.position="none") +
    theme(axis.text.x=element_text(angle=45, hjust=1))

```

Below is the same plot as above, but only showing the library size between replicates after combining the extra sequencing. Library size is more comparable between replicates now. However, it's certainly not perfect. 

```{r Combined Run Size, echo = FALSE, message = FALSE, fig.width=7, fig.height=3, fig.align='center'}

# Just the `merged` samples for a better look.
qc.after.merge %>% 
  ggplot(aes(x = as.factor(sample), y = reads_mapped, fill = as.factor(replicate))) +
    geom_bar(stat = "identity", position = position_dodge()) + 
    scale_fill_manual(values = experiment.colors) +
    xlab("Sample") +
    ylab("Number of Mapped Reads") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(legend.position="none") +
    theme(axis.text.x=element_text(angle=45, hjust=1))

```

The percent of the sample that was filtered is pretty all over the place, with some samples nearly 100% SARS-CoV-2 and others closer to 5% SARS-CoV-2. Interestingly, there are some pretty bid discrepancies between replicates. I wonder if this indicates low template number in some way? Perhaps high template number samples are more resistant to differences in the percentage of SARS-CoV-2 reads? This is the average of different sequening attemps for each sample/replicate pair.

```{r Percent Filtered, echo = FALSE, message = FALSE, fig.width=7, fig.height=3, fig.align='center'}

# Just the `merged` samples for a better look.
qc.before.merge %>% 
  ggplot(aes(x = as.factor(sample), y = percent_filtered_viral_reads, fill = as.factor(replicate))) +
    geom_bar(stat = "identity", position = position_dodge()) + 
    scale_fill_manual(values = experiment.colors) +
    xlab("Sample") +
    ylab("Number of Mapped Reads") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(legend.position="none") +
    theme(axis.text.x=element_text(angle=45, hjust=1))

```

Below is a plot of the average base quality of each sequencing run. The average base quality of all samples falls above the target score of 25, which implies an error rate of ~0.003.

```{r Average Base Quality, echo = FALSE, message = FALSE, fig.width=7, fig.height=3, fig.align='center'}

# Just the `merged` samples for a better look.
qc.after.merge %>% 
  ggplot(aes(x = as.factor(sample), y = average_quality, fill = as.factor(replicate))) +
    geom_bar(stat = "identity", position = position_dodge()) +
    geom_hline(yintercept = 25, col = "red", linetype = 3, size = 2) +
    scale_fill_manual(values = experiment.colors) +
    xlab("Sample") +
    ylab("Average Quality Score (Phred)") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(legend.position="none") +
    theme(axis.text.x=element_text(angle=45, hjust=1))

```








