---
title: "3. Replicate Analysis"
author: "Will Hannon"
date: "1/11/2021"
output: html_document
---

The goal of this notebook is compare the variants called between sequencing replicates. Replicates were performed from the same samples, but with different RT reactions. This has been shown to effectively screen out low-quality samples where the concordance between replicates is poor. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "ggrepel", "plotly", "ggpubr", "gridExtra", 'grid', "UpSetR")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Filepaths, include=FALSE}

## ==== File paths input data ==== ##

# Data from pysam/python script. It contains the % of each nucleotide
# at every covered position in the genome. This is a huge file.
pysam.data = "../../results/pileup/pysam-variants.csv"

# Data from lofreq
variant.data = "../../results/variants/variants.csv"

# Samtools average depth
average.depth.data = "../../results/coverage/merged.average.depth" 

```

Building on the previous two analyses, I removed any samples from the analysis when they fail the 'completeness'/coverage requirement of more than 80% of the genome being covered by 100X or more reads. 

```{r Filter Low Depth Sample, warning=F, message=F}

# Low quality samples - get the accessions for the samples that failed basic quality control. 
# Presently, this anything that has 20% of the genome covered by less than *100* reads. 

average.depth.df = read.table(average.depth.data, header = T) %>% 
  separate(Accession, into = c("Accession", "Replicate")) %>% 
  mutate(Run = paste0(Accession, Replicate))

low.coverage.samples = average.depth.df %>% filter(Percent < 80) %>% pull(Accession)
low.coverage.runs = average.depth.df %>% filter(Percent < 80) %>% pull(Run)

```

```{r Format Pysam data, warning=FALSE, message=FALSE}

# Import the data splitting the accession ID into the Accession and the Replicate
pysam.df = read_csv(pysam.data) %>% 
  separate(ACCESSION, into = c("ACCESSION", "REPLICATE"), sep = "_") %>% 
  mutate(REPLICATE = as.factor(REPLICATE))

# Join the custom pysam SNPs by replicate
pysam.to.join.df = pysam.df %>% 
  select(!c("SNP", "CONS"))

pysam.combined.df = full_join(
  filter(pysam.to.join.df, REPLICATE == "1"),
  filter(pysam.to.join.df, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT", "EFFECT", "GENE", "CODON_POS", "AA_CHANGE"),
  suffix = c(".one", ".two")
  ) %>% 
  select(!c("REPLICATE.one", "REPLICATE.two")) %>% 
  mutate(AF.one = if_else(is.na(AF.one), 0, AF.one)) %>% 
  mutate(AF.two = if_else(is.na(AF.two), 0, AF.two)) %>% 
  mutate(DP.one = if_else(is.na(DP.one), 0, DP.one)) %>% 
  mutate(DP.two = if_else(is.na(DP.two), 0, DP.two)) 



```

```{r Format Variant Data, warning=F, message=F}

variant.df = read_csv(variant.data) %>% 
  select(!c("Library", "Path", "Aligner", "LibraryLayout", "Virus", "Host", "Source", "Percent", "spID", "Replicate")) %>% 
  distinct() %>%  # Duplicated columns when you merge with the sample data. 
  separate(Accession, into = c("ACCESSION", "REPLICATE"), sep = "_") %>% 
  rename("AA_CHANGE" = AA_Change, "EFFECT" = Effect, "GENE" = Gene_Name)

variant.df.to.join = variant.df %>% 
  select(!c("EFFECT", "GENE", "AA_CHANGE"))

variant.combined.df = full_join(
  filter(variant.df.to.join, REPLICATE == "1"),
  filter(variant.df.to.join, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT", "Caller"),
  suffix = c(".one", ".two")
  ) %>% 
  select(!c("REPLICATE.one", "REPLICATE.two")) %>% 
  mutate(AF.one = if_else(is.na(AF.one), 0, AF.one)) %>% 
  mutate(AF.two = if_else(is.na(AF.two), 0, AF.two)) %>% 
  mutate(DP.one = if_else(is.na(DP.one), 0, DP.one)) %>% 
  mutate(DP.two = if_else(is.na(DP.two), 0, DP.two)) 



```

## Pipeline Comparison

Here I compare the variant callers to one another. I only look at SNPs because I've only benchmarked the pipeline of SNPs. Additionaly, I excluded all sites with less than 100X depth, 2% AF, and from samples that didn't meet the studies coverage requirements. 

It's a good sign tha the largest set is the overalp of all callers. Otherwise, there are no variants called by Pysam that are absent from other callers, which is also a good sign. It seems like iVar and my method hav a significant number that overlap but are absent from other callers. These are probably the two least stringent approaches, so this is to be expected. 

Main takeaway - *all of the variants that I called with Pysam have been called by at least one other benchmarked variant calling approach.* 

```{r Compare Callers, message=F, warning=F}

upset.plt = pysam.df %>% 
  mutate(Caller = "pysam") %>% 
  select(colnames(variant.df)) %>% 
  rbind(., variant.df) %>%  
  mutate(Type = case_when(nchar(REF) > 1 ~ "Del",
                          nchar(ALT) > 1 ~ "Ins",
                          nchar(REF) == 1 ~ "SNP")) %>% 
  filter(Type == "SNP") %>% 
  filter(DP >= 100) %>% # More than 100X
  filter(AF >= 0.02) %>% # More than 2% AF
  filter(!ACCESSION %in% low.coverage.samples) %>% # Not a "low coverage sample"
  mutate(SNP = paste0(REF, POS, ALT)) %>% 
  mutate(Identifier = paste(ACCESSION, REPLICATE, SNP, sep = "-")) %>% 
  select(Identifier, Caller) %>% 
  distinct() %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "Caller", values_from = "Tally", values_fill = 0) %>% 
  as.data.frame(row.names = Identifier) %>%
  upset(., sets = c("lofreq",
                          "varscan",
                          "ivar",
                          "pysam"), sets.bar.color = "#56B4E9",
      order.by = "freq", empty.intersections = "on", text.scale = 1.5)

upset.plt

png(filename = "../../results/figures/variant_caller_upset_plot.png", width = 7, height = 5)
upset.plt
dev.off()

```

## Concordance Between Replicates

The point of having replicates from the same samples with different reverse transcription reactions and library preps is to assess the quality of these samples. If variants are called at substantiall different frequences between replicates, or if variants are entirely abscent from replicates, it's possible that (a) those variants are spurious, and (b) those samples have low viral template numbers. 

To assess this, I looked at the concordance between the allele frequencies from the merged samples. Below is a plot of the concordance of allele frequency between the sequencing replicates of all libraries with *Pysam/Python* variant calling. 

```{r Concordance, message = F, warning = F, fig.width= 25, fig.height= 20, fig.align='center'}

concordance_plot_list = list()

samples = sort(unique(pysam.combined.df$ACCESSION))[which(!sort(unique(pysam.combined.df$ACCESSION)) %in% low.coverage.samples)]

included.variants.df = pysam.combined.df %>% 
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  filter(AF.one >= 0.02 & AF.two >= 0.02)
  
excluded.variants.df = pysam.combined.df %>% 
  filter(DP.one < 100 | DP.two < 100) %>% 
  filter(AF.one < 0.02 | AF.two < 0.02)

for (i in 1:length(samples)) {

  p1 = included.variants.df  %>% 
    filter(ACCESSION == samples[i]) %>% 
    ggplot(aes(x = AF.one, y = AF.two)) + 
      geom_point(data = filter(excluded.variants.df, ACCESSION == samples[i]), aes(x = AF.one, y = AF.two), size = 1, col = "#b80000", alpha = 0.5) +
      geom_point(size = 4, alpha = .75) +
      geom_abline(intercept=0, col = "blue") + 
      geom_hline(yintercept = 0.02, col = "blue", linetype = 2) + 
      geom_vline(xintercept = 0.02, col = "blue", linetype = 2) + 
      facet_wrap(~ACCESSION) +
      scale_color_manual(values = c("#080000","#b80000")) +
      scale_x_continuous(limits = c(0, 1), labels = scales::percent, guide = guide_axis(angle = 25)) + 
      scale_y_continuous(limits = c(0, 1), labels = scales::percent) + 
      xlab("") +
      ylab("") +
      theme_classic() +
      theme(text=element_text(size= 25,  family="Helvetica")) +
      theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
      theme(panel.background = element_rect(fill = NA, color = "black")) +
      theme(plot.title = element_text(hjust = 0.5)) + 
      coord_fixed()
  
  p2 = included.variants.df  %>% 
    filter(ACCESSION == samples[i]) %>% 
    ggplot(aes(x = AF.one, y = AF.two)) + 
      geom_point(data = filter(excluded.variants.df, ACCESSION == samples[i]), aes(x = AF.one, y = AF.two), size = 1.5, col = "#b80000", alpha = 0.5) +
      geom_point(size = 2.5, alpha = .75) +
      geom_abline(intercept=0, col = "blue") + 
      geom_hline(yintercept = 0.02, col = "blue", linetype = 2) + 
      geom_vline(xintercept = 0.02, col = "blue", linetype = 2) + 
      scale_color_manual(values = c("#080000","#b80000")) +
      scale_x_continuous(limits = c(0, .10), labels = scales::percent, guide = guide_axis(angle = 25)) + 
      scale_y_continuous(limits = c(0, .10), labels = scales::percent) + 
      xlab("") +
      ylab("") +
      theme_classic() +
      theme(text=element_text(size = 14,  family="Helvetica")) +
      theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
      theme(panel.background = element_rect(fill = NA, color = "black")) +
      theme(plot.title = element_text(hjust = 0.5)) + 
      theme(axis.text.x = element_blank()) + 
      coord_fixed()
  
  
  p3 = p1 + 
    annotation_custom(ggplotGrob(p2), xmin = .25, xmax = 1.15, ymin = -.05, ymax = .4) 

  concordance_plot_list[[i]] = p3
  
}

final_plot = gridExtra::grid.arrange(arrangeGrob(grobs = concordance_plot_list, 
                          left = textGrob("Allele Frequency Replicate Two", rot = 90, vjust = 1,
                                          gp = gpar(col = "black", fontsize = 35)),
                          bottom = textGrob("Allele Frequency Replicate One", hjust = .5,
                                            gp = gpar(col = "black", fontsize = 35))))

ggsave("../../results/figures/replicate_comparison_pysam.png", plot = final_plot, width = 28, height = 22, dpi = 300, units = "in")

```

In order to determine rationally which samples should be included and which should be excluded, I screen samples based on the average deviation between replicate frequencies. This method weights deviation amoung high frequency variants more than low frequency variants. 

This histogram shows which samples won't be included in the analysis because the allele frequences between replicates are to dissimilar. In this case, I cut off the long tail. All samples that had **more than 0.025 mean difference between SNPs** were not included in the downstream analysis. 

```{r Concordance Difference Histogram, message = F, warning = F, fig.width= 7, fig.height= 5, fig.align='center'}

# Cut off the long tail of the distribution
pysam.combined.df %>% 
  filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  select(ACCESSION, REF, POS, ALT, DP.one, DP.two, AF.one, AF.two) %>% 
  mutate(DIFF = abs(AF.one - AF.two)) %>% 
  group_by(ACCESSION) %>% 
  summarise(MEAN_DIFF = mean(DIFF)) %>% 
  ggplot(aes(x = MEAN_DIFF)) + 
    geom_histogram(bins = 30) + 
    geom_vline(xintercept = 0.025, col = "red") +
    xlab("Mean Difference Between Replicates") +
    ylab("Samples") +
    theme_classic() +
    theme(text=element_text(size = 18,  family="Helvetica")) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(panel.background = element_rect(fill = NA, color = "black")) +
    theme(plot.title = element_text(hjust = 0.5))
  
poor.concordace.samples = pysam.combined.df %>% 
  filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  select(ACCESSION, REF, POS, ALT, DP.one, DP.two, AF.one, AF.two) %>% 
  mutate(DIFF = abs(AF.one - AF.two)) %>% 
  group_by(ACCESSION) %>% 
  summarise(MEAN_DIFF = mean(DIFF)) %>% 
  filter(MEAN_DIFF >= 0.025) %>% 
  pull(ACCESSION)

ggsave("../../results/figures/average_allele_difference.svg", width = 7, height = 5, dpi = 300, units = "in")

poor.concordace.samples[which(!poor.concordace.samples %in% low.coverage.samples)]
```

## Concordance Between Callers

Finally, I wanted to see if the variants called by one caller have better concordance on average than another caller. 

```{r Concordance Difference All Callers, message = F, warning = F, fig.width= 10, fig.height= 10, fig.align='center'}


all.caller.df = pysam.combined.df %>% 
  mutate(Caller = "pysam") %>% 
  select(colnames(variant.combined.df)) %>% 
  rbind(variant.combined.df) %>% 
  mutate(Type = case_when(nchar(REF) > 1 ~ "Del",
                          nchar(ALT) > 1 ~ "Ins",
                          nchar(REF) == 1 ~ "SNP")) %>% 
  filter(Type == "SNP")  %>% 
  filter(!ACCESSION %in% low.coverage.samples)

included.variants.df = all.caller.df %>% 
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  filter(AF.one >= 0.02 & AF.two >= 0.02)
  
excluded.variants.df = all.caller.df %>% 
  filter(DP.one < 100 | DP.two < 100) %>% 
  filter(AF.one < 0.02 | AF.two < 0.02)

included.variants.df  %>% 
  ggplot(aes(x = AF.one, y = AF.two)) + 
    geom_point(data = filter(excluded.variants.df), aes(x = AF.one, y = AF.two), size = 1, col = "#b80000", alpha = 0.5) +
    geom_point(size = 4, alpha = .75) +
    geom_abline(intercept=0, col = "blue") + 
    geom_hline(yintercept = 0.02, col = "blue", linetype = 2) + 
    geom_vline(xintercept = 0.02, col = "blue", linetype = 2) + 
    facet_wrap(~Caller) +
    scale_color_manual(values = c("#080000","#b80000")) +
    scale_x_continuous(limits = c(0, 1)) + 
    scale_y_continuous(limits = c(0, 1)) + 
    xlab("") +
    ylab("") +
    theme_classic() +
    theme(text=element_text(size= 30,  family="Helvetica")) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(panel.background = element_rect(fill = NA, color = "black")) +
    theme(plot.title = element_text(hjust = 0.5)) + 
    coord_fixed()

ggsave("../../results/figures/variant_concordance_call_comparison.png", width = 10, height = 10, dpi = 300, units = "in")

```

`Varscan` and `Pysam` look pretty similar. iVar has terrible concordance, no idea what's going on there. It could have something to do with how the caller handles insertions and deletions in reads? `Lofreq` has a unique cluster of variants that are above 50% and below 100%. These are mostly at position 11803 and 13, which are fixed in the other tools. I don't know why it's finding these reversions, but this has been seen in other papers (Braun et. al., 2021 (a/b)).