---
title: "Replicate Analysis"
author: "Will Hannon"
date: "1/11/2021"
output: html_document
---

This notebook contains an analysis of replicate sequencing runs for the `SARS-CoV-2_Bottleneck` project. The main goals of this notebook are to characterize: 

1. Concordance between replicate sequencing runs. 
2. Identify samples with poor concordance that should be excluded from the analysis. 
3. Determine the ideal heuristic thresholds for sequencing depth and allele count. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "ggrepel", "plotly", "ggpubr", "gridExtra", 'grid')
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Filepaths, include=FALSE}

## ==== File paths input data ==== ##

# Data from pysam/python script. It contains the % of each nucleotide
# at every covered position in the genome. This is a huge file.
pysam.data = "../../results/pileup/pysam-variants.csv"
# Data from lofreq
variant.data = "../../results/variants/variants.csv"
# Samtools average depth
average.depth.data = "../../results/coverage/merged.average.depth" 

```

```{r Format data, include=FALSE, warning=FALSE, message=FALSE}

# Low quality samples 
average.depth.df = read.table(average.depth.data, header = T) %>% 
  separate(Accession, into = c("Accession", "Replicate")) %>% 
  mutate(Run = paste0(Accession, Replicate))

low.coverage.samples = average.depth.df %>% filter(Percent < 80) %>% pull(Accession)
low.coverage.runs = average.depth.df %>% filter(Percent < 80) %>% pull(Run)


# Import the data splitting the accession ID into the Accession and the Replicate
pysam.df = read_csv(pysam.data) %>% 
  separate(ACCESSION, into = c("ACCESSION", "REPLICATE"), sep = "_") %>% 
  mutate(REPLICATE = as.factor(REPLICATE))

# Join the custom pysam SNPs by replicate
pysam.to.join.df = pysam.df %>% 
  select(!c("SNP", "CONS"))

pysam.combined.df = full_join(
  filter(pysam.to.join.df, REPLICATE == "1"),
  filter(pysam.to.join.df, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT", "EFFECT", "GENE", "CODON_POS"),
  suffix = c(".one", ".two")
  ) %>% 
  select(!c("REPLICATE.one", "REPLICATE.two")) %>% 
  mutate_each(list( ~ replace(., which(is.na(.)), 0))) %>% 
  mutate(Coverage = ifelse(ACCESSION %in% low.coverage.samples, "Low", "High"))


```

## Concordance

Below is a plot of the concordance of allele frequency between the sequencing replicates of all libraries. This doesn't include any filters for depth of sequencing. I colored any runs which had a replicate less than 80% covered by more than 200 reads red. This doesn't account for all of the discrepencies in allele frequency.   

```{r}

concordance_plot_list = list()
samples = sort(unique(pysam.combined.df$ACCESSION))

included.variants.df = pysam.combined.df %>% 
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  filter(AF.one >= 0.02 & AF.two >= 0.02)
  
excluded.variants.df = pysam.combined.df %>% 
  filter(DP.one < 100 | DP.two < 100) %>% 
  filter(AF.one < 0.02 | AF.two < 0.02)

for (i in 1:length(samples)) {

  p1 = included.variants.df  %>% 
    filter(ACCESSION == samples[i]) %>% 
    ggplot(aes(x = AF.one, y = AF.two)) + 
      geom_point(data = filter(excluded.variants.df, ACCESSION == samples[i]), aes(x = AF.one, y = AF.two), size = 1, col = "#b80000", alpha = 0.5) +
      geom_point(size = 3, alpha = .75) +
      geom_abline(intercept=0, col = "blue") + 
      geom_hline(yintercept = 0.02, col = "blue", linetype = 2) + 
      geom_vline(xintercept = 0.02, col = "blue", linetype = 2) + 
      facet_wrap(~ACCESSION) +
      scale_color_manual(values = c("#080000","#b80000")) +
      scale_x_continuous(limits = c(0, 1)) + 
      scale_y_continuous(limits = c(0, 1)) + 
      xlab("") +
      ylab("") +
      theme_classic() +
      theme(text=element_text(size= 22,  family="Helvetica")) +
      theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
      theme(panel.background = element_rect(fill = NA, color = "black")) +
      theme(plot.title = element_text(hjust = 0.5)) + 
      coord_fixed(ratio = .75)
  
  p2 = included.variants.df  %>% 
    filter(ACCESSION == samples[i]) %>% 
    ggplot(aes(x = AF.one, y = AF.two)) + 
      geom_point(data = filter(excluded.variants.df, ACCESSION == samples[i]), aes(x = AF.one, y = AF.two), size = 1.5, col = "#b80000", alpha = 0.5) +
      geom_point(size = 2, alpha = .75) +
      geom_abline(intercept=0, col = "blue") + 
      geom_hline(yintercept = 0.02, col = "blue", linetype = 2) + 
      geom_vline(xintercept = 0.02, col = "blue", linetype = 2) + 
      scale_color_manual(values = c("#080000","#b80000")) +
      scale_x_continuous(limits = c(0, .10)) + 
      scale_y_continuous(limits = c(0, .10)) + 
      xlab("") +
      ylab("") +
      theme_classic() +
      theme(text=element_text(size = 12,  family="Helvetica")) +
      theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
      theme(panel.background = element_rect(fill = NA, color = "black")) +
      theme(plot.title = element_text(hjust = 0.5)) + 
      theme(axis.text.x = element_blank()) + 
      coord_fixed()
  
  
  p3 = p1 + 
    annotation_custom(ggplotGrob(p2), xmin = .25, xmax = 1.15, ymin = -.05, ymax = .4) 

  concordance_plot_list[[i]] = p3
  
}

final_plot = gridExtra::grid.arrange(arrangeGrob(grobs = concordance_plot_list, 
                          left = textGrob("Allele Frequency Replicate Two", rot = 90, vjust = 1,
                                          gp = gpar(col = "black", fontsize = 20)),
                          bottom = textGrob("Allele Frequency Replicate One", hjust = .5,
                                            gp = gpar(col = "black", fontsize = 20))))

ggsave("../../results/pysam_variant_concordance.png", plot = final_plot, width = 25, height = 20, dpi = 300, units = "in")

```


