---
title: "5. Shared Variant Analysis"
author: "Will Hannon"
date: "4/13/2021"
output: html_document
---

The goal of this notebook is to look at the iwthin host variants that are shared between crewmembers. Do any of these variants suggest that they were transmitted at low frequency between individuals? Or, are most shared variants the result of some other process - i.e. homoplasy, sequencing error, batch effect, etc..?

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "scales", "RColorBrewer", "ggrepel", "Biostrings", "gridExtra", 'grid')
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Filepaths, include=FALSE}

## ==== File paths input data ==== ##

# Data from pysam/python script. It contains the % of each nucleotide
# at every covered position in the genome. This is a huge file.
pysam.data = "../../results/pileup/pysam-variants.csv"

# Data from lofreq, varscan, and ivar
variant.data = "../../results/variants/variants.csv"

# Samtools average depth
average.depth.data = "../../results/coverage/merged.average.depth" 

# Sample data
sample_metadata = "../../config/data/Boat_Sample_Metadata.csv"

```

I'll format the variants that I called using the pileup function of the Python library `Pysam`. I'll also join the replicates.

```{r Format Pysam data, warning=FALSE, message=FALSE}

# Import the data splitting the accession ID into the Accession and the Replicate
pysam.df = read_csv(pysam.data) %>% 
  separate(ACCESSION, into = c("ACCESSION", "REPLICATE"), sep = "_") %>% 
  mutate(REPLICATE = as.factor(REPLICATE))

# Join the custom pysam SNPs by replicate
pysam.to.join.df = pysam.df %>% 
  select(!c("SNP", "CONS"))

pysam.combined.df = full_join(
  filter(pysam.to.join.df, REPLICATE == "1"),
  filter(pysam.to.join.df, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT", "EFFECT", "GENE", "CODON_POS", "AA_CHANGE"),
  suffix = c(".one", ".two")
  ) %>% 
  select(!c("REPLICATE.one", "REPLICATE.two")) %>% 
  mutate(AF.one = if_else(is.na(AF.one), 0, AF.one)) %>% 
  mutate(AF.two = if_else(is.na(AF.two), 0, AF.two)) %>% 
  mutate(DP.one = if_else(is.na(DP.one), 0, DP.one)) %>% 
  mutate(DP.two = if_else(is.na(DP.two), 0, DP.two)) 


```

Then, I'll do the sample with the variants called with three variant calling softwares: `lofreq`, `ivar`, and `varscan`. Here, I also standardize the column names with those of the pysam variants. 

```{r Format Variant Data, warning=F, message=F}

variant.df = read_csv(variant.data) %>% 
  select(!c("Library", "Path", "Aligner", "LibraryLayout", "Virus", "Host", "Source", "Percent", "spID", "Replicate")) %>% 
  distinct() %>%  # Duplicated columns when you merge with the sample data. 
  separate(Accession, into = c("ACCESSION", "REPLICATE"), sep = "_") %>% 
  dplyr::rename("AA_CHANGE" = AA_Change, "EFFECT" = Effect, "GENE" = Gene_Name)

variant.df.to.join = variant.df %>% 
  select(!c("EFFECT", "GENE", "AA_CHANGE"))

variant.combined.df = full_join(
  filter(variant.df.to.join, REPLICATE == "1"),
  filter(variant.df.to.join, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT", "Caller"),
  suffix = c(".one", ".two")
  ) %>% 
  select(!c("REPLICATE.one", "REPLICATE.two")) %>% 
  mutate(AF.one = if_else(is.na(AF.one), 0, AF.one)) %>% 
  mutate(AF.two = if_else(is.na(AF.two), 0, AF.two)) %>% 
  mutate(DP.one = if_else(is.na(DP.one), 0, DP.one)) %>% 
  mutate(DP.two = if_else(is.na(DP.two), 0, DP.two)) 

```

I'll get a list of the low coverage samples to exclude from downstream analyses. These are any samples where at least on replicate has more than 20% of the genome covered by less than 200 reads:

```{r Filter Low Depth Sample, warning=F, message=F}

# Low quality samples - get the accessions for the samples that failed basic quality control. 
# Presently, this anything that has 20% of the genome covered by less than *200* reads. 
average.depth.df = read.table(average.depth.data, header = T) %>% 
  separate(Accession, into = c("Accession", "Replicate")) %>% 
  mutate(Run = paste0(Accession, Replicate))

low.coverage.samples = average.depth.df %>% filter(Percent < 80) %>% pull(Accession)
low.coverage.runs = average.depth.df %>% filter(Percent < 80) %>% pull(Run)

print(low.coverage.samples)
```

I'll also get a list of all of the samples that have poor concordance between replicates. 

```{r Filter Poor Concordance Samples, warning=F, message=F}

poor.concordance.samples = pysam.combined.df %>% 
  filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  select(ACCESSION, REF, POS, ALT, DP.one, DP.two, AF.one, AF.two) %>% 
  mutate(DIFF = abs(AF.one - AF.two)) %>% 
  group_by(ACCESSION) %>% 
  summarise(MEAN_DIFF = mean(DIFF)) %>% 
  filter(MEAN_DIFF >= 0.025) %>% 
  pull(ACCESSION)

poor.concordance.samples = poor.concordance.samples[which(!poor.concordance.samples %in% low.coverage.samples)]

print(poor.concordance.samples)

```

Next, I wanted to select the main variants to include in the downstream analyses. To be included, the SNPs need to meet the following criteria: 

1. They must be in samples that are 'high coverage' 80% of the genome covered by 100 reads or more. 
2. They have to be samples with high concordance between replicate allele frequencies (see `sequencing-replicate-analysis.Rmd`). 
3. Variants must be present at more than 2% frequency in both of the replicates. 
4. Variants must be covered by more than 100 reads in both of the replicates. 

Finally, the variants are combined between the replicates by averaging the allele frequency. Then, alleles present at more than 50% frequency and less than 98% frequency are flipped to their non-consensus counterparts. 

```{r Prepare Final Pysam Variants, warning=F, message=F}

# Only combine those SNPs that are in both samples
selected.variant.df = inner_join(
  filter(pysam.to.join.df, REPLICATE == "1"),
  filter(pysam.to.join.df, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT", "EFFECT", "GENE", "CODON_POS", "AA_CHANGE"),
  suffix = c(".one", ".two")
  ) %>% 
  # Remove the replicate columns 
  select(!c("REPLICATE.one", "REPLICATE.two")) %>%  
  # Add in a column of SNP identifiers 
  mutate(SNP = paste0(REF, POS, ALT)) %>% 
  # Remove the low coverage samples (> 80% covered by 200+ reads)
  filter(!(ACCESSION %in% low.coverage.samples)) %>% 
  # Remove the samples with poor concordance
  filter(!(ACCESSION %in% poor.concordance.samples)) %>% 
  # Filter out the variants that don't meet the depth/af requirements 
  filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
  # Drastically speeds things up if you have a minimum depth
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  # Combine AF and DP into a single value without averaging
  mutate(AF = case_when(DP.one > DP.two ~ AF.one,
                        DP.two > DP.one ~ AF.two,
                        DP.one == DP.two ~ AF.one)) %>% 
  mutate(DP = case_when(DP.one > DP.two ~ DP.one,
                        DP.two > DP.one ~ DP.two)) %>% 
  select(!ends_with(".one")) %>% 
  select(!ends_with(".two")) 

# Filp the consensus variants
fipped.consensus.df = selected.variant.df %>% 
  filter(AF >= 0.5) %>% 
  dplyr::rename(ALT = "REF", REF = "ALT") %>% 
  mutate(AF = (1 - AF)) %>% 
  filter(AF >= 0.02) %>% 
  separate(AA_CHANGE, into = c("WT", "AAPOS"),
           sep = "((?<=[A-Za-z])(?=[0-9]))") %>% 
  separate(AAPOS, into = c("AAPOS", "MUT"),
           sep = "((?<=[0-9])(?=[A-Za-z]))") %>% 
  mutate(AA_CHANGE = paste0(MUT, AAPOS, WT)) %>% 
  select(!c("MUT", "WT", "AAPOS"))

# final.variant.df = rbind(filter(selected.variant.df, AF < 0.5), fipped.consensus.df) %>% 
#   mutate(SNV = paste0(REF, POS, ALT))

final.variant.df = selected.variant.df %>% 
  mutate(SNV = paste0(REF, POS, ALT))


```


```{r Prepare Final Caller Variants, warning=F, message=F}

# Final samples and mutations to analyze: 
# Remove low coverage samples
# Remove alleles below 2%
# Remove alleles with less than 100X coverage

caller.selected.variant.df = inner_join(
    filter(variant.df, REPLICATE == "1"),
    filter(variant.df, REPLICATE == "2"), 
    by = c("ACCESSION", "POS", "REF", "ALT", "EFFECT", "GENE", "AA_CHANGE", "Caller"),
    suffix = c(".one", ".two")
    ) %>% 
    # Remove the replicate columns 
    select(!c("REPLICATE.one", "REPLICATE.two")) %>%  
    # Add in a column of SNP identifiers 
    mutate(SNP = paste0(REF, POS, ALT)) %>% 
    # Remove the low coverage samples (> 80% covered by 200+ reads)
    filter(!(ACCESSION %in% low.coverage.samples)) %>% 
    # Remove the samples with poor concordance
    filter(!(ACCESSION %in% poor.concordance.samples)) %>% 
    # Filter out the variants that don't meet the depth/af requirements 
    filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
    # Drastically speeds things up if you have a minimum depth
    filter(DP.one >= 100 & DP.two >= 100) %>% 
    # Combine AF and DP into a single value without averaging
    mutate(AF = case_when(DP.one > DP.two ~ AF.one,
                          DP.two > DP.one ~ AF.two,
                          DP.one == DP.two ~ AF.one)) %>% 
    mutate(DP = case_when(DP.one > DP.two ~ DP.one,
                          DP.two > DP.one ~ DP.two)) %>% 
    select(!ends_with(".one")) %>% 
    select(!ends_with(".two")) 

# Filp the consensus variants
caller.fipped.consensus.df = caller.selected.variant.df %>% 
  filter(AF >= 0.5) %>% 
  dplyr::rename(ALT = "REF", REF = "ALT") %>% 
  mutate(AF = (1 - AF)) %>% 
  filter(AF >= 0.02) %>% 
  separate(AA_CHANGE, into = c("WT", "AAPOS"),
           sep = "((?<=[A-Za-z])(?=[0-9]))") %>% 
  separate(AAPOS, into = c("AAPOS", "MUT"),
           sep = "((?<=[0-9])(?=[A-Za-z]))") %>% 
  mutate(AA_CHANGE = paste0(MUT, AAPOS, WT)) %>% 
  select(!c("MUT", "WT", "AAPOS"))

caller.final.variant.df = rbind(filter(caller.selected.variant.df, AF < 0.5), caller.fipped.consensus.df) %>% 
  mutate(Type = case_when(nchar(REF) > 1 ~ "Del",
                          nchar(ALT) > 1 ~ "Ins",
                          nchar(REF) == 1 ~ "SNP")) %>% 
  filter(Type == "SNP") %>% 
  mutate(SNV = paste0(REF, POS, ALT))

```

#### Shared Intrahost Variants 

I wanted to see how the frequency of different variants changed throughout the boat between individuals. First, I did this with only the `pysam` variants. Most mutations are private to individuals. Some low frequency mutations are shared. 

```{r Per Pos Frequency - minor only, echo=F, message=F, warning=F, fig.width= 15, fig.height= 15, fig.align="center"}
 
fixed.in.the.boat = selected.variant.df %>% 
  filter(DP >= 100) %>% 
  group_by(SNP) %>% 
  count() %>% 
  filter(n == length(unique(selected.variant.df$ACCESSION))) %>% 
  pull(SNP)

final.variant.df %>% 
  mutate(ACCESSION = paste0("Crewmember-", ACCESSION)) %>% 
  filter(!(SNP %in% fixed.in.the.boat)) %>% 
  select(ACCESSION, SNP, AF) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0) %>% 
  pivot_longer(!ACCESSION, names_to = "SNP", values_to = "AF") %>%
  mutate(POS = as.numeric(parse_number(SNP))) %>%
  mutate(SNP = as.factor(SNP)) %>% 
  ggplot(aes(x = ACCESSION, y = AF)) +
    annotate("rect", xmin = -Inf, xmax =  Inf, ymin = 0, ymax = 1, fill = "#737373") +
    geom_area(aes(group = 1), fill = "#fc9403") +
    geom_line(aes(group = 1), col = "black", size = 0.2) +
    facet_wrap(~fct_reorder(SNP, POS, min), ncol = 5) +
    geom_point(size = 3) +
    ggtitle("Pysam") + 
    xlab("Crew Member") +
    ylab("Allele Frequency") + 
    scale_color_manual(values = c("#bd0000", "#0013bd", "#bd00ba"), name = "Type") +
    scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(0,1) , expand = c(0,0)) +  
    scale_x_discrete(expand = c(0,0), guide = guide_axis(angle = 90)) +    
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=24,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.background = element_rect(fill = NA, color = "black")) +
    theme(axis.text.x = element_text(size = 16)) + 
    theme(panel.spacing = unit(2, "lines")) + 
    theme(axis.ticks.length=unit(.25, "cm")) +
  theme(plot.title = element_text(hjust = 0.5))


ggsave("../../results/figures/pysam_shared_variants_frequency.png", width = 15, height = 15, dpi = 300, units = "in")

```

```{r Ordered Pysam, echo=F, message=F, warning=F, fig.width= 15, fig.height= 15, fig.align="center"}

SNPs = final.variant.df %>% 
  filter(!(SNP %in% fixed.in.the.boat)) %>% 
  pull(SNP) %>% 
  unique()

pysam_plot_list = list()

for (i in 1:length(SNPs)) {
  
  plt = final.variant.df %>% 
    # mutate(ACCESSION = paste0("Crewmember-", ACCESSION)) %>% 
    select(ACCESSION, SNP, AF) %>% 
    pivot_wider(names_from = SNP, values_from = AF, values_fill = 0) %>% 
    pivot_longer(!ACCESSION, names_to = "SNP", values_to = "AF") %>%
    mutate(POS = as.numeric(parse_number(SNP))) %>%
    mutate(SNP = as.factor(SNP)) %>% 
    filter(SNP == SNPs[i]) %>% 
    ggplot(aes(x = reorder(ACCESSION, AF), y = AF)) +
      annotate("rect", xmin = -Inf, xmax =  Inf, ymin = 0, ymax = 1, fill = "#737373") +
      geom_area(aes(group = 1), fill = "#fc9403") +
      geom_line(aes(group = 1), col = "black", size = 0.2) +
      facet_wrap(~fct_reorder(SNP, POS, min), ncol = 5) +
      geom_point(size = 3) +
      xlab("") +
      ylab("") + 
      scale_color_manual(values = c("#bd0000", "#0013bd", "#bd00ba"), name = "Type") +
      scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(0,1) , expand = c(0,0)) +  
      scale_x_discrete(expand = c(0,0), guide = guide_axis(angle = 90)) +    
      theme_classic() +
      theme(legend.position="bottom", legend.box = "horizontal") +
      theme(legend.box.background = element_rect(colour = "black")) +
      theme(text=element_text(size=24,  family="Helvetica")) +
      theme(plot.title = element_text(hjust = 0.5)) +
      theme(panel.background = element_rect(fill = NA, color = "black")) +
      theme(axis.text.x = element_text(size = 12)) + 
      theme(panel.spacing = unit(2, "lines")) + 
      theme(axis.ticks.length=unit(.25, "cm")) +
      theme(plot.title = element_text(hjust = 0.5)) + 
      theme(plot.margin=unit(c(.25, 0.2, -.4, -.25), "cm"))

  pysam_plot_list[[i]] = plt
  
}

final_plot = gridExtra::grid.arrange(arrangeGrob(grobs = pysam_plot_list, 
                          left = textGrob("Allele Frequency", rot = 90, vjust = 1,
                                          gp = gpar(col = "black", fontsize = 24)),
                          bottom = textGrob("Crew Memeber", hjust = .5,
                                            gp = gpar(col = "black", fontsize = 24))))

ggsave(plot = final_plot, "../../results/figures/pysam_shared_variants_frequency.png", width = 15, height = 15, dpi = 300, units = "in")

```

```


Same thing, but for `lofreq`:
```{r Per Pos Frequency - lofreq, echo=F, message=F, warning=F, fig.width= 15, fig.height= 15, fig.align="center"}

caller.final.variant.df %>% 
  filter(Caller == "lofreq") %>% 
  mutate(ACCESSION = paste0("Crewmember-", ACCESSION)) %>% 
  filter(!(SNP %in% fixed.in.the.boat)) %>% 
  select(ACCESSION, SNP, AF) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0) %>% 
  pivot_longer(!ACCESSION, names_to = "SNP", values_to = "AF") %>%
  mutate(POS = as.numeric(parse_number(SNP))) %>%
  mutate(SNP = as.factor(SNP)) %>% 
  ggplot(aes(x = ACCESSION, y = AF)) +
    annotate("rect", xmin = -Inf, xmax =  Inf, ymin = 0, ymax = 1, fill = "#737373") +
    geom_area(aes(group = 1), fill = "#fc9403") +
    geom_line(aes(group = 1), col = "black", size = 0.2) +
    facet_wrap(~fct_reorder(SNP, POS, min), ncol = 5) +
    geom_point(size = 3) +
    ggtitle("Lofreq") + 
    xlab("Patient") +
    ylab("Allele Frequency") + 
    scale_color_manual(values = c("#bd0000", "#0013bd", "#bd00ba"), name = "Type") +
    scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(0,1) , expand = c(0,0)) +  
    scale_x_discrete(expand = c(0,0), guide = guide_axis(angle = 90)) +    
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=24,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.background = element_rect(fill = NA, color = "black")) +
    theme(axis.text.x = element_text(size = 16)) + 
    theme(panel.spacing = unit(2, "lines")) + 
    theme(axis.ticks.length=unit(.25, "cm")) +
  theme(plot.title = element_text(hjust = 0.5))


```

Same thing, but for `varscan`:
```{r Per Pos Frequency - varscan, echo=F, message=F, warning=F, fig.width= 15, fig.height= 15, fig.align="center"}

caller.final.variant.df %>% 
  filter(Caller == "varscan") %>% 
  mutate(ACCESSION = paste0("Crewmember-", ACCESSION)) %>% 
  filter(!(SNP %in% fixed.in.the.boat)) %>% 
  select(ACCESSION, SNP, AF) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0) %>% 
  pivot_longer(!ACCESSION, names_to = "SNP", values_to = "AF") %>%
  mutate(POS = as.numeric(parse_number(SNP))) %>%
  mutate(SNP = as.factor(SNP)) %>% 
  ggplot(aes(x = ACCESSION, y = AF)) +
    annotate("rect", xmin = -Inf, xmax =  Inf, ymin = 0, ymax = 1, fill = "#737373") +
    geom_area(aes(group = 1), fill = "#fc9403") +
    geom_line(aes(group = 1), col = "black", size = 0.2) +
    facet_wrap(~fct_reorder(SNP, POS, min), ncol = 5) +
    geom_point(size = 3) +
    ggtitle("Varscan") + 
    xlab("Patient") +
    ylab("Allele Frequency") + 
    scale_color_manual(values = c("#bd0000", "#0013bd", "#bd00ba"), name = "Type") +
    scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(0,1) , expand = c(0,0)) +  
    scale_x_discrete(expand = c(0,0), guide = guide_axis(angle = 90)) +    
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=24,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.background = element_rect(fill = NA, color = "black")) +
    theme(axis.text.x = element_text(size = 16)) + 
    theme(panel.spacing = unit(2, "lines")) + 
    theme(axis.ticks.length=unit(.25, "cm")) +
  theme(plot.title = element_text(hjust = 0.5))


```

Same thing, but for `ivar`:
```{r Per Pos Frequency - ivar, echo=F, message=F, warning=F, fig.width= 15, fig.height= 15, fig.align="center"}

caller.final.variant.df %>% 
  filter(Caller == "ivar") %>% 
  mutate(ACCESSION = paste0("Crewmember-", ACCESSION)) %>% 
  filter(!(SNP %in% fixed.in.the.boat)) %>% 
  select(ACCESSION, SNP, AF) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0) %>% 
  pivot_longer(!ACCESSION, names_to = "SNP", values_to = "AF") %>%
  mutate(POS = as.numeric(parse_number(SNP))) %>%
  mutate(SNP = as.factor(SNP)) %>% 
  ggplot(aes(x = ACCESSION, y = AF)) +
    annotate("rect", xmin = -Inf, xmax =  Inf, ymin = 0, ymax = 1, fill = "#737373") +
    geom_area(aes(group = 1), fill = "#fc9403") +
    geom_line(aes(group = 1), col = "black", size = 0.2) +
    facet_wrap(~fct_reorder(SNP, POS, min), ncol = 5) +
    geom_point(size = 3) +
    xlab("Patient") +
    ylab("Allele Frequency") + 
    scale_color_manual(values = c("#bd0000", "#0013bd", "#bd00ba"), name = "Type") +
    scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(0,1) , expand = c(0,0)) +  
    scale_x_discrete(expand = c(0,0), guide = guide_axis(angle = 90)) +    
    ggtitle("Ivar") + 
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=24,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.background = element_rect(fill = NA, color = "black")) +
    theme(axis.text.x = element_text(size = 16)) + 
    theme(panel.spacing = unit(2, "lines")) + 
    theme(axis.ticks.length=unit(.25, "cm")) +
  theme(plot.title = element_text(hjust = 0.5))


```

```{r Shared Variants by Method}

# Combine all of the methods
combined.final.variant.df = rbind(caller.final.variant.df, select(mutate(final.variant.df, Caller = "pysam", Type = "SNP"), colnames(caller.final.variant.df)))

# What are the variants that show up in more than one person? 
shared.variants = combined.final.variant.df %>% 
  group_by(POS, REF, ALT, EFFECT, GENE, AA_CHANGE, Caller) %>% 
  count() %>%
  filter(Caller == "pysam") %>% 
  filter(n > 1) %>% 


print(shared.variants)

```

The variants that show up in multiple individuals are mostly the same between the four methods. There are two key differences: (1) One site that is fixed in multiple methods shows up as a Wuhan-1 reversion in `lofreq` - 1108, (2) There are a string of mutations at the end of the genome that show up in `ivar` and `lofreq`. 

Otherwise, there are four shared minor variants that I called with `pysam`: `r (unique(shared.variants$POS))`. Three out four are called with all three methods. One (`14335`) is missing from `varscan`. 

#### Low Frequnecy Shared Variants

```{r Odd Variants}

# Read in the reference as a indexable string
reference = readDNAStringSet("../../config/ref/SARS2.fa")

# Get the shared positions
shared.positions = unique(shared.variants$POS)
shared.snps = shared.variants %>% 
  mutate(SNP = paste0(REF, POS, ALT)) %>% 
  pull(SNP) %>% 
  unique()

# For each of the shared positions, get the sequence context, number of people, strand bias, and position bias
for (i in 1:length(shared.positions)){
  
  pos = shared.positions[i]
  snp = shared.snps[i]
  print(paste("Position:", snp))
  
  # Get the number of people with this variant
  n.crew = final.variant.df %>% filter(POS == pos) %>% pull(ACCESSION) %>% length()
  t.crew = final.variant.df %>% pull(ACCESSION) %>% unique() %>% length()
  print(paste0("   ", n.crew, "/", t.crew, " people have this variant."))
  
  # Get the sequence context
  substring = substr(reference, start = pos-3, stop = pos+3)
  print(paste("   Sequence Context:", substring))
  
  # Get the mean read positions
  readpos.df = pysam.combined.df %>% 
    filter(!(ACCESSION %in% low.coverage.samples)) %>% 
    filter(!(ACCESSION %in% poor.concordance.samples)) %>% 
    filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
    filter(POS == pos) %>% 
    select(POS, MEAN_READPOS.one, MEAN_READPOS.two) %>% 
    pivot_longer(!POS, names_to = "REP", values_to = "READ_POSITION") %>% 
    group_by(REP) %>% 
    summarize(Mean = mean(READ_POSITION), .groups = 'drop')
  
  print(paste("   Mean Read Position Replicate 1:", filter(readpos.df, REP == "MEAN_READPOS.one")$Mean))
  print(paste("   Mean Read Position Replicate 2:", filter(readpos.df, REP == "MEAN_READPOS.two")$Mean))


  # Get the strand bias
  readpos.df = pysam.combined.df %>% 
    filter(!(ACCESSION %in% low.coverage.samples)) %>% 
    filter(!(ACCESSION %in% poor.concordance.samples)) %>% 
    filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
    filter(POS == pos) %>% 
    select(POS, STRAND_RATIO.one, STRAND_RATIO.two) %>% 
    pivot_longer(!POS, names_to = "REP", values_to = "BIAS") %>% 
    group_by(REP) %>% 
    summarize(Mean = mean(BIAS), .groups = 'drop')
  
  print(paste("   Mean Strand Ratio Replicate 1:", filter(readpos.df, REP == "STRAND_RATIO.one")$Mean))
  print(paste("   Mean Strand Ratio Replicate 2:", filter(readpos.df, REP == "STRAND_RATIO.two")$Mean))
  print("----------------------------------------------")

}

```

(1) `C9502T` is clearly a homopolymer region. 

(2) `G14335T` might be considered a homopolymeric? However, the strand bias and position bias are pretty off the charts for this variant. 
(3) `A4229C` is similar to above. There is a pretty clear strand/read position bias. 

(4) `T18402A` is passable. 

Oddly, in all four cases, the base chages to the same base as its immediate neighbor. 

#### Replicate Frequency differences. 

This is part of the final supplementary figure showing that the freqency betweem replicates is off for these alleles. 

```{r Frequency Discrepency, warning=F, message=F, fig.width= 9, fig.height= 7, fig.align="center"}

pysam.combined.df %>% 
    filter(!(ACCESSION %in% low.coverage.samples)) %>% 
    filter(!(ACCESSION %in% poor.concordance.samples)) %>% 
    filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
    filter(POS %in% shared.positions) %>% 
    mutate(SNP = paste0(REF, POS, ALT)) %>% 
    ggplot(aes(y = AF.one, x = AF.two)) +
      geom_abline(intercept = 0, col = "red", size = 1, linetype = 2) + 
      geom_point(size = 2) + 
      geom_smooth(method = "lm", se = FALSE) +
      facet_wrap(~SNP) +
      scale_y_continuous(limits = c(0, .10),  labels = scales::percent) +
      scale_x_continuous(limits = c(0, .10), labels = scales::percent, guide = guide_axis(angle = 25)) +
      xlab("Allele Frequency Replicate Two") + 
      ylab("Allele Frequency Replicate One") + 
      theme_bw(20)

ggsave("../../results/figures/discrepency_in_frequency_between_replicates.png", width = 9, height = 7, dpi = 300, units = "in")

```

What is the "null" distributions for allele frequency correlation? 

```{r Frequency Null, warning=F, message=F, fig.width= 9, fig.height= 7, fig.align="center"}

snps.for.null = final.variant.df %>% 
  filter(!SNP %in% shared.snps) %>% 
  pull(POS) %>% 
  unique()

# Expected correlation 
pysam.combined.df %>% 
    filter(!(ACCESSION %in% low.coverage.samples)) %>% 
    filter(!(ACCESSION %in% poor.concordance.samples)) %>% 
    filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
    filter(POS %in% snps.for.null) %>% 
    mutate(SNP = paste0(REF, POS, ALT)) %>% 
    ggplot(aes(y = AF.one, x = AF.two)) +
      geom_abline(intercept = 0, col = "red", size = 1, linetype = 2) + 
      geom_point(size = 2) + 
      geom_smooth(method = "lm", se = FALSE) +
      scale_y_continuous(limits = c(0, 1),  labels = scales::percent) +
      scale_x_continuous(limits = c(0, 1), labels = scales::percent, guide = guide_axis(angle = 25)) +
      xlab("Allele Frequency Replicate Two") + 
      ylab("Allele Frequency Replicate One") + 
      theme_bw(20)

```

#### END

```{r}
write_csv(final.variant.df, path = "../../config/data/intrahost_variants.csv")
```

