---
title: "Varaint Features"
author: "Will Hannon"
date: "1/11/2021"
output: html_document
---

This notebook contains an analysis of the features of minor variants for the `SARS-CoV-2_Bottleneck` project. The main goals of this notebook are to characterize: 

1. Identify distribution of minor variants over the genome.
2. Determine the ratio of tansitions to transversions. 
3. Show the mutation specturm.

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "scales", "RColorBrewer")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Filepaths, include=FALSE}

## ==== File paths input data ==== ##

# Data from pysam/python script. It contains the % of each nucleotide
# at every covered position in the genome. This is a huge file.
pysam.data = "../../results/pileup/pysam-variants.csv"

# Data from lofreq
variant.data = "../../results/variants/variants.csv"

# Samtools average depth
average.depth.data = "../../results/coverage/merged.average.depth" 

# Annotation of SARS-CoV-2 gene positions
annotation.data = "../../config/sars_cov_2_annot.csv"

```

```{r Format data, include=FALSE, warning=FALSE, message=FALSE}

# Import the data splitting the accession ID into the Accession and the Replicate
pysam.df = read_csv(pysam.data) %>% 
  separate(ACCESSION, into = c("ACCESSION", "REPLICATE"), sep = "_") %>% 
  mutate(REPLICATE = as.factor(REPLICATE))

# Get only the consensus alleles
pysam.consensus.df = pysam.df %>% 
  filter(CONS == TRUE)

# Get only the SNPs
pysam.snp.df = pysam.df %>% 
  filter(SNP == TRUE) 

# Get the prevalence of SNPs between replicates by filtering 
# and joining by the samples, position in the genome, reference
# allele (this is redundant), and the alternative allele (SNP).
pysam.combined.snps.df = full_join(
  filter(pysam.snp.df, REPLICATE == "1"),
  filter(pysam.snp.df, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT"),
  suffix = c(".one", ".two")
  ) %>% 
  select(!c("REPLICATE.one", "REPLICATE.two", "SNP.one", "SNP.two", "CONS.one", "CONS.two")) %>% 
  mutate_each(list( ~ replace(., which(is.na(.)), 0)))

# Varinat data from lofreq and varscan
variant.df = read_csv(variant.data) %>% 
  filter(Library == 1) %>% 
  filter(Caller == "lofreq") %>% 
  rename(ACCESSION = "spID", REPLICATE = "Replicate") %>% 
  select(ACCESSION, REPLICATE, POS, DP, REF, ALT, AF) %>% 
  mutate(SNP = TRUE, CONS = ifelse(AF >= 0.5, TRUE, FALSE))

variant.combined.snps.df = full_join(
  filter(variant.df, REPLICATE == "1"),
  filter(variant.df, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT"),
  suffix = c(".one", ".two")
  ) %>% 
  select(!c("REPLICATE.one", "REPLICATE.two", "SNP.one", "SNP.two", "CONS.one", "CONS.two")) %>% 
  mutate_each(list( ~ replace(., which(is.na(.)), 0)))


# Low quality samples 
average.depth.df = read.table(average.depth.data, header = T) %>% 
  separate(Accession, into = c("Accession", "Replicate")) %>% 
  mutate(Run = paste0(Accession, Replicate))

low.coverage.samples = average.depth.df %>% filter(Percent < 80) %>% pull(Accession)
low.coverage.runs = average.depth.df %>% filter(Percent < 80) %>% pull(Run)

```

```{r Annotations, echo=F, message=F, warning=F, fig.width=25, fig.height=20, fig.align="center"}

# Read in the SARS-CoV-2 gene annotations
annotation.df = read_csv(annotation.data)

# Generate positions for lables 
label = annotation.df %>% 
  mutate(POS = (start+end)/2) %>% 
  mutate(AF = -(.11/2)) %>% 
  select(gene, POS, AF) 

# Get the interval positions for these genes 
ORF1ab = annotation.df %>% dplyr::filter(gene == "Orf1ab")
Spike = annotation.df %>% dplyr::filter(gene == "Spike")
ORF3a = annotation.df %>% dplyr::filter(gene == "ORF3a")
Envelope = annotation.df %>% dplyr::filter(gene == "Envelope")
Membrane = annotation.df %>% dplyr::filter(gene == "Membrane")
ORF6 = annotation.df %>% dplyr::filter(gene == "ORF6")
ORF7a = annotation.df %>% dplyr::filter(gene == "ORF7a")
ORF8 = annotation.df %>% dplyr::filter(gene == "ORF8")
Nucleocapsid = annotation.df %>% dplyr::filter(gene == "Nucleocapsid")
ORF10 = annotation.df %>% dplyr::filter(gene == "ORF10")

# Color scheme
effect_colors = c("#e2e627", "#13a13b", "#990505")
```

```{r}

# Final samples and mutations to analyze: 
# Remove low coverage samples
# Remove alleles below 2%
# Remove alleles with less than 100X coverage
# Remove samples with poor replicate condordance: 10089, 10117, *10088*

selected.variant.df = pysam.combined.snps.df %>% 
  mutate(Coverage = ifelse(ACCESSION %in% low.coverage.samples, "Low", "High")) %>% 
  filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  filter(Coverage == "High") %>% 
  filter(!(ACCESSION %in% c("10089", "10117", "10088"))) %>% 
  # Allele frequency is the average of the two replicates
  mutate(AF = (AF.one + AF.two)/2) %>% 
  select(ACCESSION, POS, REF, ALT, AF, AA_CHANGE = "AA_CHANGE.one", EFFECT = "EFFECT.one", GENE = "GENE.one") 

# Filp the consensus variants
fipped.cons = selected.variant.df %>% 
  filter(AF >= 0.5) %>% 
  rename(ALT = "REF", REF = "ALT") %>% 
  mutate(AF = (1 - AF)) %>% 
  filter(AF >= 0.02) %>% 
  separate(AA_CHANGE, into = c("WT", "AAPOS"),
           sep = "((?<=[A-Za-z])(?=[0-9]))") %>% 
  separate(AAPOS, into = c("AAPOS", "MUT"),
           sep = "((?<=[0-9])(?=[A-Za-z]))") %>% 
  mutate(AA_CHANGE = paste0(MUT, AAPOS, WT)) %>% 
  select(!c("MUT", "WT", "AAPOS"))

final.variant.df = rbind(filter(selected.variant.df, AF < 0.5), fipped.cons) %>% 
  mutate(SNV = paste0(REF, POS, ALT))

```

```{r}

average.number.of.minor.variants = final.variant.df %>% 
  group_by(ACCESSION) %>% 
  count() %>% 
  pull(n) %>% 
  mean(.) %>% 
  round(.)

final.variant.df %>% 
  group_by(ACCESSION) %>% 
  count() %>% 
  ggplot(aes(x = n, y = reorder(ACCESSION, -n))) + 
    geom_bar(stat = 'identity') + 
    geom_text(aes(x = 4.5, y = "10040"), label = paste("Average ~ ", average.number.of.minor.variants), size = 7) +
    ylab('Crewmember') + 
    xlab('Number of minor variants') +
    theme_bw(18)

ggsave(filename = "../../config/figures/figure-3-B.svg", width = 6, height = 4)
```

There is a mean of ~ 4 minor variants per sample. 


```{r}

final.variant.df %>% 
  ggplot(aes(x = AF)) + 
    geom_histogram(bins = 20) + 
    ylab("Count") + 
    xlab("Allele Frequency") +
    theme_bw(18) +
    theme(plot.title = element_text(hjust = 0.5)) 
  
ggsave(filename = "../../config/figures/figure-3-C.svg", width = 6, height = 4)

```


```{r Lofreq Minor Variants Distribution, echo=F, message=F, warning=F, fig.width=11, fig.height=4, fig.align="center"}

label = annotation.df %>% 
  mutate(POS = (start+end)/2) %>% 
  mutate(AF = -(.1/2)) %>% 
  select(gene, POS, AF) 

selected.variant.df %>% 
  filter(AF <= .98) %>% 
  ggplot(aes(x = POS, y = AF, col = ACCESSION))+
    geom_point(size = 4, stroke = 2, alpha = 0.5) +
    xlab("Position") +
    ylab("Allele Frequency") +
    scale_x_continuous(breaks=c(0,5000,10000,15000,20000,25000,30000))  +
    scale_color_discrete(name = "Crewmembers") +
    annotate("rect", xmin = 0, xmax = ORF1ab[[2]]-1, ymin = -.1, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate("rect", xmin = ORF1ab[[2]], xmax = ORF1ab[[3]], ymin =  -.1, ymax = -.01 ,
               alpha = .2, fill = "dodgerblue2", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][1], y = label[[3]][1], label = label[[1]][1], family = "mono", size = 7) +
    annotate("rect", xmin = Spike[[2]], xmax = Spike[[3]], ymin =   -.1, ymax = -.01 ,
             alpha = .2, fill = "yellow3", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][2], y = label[[3]][2], label = label[[1]][2], family = "mono", size = 7) +
    annotate("rect", xmin = ORF3a[[2]], xmax = ORF3a[[3]], ymin =   -.1, ymax = -.01 ,
             alpha = .2, fill = "#FF7F00", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][3], y = label[[3]][3], label = "3a", family = "mono", size = 7) +
    annotate("rect", xmin = Envelope[[2]], xmax = Envelope[[3]], ymin =   -.1, ymax = -.01 ,
             alpha = .2, fill =  "#6A3D9A", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][4], y = label[[3]][4], label = "E", family = "mono", size = 7) +
    annotate("rect", xmin = Membrane[[2]], xmax = Membrane[[3]], ymin =   -.1, ymax = -.01 ,
             alpha = .2, fill =  "green4", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][5], y = label[[3]][5], label = "M", family = "mono", size = 7) +
    annotate("rect", xmin = ORF6[[2]], xmax = ORF6[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "#E31A1C", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][6], y = label[[3]][6], label = "6", family = "mono", size = 7) +
    annotate("rect", xmin = ORF7a[[2]], xmax = ORF7a[[3]], ymin =   -.1, ymax = -.01 ,
             alpha = .2, fill = "#FB9A99", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][7], y = label[[3]][7], label = "7a", family = "mono", size = 7) +
    annotate("rect", xmin = ORF8[[2]], xmax = ORF8[[3]], ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "orchid1", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][8], y = label[[3]][8], label = "8", family = "mono", size = 7) +
    annotate("rect", xmin = Nucleocapsid[[2]], xmax = Nucleocapsid[[3]], ymin =   -.1, ymax = -.01 ,
             alpha = .2, fill = "maroon", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][9], y = label[[3]][9], label = "N", family = "mono", size = 7) +
    annotate("rect", xmin = ORF10[[2]], xmax = ORF10[[3]], ymin =   -.1, ymax = -.01 ,
             alpha = .2, fill =  "skyblue2", col = "black", size = .1) +
    annotate("rect", xmin = ORF10[[3]]+1, xmax = 29903, ymin =  -.1, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate(geom = "text", x = -200, y = label[[3]][1], label = "5'", family = "mono", size = 7) +
    annotate(geom = "text", x = 30200, y = label[[3]][1], label = "3'", family = "mono", size = 7) +
    #annotate(geom = "text", x = label[[2]][10], y = label[[3]][10], label = "10", family = "mono") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=24)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x = element_text(hjust= 1)) 

ggsave("../../config/figures/figure-3-A.svg", width = 18, height = 6, dpi = 300, units = "in")

```


```{r}

final.variant.df %>% 
  group_by(SNV) %>% 
  count() %>% 
  arrange(-n) %>% 
  ggplot(aes(x = n)) + 
    geom_histogram(bins = 30) + 
    xlab("Number of samples sharing site") +
    ylab("Number of sites") +
    scale_x_continuous(breaks = c(2, 4, 6, 8)) +
    theme_bw(20)
  
```


```{r Percent by sample, message=FALSE, warning=FALSE, fig.width=8, fig.height=5, fig.align='center'}

shared.snps = final.variant.df %>% 
  group_by(SNV) %>% 
  count() %>% 
  filter(n > 1) %>% 
  pull(SNV)

shared.snps.df = final.variant.df[which(final.variant.df$SNV %in% shared.snps),]

shared.snps.df %>% 
  ggplot(aes(x = reorder(ACCESSION, AF), y = AF)) +
    geom_point(size = 2) + 
    facet_wrap(~SNV) + 
    xlab("Sample") + 
    ylab("Allele Frequency") +
    scale_y_continuous(labels = percent, limits = c(0, .15)) +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1))




```

Does T18402A ever reach 100% frequency? 

```{r Probably an artifact, message=FALSE, warning=FALSE, fig.width=6, fig.height=3, fig.align='center'}

good.runs = average.depth.df %>% 
  mutate(Run = paste(Accession, Replicate, sep = "_")) %>% 
  filter(Percent >= 80) %>% 
  pull(Run)

pysam.df %>% 
  mutate(Run = paste(ACCESSION, REPLICATE, sep = "_")) %>% 
  filter(Run %in% good.runs) %>% 
  filter(POS == 18402) %>% 
  filter(AF >= 0.02) %>% 
  ggplot(aes(x = reorder(ACCESSION, AF), y = AF)) +
    geom_point(size = 3) + 
    facet_wrap(~REPLICATE) +
    xlab("Sample") + 
    ylab("Allele Frequency") +
    ggtitle("T18402A is likely an artifact") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1))



```


```{r Minor Varaints, message=FALSE, warning=FALSE, fig.width=10, fig.height=30, fig.align='center'}

high.confidence.snps = final.variant.df %>% 
  pull(POS) %>% 
  unique(.) 

good.runs = average.depth.df %>% 
  mutate(Run = paste(Accession, Replicate, sep = "_")) %>% 
  filter(Percent >= 80) %>% 
  pull(Run)

pysam.df %>% 
  mutate(Run = paste(ACCESSION, REPLICATE, sep = "_")) %>% 
  filter(Run %in% good.runs) %>% 
  filter(POS %in% high.confidence.snps) %>% 
  filter(AF >= 0.02) %>% 
  ggplot(aes(x = reorder(ACCESSION, AF), y = AF)) +
    geom_point(size = 3) + 
    facet_grid(cols = vars(REPLICATE), rows = vars(POS)) +
    xlab("Sample") + 
    ylab("Allele Frequency") +
    theme_bw() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=15)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1))

```


```{r Fixed Mutations, message=FALSE, warning=FALSE, fig.width=10, fig.height=30, fig.align='center'}

high.confidence.fixed = selected.variant.df %>% 
  filter(AF >= 0.5) %>% 
  pull(POS) %>% 
  unique(.) 

good.runs = average.depth.df %>% 
  mutate(Run = paste(Accession, Replicate, sep = "_")) %>% 
  filter(Percent >= 80) %>% 
  pull(Run)

pysam.df %>% 
  mutate(Run = paste(ACCESSION, REPLICATE, sep = "_")) %>% 
  filter(Run %in% good.runs) %>% 
  filter(POS %in% high.confidence.fixed) %>% 
  filter(AF >= 0.02) %>% 
  ggplot(aes(x = reorder(ACCESSION, AF), y = AF)) +
    geom_point(size = 3) + 
    facet_grid(cols = vars(REPLICATE), rows = vars(POS)) +
    xlab("Sample") + 
    ylab("Allele Frequency") +
    theme_bw() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=15)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1))

```


```{r Summary Figure, message=FALSE, warning=FALSE, fig.width=16, fig.height=4, fig.align='center'}

# Good samples where both replicates have more than 80% coverage (14)
samples.to.include = average.depth.df %>% 
  mutate(Run = paste(Accession, Replicate, sep = "_")) %>% 
  filter(Percent >= 80) %>% 
  group_by(Accession) %>% 
  count() %>% 
  filter(n == 2) %>% 
  pull(Accession)

# Mutations from the reference
mutations.to.include = selected.variant.df %>% pull(POS)

pysam.df %>% 
  filter(ACCESSION %in% samples.to.include) %>% 
  filter(POS %in% mutations.to.include) %>% 
  filter(AF > 0.02) %>% 
  ggplot(aes(x = reorder(as.factor(POS), AF), y = AF, col = REPLICATE)) + 
    geom_jitter(size = 5, width = .15) +
    xlab("Genomic Position with SNP") + 
    ylab("Allele Frequency") +
    scale_color_manual(values = c("#3232a8", "#b03733"), name = "Replicate") +
    theme_bw() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=28)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1)) +
    theme(panel.grid.major.x = element_line(size = 1.5, linetype = "solid"))

  

```


```{r Summary Figure, message=FALSE, warning=FALSE, fig.width=18, fig.height=5, fig.align='center'}

  
pallete = c("#4287f5", 
            "#b5183a", 
            "#eaed24", 
            "#22e051", 
            "#dbab18", 
            "#e815c5",
            "#38fffc", 
            "#ff38b6", 
            "#bda679", 
            "#ffb3b3", 
            "#12128a", 
            "#a1a1a1", 
            "#000000",
            "#156b00")

# Mutations from the reference exlucing the 'boat consensus' 
boat.consensus = selected.variant.df %>% 
  filter(AF > 0.5) %>% 
  mutate(SNV = paste0(REF, POS, ALT)) %>% 
  group_by(SNV) %>% 
  count() %>% 
  filter(n == 12) %>% 
  pull(SNV)

positions.to.include = selected.variant.df %>% 
  mutate(SNV = paste0(REF, POS, ALT)) %>% 
  filter(!SNV %in% boat.consensus) %>% 
  pull(POS)

pysam.combined.snps.df %>% 
  filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  mutate(AF = (AF.one + AF.two)/2) %>% 
  select(ACCESSION, POS, REF, ALT, AF, AA_CHANGE = "AA_CHANGE.one", EFFECT = "EFFECT.one", GENE = "GENE.one") %>% 
  mutate(SNV = paste0(REF, POS, ALT)) %>% 
  filter(AF > 0.02) %>%   
  filter(ACCESSION %in% samples.to.include) %>% 
  filter(POS %in% positions.to.include) %>% 
  ggplot(aes(x = reorder(as.factor(POS), AF), y = AF, col = ACCESSION)) + 
    geom_jitter(size = 5, width = .25) +
    xlab("Genomic Position with SNP") + 
    ylab("Allele Frequency") + 
    scale_color_manual(values = pallete) +
    theme_bw() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=28)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1)) +
    theme(panel.grid.major.x = element_line(size = 1.5, linetype = "solid")) +
    theme(legend.position = "none")


```

```{r}

pysam.df %>% 
  filter(ACCESSION %in% samples.to.include) %>% 
  filter(POS %in% mutations.to.include) %>% 
  filter(AF >= 0.02) %>% 
  mutate(High_Confidence = ifelse(AF >= 0.05, "YES", "NO")) %>% 
  ggplot(aes(x = STRAND_RATIO, fill = High_Confidence)) + 
    geom_histogram() +
    xlab("Strand Ratio") + 
    ylab("Count") + 
    scale_fill_manual(values = c("#fa130f", "#262aff"), name = "Greater than 5%") +
    theme_bw(15)

pysam.df %>% 
  filter(ACCESSION %in% samples.to.include) %>% 
  filter(POS %in% mutations.to.include) %>% 
  filter(AF >= 0.02) %>% 
  mutate(High_Confidence = ifelse(AF >= 0.05, "YES", "NO")) %>% 
  ggplot(aes(x = MEAN_READPOS, fill = High_Confidence)) + 
    geom_histogram()+
    xlab("Mean Read Position") + 
    ylab("Count") + 
    scale_fill_manual(values = c("#fa130f", "#262aff"), name = "Greater than 5%") +
    theme_bw(15)

```

