---
title: "Varaint Features"
author: "Will Hannon"
date: "1/11/2021"
output: html_document
---

This notebook contains an analysis of the features of minor variants for the `SARS-CoV-2_Bottleneck` project. The main goals of this notebook are to characterize: 

1. Identify distribution of minor variants over the genome.
2. Determine the ratio of tansitions to transversions. 
3. Show the mutation specturm.

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "scales")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Filepaths, include=FALSE}

## ==== File paths input data ==== ##

# Data from pysam/python script. It contains the % of each nucleotide
# at every covered position in the genome. This is a huge file.
pysam.data = "../../results/pileup/pysam-variants.csv"
# Data from lofreq
variant.data = "../../results/variants/variants.csv"
# Samtools average depth
average.depth.data = "../../results/coverage/merged.average.depth" 

```

```{r Format data, include=FALSE, warning=FALSE, message=FALSE}

# Import the data splitting the accession ID into the Accession and the Replicate
pysam.df = read_csv(pysam.data) %>% 
  separate(ACCESSION, into = c("ACCESSION", "REPLICATE"), sep = "_") %>% 
  mutate(REPLICATE = as.factor(REPLICATE))

# Get only the consensus alleles
pysam.consensus.df = pysam.df %>% 
  filter(CONS == TRUE)

# Get only the SNPs
pysam.snp.df = pysam.df %>% 
  filter(SNP == TRUE) 

# Get the prevalence of SNPs between replicates by filtering 
# and joining by the samples, position in the genome, reference
# allele (this is redundant), and the alternative allele (SNP).
pysam.combined.snps.df = full_join(
  filter(pysam.snp.df, REPLICATE == "1"),
  filter(pysam.snp.df, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT"),
  suffix = c(".one", ".two")
  ) %>% 
  select(!c("REPLICATE.one", "REPLICATE.two", "SNP.one", "SNP.two", "CONS.one", "CONS.two")) %>% 
  mutate_each(list( ~ replace(., which(is.na(.)), 0)))

# Varinat data from lofreq and varscan
variant.df = read_csv(variant.data) %>% 
  filter(Library == 1) %>% 
  filter(Caller == "lofreq") %>% 
  rename(ACCESSION = "spID", REPLICATE = "Replicate") %>% 
  select(ACCESSION, REPLICATE, POS, DP, REF, ALT, AF) %>% 
  mutate(SNP = TRUE, CONS = ifelse(AF >= 0.5, TRUE, FALSE))

variant.combined.snps.df = full_join(
  filter(variant.df, REPLICATE == "1"),
  filter(variant.df, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT"),
  suffix = c(".one", ".two")
  ) %>% 
  select(!c("REPLICATE.one", "REPLICATE.two", "SNP.one", "SNP.two", "CONS.one", "CONS.two")) %>% 
  mutate_each(list( ~ replace(., which(is.na(.)), 0)))


# Low quality samples 
average.depth.df = read.table(average.depth.data, header = T) %>% 
  separate(Accession, into = c("Accession", "Replicate")) %>% 
  mutate(Run = paste0(Accession, Replicate))

low.coverage.samples = average.depth.df %>% filter(Percent < 80) %>% pull(Accession)
low.coverage.runs = average.depth.df %>% filter(Percent < 80) %>% pull(Run)

```

```{r}

# Final samples and mutations to analyze: 
# Remove low coverage samples
# Remove alleles below 2%
# Remove alleles with less than 100X coverage
# Remove samples with poor replicate condordance: 10089, 10117, *10088*

selected.variant.df = pysam.combined.snps.df %>% 
  mutate(Coverage = ifelse(ACCESSION %in% low.coverage.samples, "Low", "High")) %>% 
  filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  filter(Coverage == "High") %>% 
  filter(!(ACCESSION %in% c("10089", "10117"))) %>% 
  # Allele frequency is the average of the two replicates
  mutate(AF = (AF.one + AF.two)/2) %>% 
  select(ACCESSION, POS, REF, ALT, AF, AA_CHANGE = "AA_CHANGE.one", EFFECT = "EFFECT.one", GENE = "GENE.one") 

# Filp the consensus variants
fipped.cons = selected.variant.df %>% 
  filter(AF >= 0.5) %>% 
  rename(ALT = "REF", REF = "ALT") %>% 
  mutate(AF = (1 - AF)) %>% 
  filter(AF >= 0.02) %>% 
  separate(AA_CHANGE, into = c("WT", "AAPOS"),
           sep = "((?<=[A-Za-z])(?=[0-9]))") %>% 
  separate(AAPOS, into = c("AAPOS", "MUT"),
           sep = "((?<=[0-9])(?=[A-Za-z]))") %>% 
  mutate(AA_CHANGE = paste0(MUT, AAPOS, WT)) %>% 
  select(!c("MUT", "WT", "AAPOS"))

final.variant.df = rbind(filter(selected.variant.df, AF < 0.5), fipped.cons) %>% 
  mutate(SNV = paste0(REF, POS, ALT))

```

```{r}

average.number.of.minor.variants = final.variant.df %>% 
  group_by(ACCESSION) %>% 
  count() %>% 
  pull(n) %>% 
  mean(.) %>% 
  round(.)

final.variant.df %>% 
  group_by(ACCESSION) %>% 
  count() %>% 
  ggplot(aes(x = n, y = reorder(ACCESSION, -n))) + 
    geom_bar(stat = 'identity') + 
    ylab('Sample') + 
    xlab('Number of minor variants') +
    theme_bw(18)

```

There is a mean of ~ 4 minor variants per sample. 


```{r}

final.variant.df %>% 
  ggplot(aes(x = AF)) + 
    geom_histogram(bins = 20) + 
    ylab("Count") + 
    xlab("Allele Frequency") +
    theme_bw(18) +
    theme(plot.title = element_text(hjust = 0.5)) 
  
```

```{r}

final.variant.df %>% 
  group_by(SNV) %>% 
  count() %>% 
  arrange(-n) %>% 
  ggplot(aes(x = n)) + 
    geom_histogram(bins = 30) + 
    xlab("Number of samples sharing site") +
    ylab("Number of sites") +
    scale_x_continuous(breaks = c(2, 4, 6, 8)) +
    theme_bw(20)
  
```


```{r Percent by sample, message=FALSE, warning=FALSE, fig.width=8, fig.height=5, fig.align='center'}

shared.snps = final.variant.df %>% 
  group_by(SNV) %>% 
  count() %>% 
  filter(n > 1) %>% 
  pull(SNV)

shared.snps.df = final.variant.df[which(final.variant.df$SNV %in% shared.snps),]

shared.snps.df %>% 
  ggplot(aes(x = reorder(ACCESSION, AF), y = AF)) +
    geom_point(size = 2) + 
    facet_wrap(~SNV) + 
    xlab("Sample") + 
    ylab("Allele Frequency") +
    scale_y_continuous(labels = percent, limits = c(0, .15)) +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1))




```

Does T18402A ever reach 100% frequency? 

```{r Probably an artifact, message=FALSE, warning=FALSE, fig.width=6, fig.height=3, fig.align='center'}

good.runs = average.depth.df %>% 
  mutate(Run = paste(Accession, Replicate, sep = "_")) %>% 
  filter(Percent >= 80) %>% 
  pull(Run)

pysam.df %>% 
  mutate(Run = paste(ACCESSION, REPLICATE, sep = "_")) %>% 
  filter(Run %in% good.runs) %>% 
  filter(POS == 18402) %>% 
  filter(AF >= 0.02) %>% 
  ggplot(aes(x = reorder(ACCESSION, AF), y = AF)) +
    geom_point(size = 3) + 
    facet_wrap(~REPLICATE) +
    xlab("Sample") + 
    ylab("Allele Frequency") +
    ggtitle("T18402A is likely an artifact") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1))



```


```{r Minor Varaints, message=FALSE, warning=FALSE, fig.width=10, fig.height=30, fig.align='center'}

high.confidence.snps = final.variant.df %>% 
  pull(POS) %>% 
  unique(.) 

good.runs = average.depth.df %>% 
  mutate(Run = paste(Accession, Replicate, sep = "_")) %>% 
  filter(Percent >= 80) %>% 
  pull(Run)

pysam.df %>% 
  mutate(Run = paste(ACCESSION, REPLICATE, sep = "_")) %>% 
  filter(Run %in% good.runs) %>% 
  filter(POS %in% high.confidence.snps) %>% 
  filter(AF >= 0.02) %>% 
  ggplot(aes(x = reorder(ACCESSION, AF), y = AF)) +
    geom_point(size = 3) + 
    facet_grid(cols = vars(REPLICATE), rows = vars(POS)) +
    xlab("Sample") + 
    ylab("Allele Frequency") +
    theme_bw() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=15)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1))

```


```{r Fixed Mutations, message=FALSE, warning=FALSE, fig.width=10, fig.height=30, fig.align='center'}

high.confidence.fixed = selected.variant.df %>% 
  filter(AF >= 0.5) %>% 
  pull(POS) %>% 
  unique(.) 

good.runs = average.depth.df %>% 
  mutate(Run = paste(Accession, Replicate, sep = "_")) %>% 
  filter(Percent >= 80) %>% 
  pull(Run)

pysam.df %>% 
  mutate(Run = paste(ACCESSION, REPLICATE, sep = "_")) %>% 
  filter(Run %in% good.runs) %>% 
  filter(POS %in% high.confidence.fixed) %>% 
  filter(AF >= 0.02) %>% 
  ggplot(aes(x = reorder(ACCESSION, AF), y = AF)) +
    geom_point(size = 3) + 
    facet_grid(cols = vars(REPLICATE), rows = vars(POS)) +
    xlab("Sample") + 
    ylab("Allele Frequency") +
    theme_bw() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=15)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(axis.text.x=element_text(angle=45, hjust=1))

```