---
title: "4. Intrahost Variant Features"
author: "Will Hannon"
date: "1/11/2021"
output: html_document
---

The goal of this notebook is to look at the features of the intrahost variants from each of the boat passangers. The idea is to see where these fall in the genome, how many are there per sample, and what's the distribution of frequencies. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "scales", "RColorBrewer", "ggrepel")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Filepaths, include=FALSE}

## ==== File paths input data ==== ##

# Data from pysam/python script. It contains the % of each nucleotide
# at every covered position in the genome. This is a huge file.
pysam.data = "../../results/pileup/pysam-variants.csv"

# Data from lofreq, varscan, and ivar
variant.data = "../../results/variants/variants.csv"

# Samtools average depth
average.depth.data = "../../results/coverage/merged.average.depth" 

# Annotation of SARS-CoV-2 gene positions
annotation.data = "../../config/data/sars_cov_2_annot.csv"

# Sample data
sample_metadata = "../../config/data/Boat_Sample_Metadata.csv"

```

I'll format the variants that I called using the pileup function of the Python library `Pysam`. I'll also join the replicates.

```{r Format Pysam data, warning=FALSE, message=FALSE}

# Import the data splitting the accession ID into the Accession and the Replicate
pysam.df = read_csv(pysam.data) %>% 
  separate(ACCESSION, into = c("ACCESSION", "REPLICATE"), sep = "_") %>% 
  mutate(REPLICATE = as.factor(REPLICATE))

# Join the custom pysam SNPs by replicate
pysam.to.join.df = pysam.df %>% 
  select(!c("SNP", "CONS"))


pysam.combined.df = full_join(
  filter(pysam.to.join.df, REPLICATE == "1"),
  filter(pysam.to.join.df, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT", "EFFECT", "GENE", "CODON_POS", "AA_CHANGE"),
  suffix = c(".one", ".two")
  ) %>% 
  select(!c("REPLICATE.one", "REPLICATE.two")) %>% 
  mutate(AF.one = if_else(is.na(AF.one), 0, AF.one)) %>% 
  mutate(AF.two = if_else(is.na(AF.two), 0, AF.two)) %>% 
  mutate(DP.one = if_else(is.na(DP.one), 0, DP.one)) %>% 
  mutate(DP.two = if_else(is.na(DP.two), 0, DP.two)) 


```

Then, I'll do the sample with the variants called with three variant calling softwares: `lofreq`, `ivar`, and `varscan`. Here, I also standardize the column names with those of the pysam variants. 

```{r Format Variant Data, warning=F, message=F}

variant.df = read_csv(variant.data) %>% 
  select(!c("Library", "Path", "Aligner", "LibraryLayout", "Virus", "Host", "Source", "Percent", "spID", "Replicate")) %>% 
  distinct() %>%  # Duplicated columns when you merge with the sample data. 
  separate(Accession, into = c("ACCESSION", "REPLICATE"), sep = "_") %>% 
  dplyr::rename("AA_CHANGE" = AA_Change, "EFFECT" = Effect, "GENE" = Gene_Name)

variant.df.to.join = variant.df %>% 
  select(!c("EFFECT", "GENE", "AA_CHANGE"))

variant.combined.df = full_join(
  filter(variant.df.to.join, REPLICATE == "1"),
  filter(variant.df.to.join, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT", "Caller"),
  suffix = c(".one", ".two")
  ) %>% 
  select(!c("REPLICATE.one", "REPLICATE.two")) %>% 
  mutate(AF.one = if_else(is.na(AF.one), 0, AF.one)) %>% 
  mutate(AF.two = if_else(is.na(AF.two), 0, AF.two)) %>% 
  mutate(DP.one = if_else(is.na(DP.one), 0, DP.one)) %>% 
  mutate(DP.two = if_else(is.na(DP.two), 0, DP.two)) 

```

```{r Annotations, message=F, warning=F, fig.width=25, fig.height=20, fig.align="center"}

# Read in the SARS-CoV-2 gene annotations
annotation.df = read_csv(annotation.data)

# Generate positions for lables 
label = annotation.df %>% 
  mutate(POS = (start+end)/2) %>% 
  mutate(AF = -(.11/2)) %>% 
  select(gene, POS, AF) 

# Get the interval positions for these genes 
ORF1ab = annotation.df %>% dplyr::filter(gene == "Orf1ab")
Spike = annotation.df %>% dplyr::filter(gene == "Spike")
ORF3a = annotation.df %>% dplyr::filter(gene == "ORF3a")
Envelope = annotation.df %>% dplyr::filter(gene == "Envelope")
Membrane = annotation.df %>% dplyr::filter(gene == "Membrane")
ORF6 = annotation.df %>% dplyr::filter(gene == "ORF6")
ORF7a = annotation.df %>% dplyr::filter(gene == "ORF7a")
ORF8 = annotation.df %>% dplyr::filter(gene == "ORF8")
Nucleocapsid = annotation.df %>% dplyr::filter(gene == "Nucleocapsid")
ORF10 = annotation.df %>% dplyr::filter(gene == "ORF10")

# Color scheme
effect_colors = c("#e2e627", "#13a13b", "#990505")
```

I'll get a list of the low coverage samples to exclude from downstream analyses. These are any samples where at least on replicate has more than 20% of the genome covered by less than 200 reads:

```{r Filter Low Depth Sample, warning=F, message=F}

# Low quality samples - get the accessions for the samples that failed basic quality control. 
# Presently, this anything that has 20% of the genome covered by less than *200* reads. 
average.depth.df = read.table(average.depth.data, header = T) %>% 
  separate(Accession, into = c("Accession", "Replicate")) %>% 
  mutate(Run = paste0(Accession, Replicate))

low.coverage.samples = average.depth.df %>% filter(Percent < 80) %>% pull(Accession)
low.coverage.runs = average.depth.df %>% filter(Percent < 80) %>% pull(Run)

print(low.coverage.samples)
```

I'll also get a list of all of the samples that have poor concordance between replicates. 

```{r Filter Poor Concordance Samples, warning=F, message=F}

poor.concordance.samples = pysam.combined.df %>% 
  filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  select(ACCESSION, REF, POS, ALT, DP.one, DP.two, AF.one, AF.two) %>% 
  mutate(DIFF = abs(AF.one - AF.two)) %>% 
  group_by(ACCESSION) %>% 
  summarise(MEAN_DIFF = mean(DIFF)) %>% 
  filter(MEAN_DIFF >= 0.025) %>% 
  pull(ACCESSION)

poor.concordance.samples = poor.concordance.samples[which(!poor.concordance.samples %in% low.coverage.samples)]

print(poor.concordance.samples)

```

Next, I wanted to select the main variants to include in the downstream analyses. To be included, the SNPs need to meet the following criteria: 

1. They must be in samples that are 'high coverage' 80% of the genome covered by 100 reads or more. 
2. They have to be samples with high concordance between replicate allele frequencies (see `sequencing-replicate-analysis.Rmd`). 
3. Variants must be present at more than 2% frequency in both of the replicates. 
4. Variants must be covered by more than 100 reads in both of the replicates. 

Finally, the variants are combined between the replicates by averaging the allele frequency. Then, alleles present at more than 50% frequency and less than 98% frequency are flipped to their non-consensus counterparts. 

```{r Prepare Final Pysam Variants, warning=F, message=F}

# Only combine those SNPs that are in both samples
selected.variant.df = inner_join(
  filter(pysam.to.join.df, REPLICATE == "1"),
  filter(pysam.to.join.df, REPLICATE == "2"), 
  by = c("ACCESSION", "POS", "REF", "ALT", "EFFECT", "GENE", "CODON_POS", "AA_CHANGE"),
  suffix = c(".one", ".two")
  ) %>% 
  # Remove the replicate columns 
  select(!c("REPLICATE.one", "REPLICATE.two")) %>%  
  # Add in a column of SNP identifiers 
  mutate(SNP = paste0(REF, POS, ALT)) %>% 
  # Remove the low coverage samples (> 80% covered by 200+ reads)
  filter(!(ACCESSION %in% low.coverage.samples)) %>% 
  # Remove the samples with poor concordance
  filter(!(ACCESSION %in% poor.concordance.samples)) %>% 
  # Filter out the variants that don't meet the depth/af requirements 
  filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
  # Drastically speeds things up if you have a minimum depth
  filter(DP.one >= 100 & DP.two >= 100) %>% 
  # Combine AF and DP into a single value without averaging
  mutate(AF = case_when(DP.one > DP.two ~ AF.one,
                        DP.two > DP.one ~ AF.two,
                        DP.one == DP.two ~ AF.one)) %>% 
  mutate(DP = case_when(DP.one > DP.two ~ DP.one,
                        DP.two > DP.one ~ DP.two)) %>% 
  select(!ends_with(".one")) %>% 
  select(!ends_with(".two")) 

# Filp the consensus variants
fipped.consensus.df = selected.variant.df %>% 
  filter(AF >= 0.5) %>% 
  dplyr::rename(ALT = "REF", REF = "ALT") %>% 
  mutate(AF = (1 - AF)) %>% 
  filter(AF >= 0.02) %>% 
  separate(AA_CHANGE, into = c("WT", "AAPOS"),
           sep = "((?<=[A-Za-z])(?=[0-9]))") %>% 
  separate(AAPOS, into = c("AAPOS", "MUT"),
           sep = "((?<=[0-9])(?=[A-Za-z]))") %>% 
  mutate(AA_CHANGE = paste0(MUT, AAPOS, WT)) %>% 
  select(!c("MUT", "WT", "AAPOS"))

final.variant.df = rbind(filter(selected.variant.df, AF < 0.5), fipped.consensus.df) %>% 
  mutate(SNV = paste0(REF, POS, ALT))

```


```{r Prepare Final Caller Variants, warning=F, message=F}

# Final samples and mutations to analyze: 
# Remove low coverage samples
# Remove alleles below 2%
# Remove alleles with less than 100X coverage

caller.selected.variant.df = inner_join(
    filter(variant.df, REPLICATE == "1"),
    filter(variant.df, REPLICATE == "2"), 
    by = c("ACCESSION", "POS", "REF", "ALT", "EFFECT", "GENE", "AA_CHANGE", "Caller"),
    suffix = c(".one", ".two")
    ) %>% 
    # Remove the replicate columns 
    select(!c("REPLICATE.one", "REPLICATE.two")) %>%  
    # Add in a column of SNP identifiers 
    mutate(SNP = paste0(REF, POS, ALT)) %>% 
    # Remove the low coverage samples (> 80% covered by 200+ reads)
    filter(!(ACCESSION %in% low.coverage.samples)) %>% 
    # Remove the samples with poor concordance
    filter(!(ACCESSION %in% poor.concordance.samples)) %>% 
    # Filter out the variants that don't meet the depth/af requirements 
    filter(AF.one >= 0.02 & AF.two >= 0.02) %>% 
    # Drastically speeds things up if you have a minimum depth
    filter(DP.one >= 100 & DP.two >= 100) %>% 
    # Combine AF and DP into a single value without averaging
    mutate(AF = case_when(DP.one > DP.two ~ AF.one,
                          DP.two > DP.one ~ AF.two,
                          DP.one == DP.two ~ AF.one)) %>% 
    mutate(DP = case_when(DP.one > DP.two ~ DP.one,
                          DP.two > DP.one ~ DP.two)) %>% 
    select(!ends_with(".one")) %>% 
    select(!ends_with(".two")) 

# Filp the consensus variants
caller.fipped.consensus.df = caller.selected.variant.df %>% 
  filter(AF >= 0.5) %>% 
  dplyr::rename(ALT = "REF", REF = "ALT") %>% 
  mutate(AF = (1 - AF)) %>% 
  filter(AF >= 0.02) %>% 
  separate(AA_CHANGE, into = c("WT", "AAPOS"),
           sep = "((?<=[A-Za-z])(?=[0-9]))") %>% 
  separate(AAPOS, into = c("AAPOS", "MUT"),
           sep = "((?<=[0-9])(?=[A-Za-z]))") %>% 
  mutate(AA_CHANGE = paste0(MUT, AAPOS, WT)) %>% 
  select(!c("MUT", "WT", "AAPOS"))

caller.final.variant.df = rbind(filter(caller.selected.variant.df, AF < 0.5), caller.fipped.consensus.df) %>% 
  mutate(Type = case_when(nchar(REF) > 1 ~ "Del",
                          nchar(ALT) > 1 ~ "Ins",
                          nchar(REF) == 1 ~ "SNP")) %>% 
  filter(Type == "SNP") %>% 
  mutate(SNV = paste0(REF, POS, ALT))

```

## SNPs vs. Ct

Ideally, I would have a direct measurement of the number of viral transcripts in these samples to compare with iSNVs. However, all I have is the Ct values from the original samples as a benchmark. The goal here is to see if there is a correlation between Ct and intrahost variants that are called in both samples. Included iSNVs are the SNPs that are present in both replicates with more than 100X reads covering them, more than 2% frequency, no InDels, and not major variants (less than 50% frequency).

```{r Pysam iSNVs vs. Ct, warning=F, message=F, width = 7, height = 5}

ct_df = read_csv(sample_metadata) %>% 
  dplyr::rename("ACCESSION" = SpID, "Ct" = avg_ct) %>% 
  select(ACCESSION, Ct) %>% 
  mutate(ACCESSION = as.character(ACCESSION))

iSNV_count.df = final.variant.df %>% 
  group_by(ACCESSION) %>% 
  count()

left_join(iSNV_count.df, ct_df, by = "ACCESSION") %>% 
  ggplot(aes(x = Ct, y = n)) +
    geom_point(size = 3) +
    scale_x_reverse() + 
    ylab('iSNV Count') + 
    xlab('Cycle Threshold') +
    theme_bw(18)

ggsave(filename = "../../results/figures/pysam_snps_vs_ct.svg", width = 7, height = 5)

```

It's clear that Ct value doesn't correlate with the number of iSNVs called after filtering. This could have to do with (1) Ct value not being a close enough proxy for viral template number, (2) The number of iSNVs being too low to record any signal here, or (3) the Ct value being low enough across all these samples that RNA levels aren't the limiting factor in our sensitivity to call variants. 

```{r Variant Caller iSNVs vs. Ct, warning=F, message=F, width = 7, height = 5}

all.final.variants.df = rbind(select(caller.final.variant.df, !Type), select(mutate(final.variant.df, Caller = "pysam"), colnames(select(caller.final.variant.df, !Type))))

all_iSNV_count.df = all.final.variants.df %>% 
  group_by(ACCESSION, Caller) %>% 
  count()

left_join(all_iSNV_count.df, ct_df, by = "ACCESSION") %>% 
  ggplot(aes(x = Ct, y = n)) +
    geom_point(size = 3) +
    facet_wrap(~Caller) +
    scale_x_reverse() + 
    ylab('iSNV Count') + 
    xlab('Cycle Threshold') +
    theme_bw(18)

ggsave(filename = "../../results/figures/all_callers_snps_vs_ct.svg", width = 7, height = 5)

```

Although the number of "passing" variants is a little bit different between the callers, the relationship between iSNVs and Ct holds in all three methods. 

## Average Number of iSNVs

I wanted to see how many intrahost variants there are on average across samples. Just like above, these needed to pass quality controls. 

```{r Average SNPs, warning=F, message=F, width = 8, height = 5}

average.number.of.minor.variants = final.variant.df %>% 
  group_by(ACCESSION) %>% 
  count() %>% 
  pull(n) %>% 
  mean(.) %>% 
  round(.)

final.variant.df %>% 
  group_by(ACCESSION) %>% 
  count() %>% 
  ggplot(aes(x = n, y = reorder(ACCESSION, -n))) + 
    geom_bar(stat = 'identity') + 
    geom_text(aes(x = 4.5, y = "10040"), label = paste("Average ~ ", average.number.of.minor.variants), size = 6) +
    ylab('Crew Member') + 
    xlab('Number of minor variants') +
    theme_bw(18)

ggsave(filename = "../../results/figures/average_number_of_iSNVs.svg", width = 7, height = 5)

```

There is a mean of ~ 3 minor variants per sample. 

## Average Variant Frequency 

Next, I wanted to look at the average frequency of these intrahost variants, as well as their distributioon across the creq members. 

```{r MAF spectrum, warning=F, message=F, width = 7, height = 5}

final.variant.df %>% 
  ggplot(aes(x = AF)) + 
    geom_histogram(bins = 20) + 
    ylab("Count") + 
    xlab("Allele Frequency") +
    theme_bw(18) +
    theme(plot.title = element_text(hjust = 0.5)) 
  
ggsave(filename = "../../results/figures/MAF_specturm.svg", width = 6, height = 4)

```

The vast majority of intrahost variants are present at less than 10% frequency. However, there are some present at quite high frequency, even approaching 50%. These are particularly interesting if they appear elsewhere in the phylogeny of the boat samples or washington state. 

## Variant Position 

This next plot shows the positions of each variant in the genome, colored by what the mutational effect is, and labeled with the sample that it belongs to. Not a particularly useful plot, but most papers of intrahost mutations have some version of these. At least you can see that variants at low frequency are much more likely to be shared than those at high frequency. This discepecny will be adressed more specifically in `shared-variants-analysis.Rmd`. 

```{r Minor Variants Distribution, message=F, warning=F, fig.width=18, fig.height=6, fig.align="center"}

# Color scheme
effect_colors = c("#990505", "#13a13b")

label = annotation.df %>% 
  mutate(POS = (start+end)/2) %>% 
  mutate(AF = -(.075/2)) %>% 
  select(gene, POS, AF) 

final.variant.df %>% 
  mutate(EFFECT = if_else(is.na(EFFECT), "synonymous", EFFECT)) %>% 
  ggplot(aes(x = POS, y = AF, col = EFFECT, fill = EFFECT))+
    geom_point(stat="identity", size = 3, stroke = 2, alpha = 0.5) + 
    # geom_bar(stat="identity", position = position_identity(), width=.01, alpha = 0.5) +
    geom_text_repel(aes(label = ACCESSION)) +
    xlab("Position") +
    ylab("Allele Frequency") +
    scale_x_continuous(breaks=c(0,5000,10000,15000,20000,25000,30000))  +
    scale_color_manual(name = "Effect", values=effect_colors) +
    scale_fill_manual(name = "Effect", values=effect_colors) +
    annotate("rect", xmin = 0, xmax = ORF1ab[[2]]-1, ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) + 
    annotate("rect", xmin = ORF1ab[[2]], xmax = ORF1ab[[3]], ymin =  -.075, ymax = -.01 ,
               alpha = .2, fill = "dodgerblue2", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][1], y = label[[3]][1], label = label[[1]][1], family = "helvetica", size = 7) +
    annotate("rect", xmin = Spike[[2]], xmax = Spike[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "yellow3", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][2], y = label[[3]][2], label = label[[1]][2], family = "helvetica", size = 7) +
    annotate("rect", xmin = ORF3a[[2]], xmax = ORF3a[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "#FF7F00", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][3], y = label[[3]][3], label = "3a", family = "helvetica", size = 7) +
    annotate("rect", xmin = Envelope[[2]], xmax = Envelope[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill =  "#6A3D9A", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][4], y = label[[3]][4], label = "E", family = "helvetica", size = 7) +
    annotate("rect", xmin = Membrane[[2]], xmax = Membrane[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill =  "green4", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][5], y = label[[3]][5], label = "M", family = "helvetica", size = 7) +
    annotate("rect", xmin = ORF6[[2]], xmax = ORF6[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "#E31A1C", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][6], y = label[[3]][6], label = "6", family = "helvetica", size = 7) +
    annotate("rect", xmin = ORF7a[[2]], xmax = ORF7a[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "#FB9A99", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][7], y = label[[3]][7], label = "7a", family = "helvetica", size = 7) +
    annotate("rect", xmin = ORF8[[2]], xmax = ORF8[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "orchid1", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][8], y = label[[3]][8], label = "8", family = "helvetica", size = 7) +
    annotate("rect", xmin = Nucleocapsid[[2]], xmax = Nucleocapsid[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "maroon", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][9], y = label[[3]][9], label = "N", family = "helvetica", size = 7) +
    annotate("rect", xmin = ORF10[[2]], xmax = ORF10[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill =  "skyblue2", col = "black", size = .1) +
    annotate("rect", xmin = ORF10[[3]]+1, xmax = 29903, ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate(geom = "text", x = -200, y = label[[3]][1], label = "5'", family = "helvetica", size = 7) +
    annotate(geom = "text", x = 30200, y = label[[3]][1], label = "3'", family = "helvetica", size = 7) +
    #annotate(geom = "text", x = label[[2]][10], y = label[[3]][10], label = "10", family = "helvetica") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=24)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) 

ggsave("../../results/figures/variant_postition_in_genome.svg", width = 18, height = 6, dpi = 300, units = "in")

```

```{r Fixed Variants Distribution per person, message=F, warning=F, fig.width=15, fig.height=15 fig.align="center"}

label = annotation.df %>% 
  mutate(POS = (start+end)/2) %>% 
  mutate(AF = -(.075/2)) %>% 
  select(gene, POS, AF) 

selected.variant.df %>% 
  filter(AF >= 0.9) %>% 
  mutate(ACCESSION = paste0("Crew Member ", ACCESSION)) %>% 
  mutate(EFFECT = if_else(is.na(EFFECT), "synonymous", EFFECT)) %>% 
  ggplot(aes(x = POS, y = AF, col = EFFECT, fill = EFFECT))+
    geom_point(stat="identity", size = 3, stroke = 2, alpha = 0.5) + 
    geom_bar(stat="identity", position = position_identity(), width=.01, alpha = 0.5) +
    geom_text_repel(aes(label = SNP),box.padding = 1.5, size = 5, nudge_y = -.005) +
    facet_wrap(~ACCESSION, ncol = 3) +
    xlab("Position") +
    ylab("Allele Frequency") +
    scale_x_continuous(breaks=c(0,5000,10000,15000,20000,25000,30000))  +
    scale_y_continuous(breaks=c(0,0.25,0.5,0.75,1), limits = c(-0.075, 1.1))  +
    scale_color_manual(name = "Effect", values=effect_colors) +
    scale_fill_manual(name = "Effect", values=effect_colors) +
    annotate("rect", xmin = 0, xmax = ORF1ab[[2]]-1, ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) + 
    annotate("rect", xmin = ORF1ab[[2]], xmax = ORF1ab[[3]], ymin =  -.075, ymax = -.01 ,
               alpha = .2, fill = "dodgerblue2", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][1], y = label[[3]][1], label = label[[1]][1], size = 7) +
    annotate("rect", xmin = Spike[[2]], xmax = Spike[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "yellow3", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][2], y = label[[3]][2], label = label[[1]][2], size = 7) +
    annotate("rect", xmin = ORF3a[[2]], xmax = ORF3a[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "#FF7F00", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][3], y = label[[3]][3], label = "3a", size = 7) +
    annotate("rect", xmin = Envelope[[2]], xmax = Envelope[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill =  "#6A3D9A", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][4], y = label[[3]][4], label = "E", size = 7) +
    annotate("rect", xmin = Membrane[[2]], xmax = Membrane[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill =  "green4", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][5], y = label[[3]][5], label = "M", size = 7) +
    annotate("rect", xmin = ORF6[[2]], xmax = ORF6[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "#E31A1C", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][6], y = label[[3]][6], label = "6", size = 7) +
    annotate("rect", xmin = ORF7a[[2]], xmax = ORF7a[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "#FB9A99", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][7], y = label[[3]][7], label = "7a", family = "helvetica", size = 7) +
    annotate("rect", xmin = ORF8[[2]], xmax = ORF8[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "orchid1", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][8], y = label[[3]][8], label = "8", size = 7) +
    annotate("rect", xmin = Nucleocapsid[[2]], xmax = Nucleocapsid[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "maroon", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][9], y = label[[3]][9], label = "N", size = 7) +
    annotate("rect", xmin = ORF10[[2]], xmax = ORF10[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill =  "skyblue2", col = "black", size = .1) +
    annotate("rect", xmin = ORF10[[3]]+1, xmax = 29903, ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate(geom = "text", x = -400, y = label[[3]][1], label = "5'", size = 7) +
    annotate(geom = "text", x = 30400, y = label[[3]][1], label = "3'", size = 7) +
    #annotate(geom = "text", x = label[[2]][10], y = label[[3]][10], label = "10", family = "helvetica") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=24)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) 

ggsave("../../results/figures/major_variant_postition_in_genome_per_person.svg", width = 20, height = 20, dpi = 300, units = "in")

```
```{r Minor Variants Distribution per person, message=F, warning=F, fig.width=15, fig.height=15 fig.align="center"}

label = annotation.df %>% 
  mutate(POS = (start+end)/2) %>% 
  mutate(AF = -(.075/2)) %>% 
  select(gene, POS, AF) 

final.variant.df %>% 
  mutate(ACCESSION = paste0("Crew Member ", ACCESSION)) %>% 
  mutate(EFFECT = if_else(is.na(EFFECT), "synonymous", EFFECT)) %>% 
  ggplot(aes(x = POS, y = AF, col = EFFECT, fill = EFFECT))+
    geom_point(stat="identity", size = 3, stroke = 2, alpha = 0.5) + 
    geom_bar(stat="identity", position = position_identity(), width=.01, alpha = 0.5) +
    geom_text_repel(aes(label = SNP), nudge_y = 0.05, box.padding = 1, size = 5) +
    facet_wrap(~ACCESSION, ncol = 3) +
    xlab("Position") +
    ylab("Allele Frequency") +
    scale_x_continuous(breaks=c(0,5000,10000,15000,20000,25000,30000))  +
    scale_y_continuous(breaks=c(0,.1,.2,.3,.4,.5), limits = c(-0.075, .5))  +
    scale_color_manual(name = "Effect", values=effect_colors) +
    scale_fill_manual(name = "Effect", values=effect_colors) +
    annotate("rect", xmin = 0, xmax = ORF1ab[[2]]-1, ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) + 
    annotate("rect", xmin = ORF1ab[[2]], xmax = ORF1ab[[3]], ymin =  -.075, ymax = -.01 ,
               alpha = .2, fill = "dodgerblue2", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][1], y = label[[3]][1], label = label[[1]][1], size = 7) +
    annotate("rect", xmin = Spike[[2]], xmax = Spike[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "yellow3", col = "black", size = .1) +
    annotate(geom = "text", x = label[[2]][2], y = label[[3]][2], label = label[[1]][2], size = 7) +
    annotate("rect", xmin = ORF3a[[2]], xmax = ORF3a[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "#FF7F00", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][3], y = label[[3]][3], label = "3a", size = 7) +
    annotate("rect", xmin = Envelope[[2]], xmax = Envelope[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill =  "#6A3D9A", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][4], y = label[[3]][4], label = "E", size = 7) +
    annotate("rect", xmin = Membrane[[2]], xmax = Membrane[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill =  "green4", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][5], y = label[[3]][5], label = "M", size = 7) +
    annotate("rect", xmin = ORF6[[2]], xmax = ORF6[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "#E31A1C", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][6], y = label[[3]][6], label = "6", size = 7) +
    annotate("rect", xmin = ORF7a[[2]], xmax = ORF7a[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "#FB9A99", col = "black", size = .1) +
    #annotate(geom = "text", x = label[[2]][7], y = label[[3]][7], label = "7a", family = "helvetica", size = 7) +
    annotate("rect", xmin = ORF8[[2]], xmax = ORF8[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "orchid1", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][8], y = label[[3]][8], label = "8", size = 7) +
    annotate("rect", xmin = Nucleocapsid[[2]], xmax = Nucleocapsid[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "maroon", col = "black", size = .1) +
    # annotate(geom = "text", x = label[[2]][9], y = label[[3]][9], label = "N", size = 7) +
    annotate("rect", xmin = ORF10[[2]], xmax = ORF10[[3]], ymin = -.075, ymax = -.01 ,
             alpha = .2, fill =  "skyblue2", col = "black", size = .1) +
    annotate("rect", xmin = ORF10[[3]]+1, xmax = 29903, ymin = -.075, ymax = -.01 ,
             alpha = .2, fill = "grey", col = "black", size = .1) +
    annotate(geom = "text", x = -400, y = label[[3]][1], label = "5'", size = 7) +
    annotate(geom = "text", x = 30400, y = label[[3]][1], label = "3'", size = 7) +
    #annotate(geom = "text", x = label[[2]][10], y = label[[3]][10], label = "10", family = "helvetica") +
    theme_classic() +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=24)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) 

ggsave("../../results/figures/minor_variant_postition_in_genome_per_person.svg", width = 20, height = 20, dpi = 300, units = "in")

```
